<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Align the sequencing reads &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Manuals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Align the sequencing reads</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/manuals/workflow/align.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="align-the-sequencing-reads">
<h1>Align the sequencing reads<a class="headerlink" href="#align-the-sequencing-reads" title="Permalink to this headline"></a></h1>
<section id="align-input-files">
<h2>Align: input files<a class="headerlink" href="#align-input-files" title="Permalink to this headline"></a></h2>
<section id="input-file-fasta">
<h3>Input file: FASTA<a class="headerlink" href="#input-file-fasta" title="Permalink to this headline"></a></h3>
<p>Alignment requires exactly one file of reference sequences, which must
be in FASTA format.
See <a class="reference internal" href="../../formats/data/fasta.html"><span class="doc">FASTA: Reference sequences</span></a> for details.
You can use the <a class="reference internal" href="../util/fastaclean.html"><span class="doc">FASTA file cleaner</span></a> to clean the FASTA file before
alignment, if necessary.</p>
</section>
<section id="input-file-s-fastq">
<h3>Input file(s): FASTQ<a class="headerlink" href="#input-file-s-fastq" title="Permalink to this headline"></a></h3>
<p>Alignment requires one or more files of sequencing reads, which must be
in FASTQ format.
See <a class="reference internal" href="../../formats/data/fastq.html"><span class="doc">FASTQ: Sequencing Reads</span></a> for details.</p>
<section id="fastq-files-of-single-end-versus-paired-end-reads">
<h4>FASTQ files of single-end versus paired-end reads<a class="headerlink" href="#fastq-files-of-single-end-versus-paired-end-reads" title="Permalink to this headline"></a></h4>
<p>Sequencing reads can be single-end or paired-end (for more details, see
<a class="reference internal" href="../../formats/data/index.html#fastq-endedness"><span class="std std-ref">Endedness: single-end and paired-end reads</span></a>).
The SEISMIC-RNA workflow can handle both single- and paired-end reads,
provided that no FASTQ file contains a mixture of both types of reads.</p>
</section>
<section id="fastq-files-of-whole-samples-versus-demultiplexed-samples">
<h4>FASTQ files of whole samples versus demultiplexed samples<a class="headerlink" href="#fastq-files-of-whole-samples-versus-demultiplexed-samples" title="Permalink to this headline"></a></h4>
<p>FASTQ files can contain reads from a whole sample (which can possibly
come from many different reference sequences) or can be demultiplexed
before alignment so that they contain reads from one reference sequence.</p>
</section>
<section id="fastq-file-types">
<h4>FASTQ file types<a class="headerlink" href="#fastq-file-types" title="Permalink to this headline"></a></h4>
<p>When giving FASTQ files on the command line, indicate the type of reads
(single-end, paired-end interleaved, or paired-end in separate files)
and the type of content (whole-sample or demultiplexed) via the name of
the option with which you give the file.
Each option has a “long form” (starting with <code class="docutils literal notranslate"><span class="pre">--</span></code>) and a “short form”
(starting with <code class="docutils literal notranslate"><span class="pre">-</span></code> and comprising a single character), which are both
indicated in this table, separated by a <code class="docutils literal notranslate"><span class="pre">/</span></code>:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 45%" />
<col style="width: 26%" />
<col style="width: 29%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Endedness</p></th>
<th class="head"><p>Whole Sample</p></th>
<th class="head"><p>Demultiplexed</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>paired-end, separate files</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqx/-x</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqx/-X</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>paired-end, interleaved</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqy/-y</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqy/-Y</span></code></p></td>
</tr>
<tr class="row-even"><td><p>single-end</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqz/-z</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqz/-Z</span></code></p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As with all options, the long and short forms are interchangable.
For brevity, the rest of this manual uses only the short forms.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If using paired-end reads in separate FASTQ files, both files of 1st
and 2nd mates are required. See <a class="reference internal" href="index.html#fastq-pair"><span class="std std-ref">Providing one pair of paired-end FASTQ files</span></a> for instructions.</p>
</div>
</section>
<section id="providing-one-fastq-file-single-end-or-interleaved-paired-end">
<h4>Providing one FASTQ file (single-end or interleaved paired-end)<a class="headerlink" href="#providing-one-fastq-file-single-end-or-interleaved-paired-end" title="Permalink to this headline"></a></h4>
<p>For example, to align a FASTQ of single-end reads from a whole sample,
use the option <code class="docutils literal notranslate"><span class="pre">-z</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs.fa}</span></code> is the path to the FASTA file of reference sequences
and <code class="docutils literal notranslate"><span class="pre">{sample.fq.gz}</span></code> is the path to the FASTQ file of the sample.</p>
<p>For a FASTQ of paired-end, interleaved reads that were demultiplexed,
use the option <code class="docutils literal notranslate"><span class="pre">-Y</span></code> instead:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">Y</span> <span class="p">{</span><span class="n">sample</span><span class="o">/</span><span class="n">ref</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample/ref.fq.gz}</span></code> is the path to the FASTQ file containing
reads from only one reference in the sample.</p>
</section>
<section id="providing-one-pair-of-paired-end-fastq-files">
<span id="fastq-pair"></span><h4>Providing one pair of paired-end FASTQ files<a class="headerlink" href="#providing-one-pair-of-paired-end-fastq-files" title="Permalink to this headline"></a></h4>
<p>If your reads are paired-end and you have one FASTQ file containing all
1st mates and another containing all 2nd mates, then you will need to
provide both FASTQ files.
There are two methods:</p>
<ol class="arabic">
<li><p>Use the option <code class="docutils literal notranslate"><span class="pre">-x</span></code> (or <code class="docutils literal notranslate"><span class="pre">-X</span></code>) twice, once per FASTQ file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample_R1.fq.gz}</span></code> and <code class="docutils literal notranslate"><span class="pre">{sample_R2.fq.gz}</span></code> are the paths
to the FASTQ files of the 1st and 2nd mates, respectively.</p>
</li>
<li><p>Make a new directory, move both FASTQ files into that directory, and
provide the path to that directory with <code class="docutils literal notranslate"><span class="pre">-x</span></code> (or <code class="docutils literal notranslate"><span class="pre">-X</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
<span class="n">mv</span> <span class="p">{</span><span class="n">sample_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="p">{</span><span class="n">sample_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
<span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the new directory for both FASTQ files.</p>
</li>
</ol>
</section>
<section id="providing-multiple-fastq-files-or-pairs-or-paired-end-fastq-files">
<h4>Providing multiple FASTQ files or pairs or paired-end FASTQ files<a class="headerlink" href="#providing-multiple-fastq-files-or-pairs-or-paired-end-fastq-files" title="Permalink to this headline"></a></h4>
<p>There are three ways to align multiple FASTQ files (or pairs) at once:</p>
<ol class="arabic">
<li><p>Use options more than once.
The options for FASTQ files can all be given multiple times, and can
even be mixed in one command.
For example, to align a pair of paired-end FASTQ files (sample 1),
an interleaved paired-end FASTQ file (sample 2), and two single-end
FASTQ files (samples 3 and 4), use the following options:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample1_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample1_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">y</span> <span class="p">{</span><span class="n">sample2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample3</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample4</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>This method is most useful when you have a few FASTQ files.</p>
</li>
<li><p>Group FASTQ files of the same type into a directory.
For example, suppose you have 63 FASTQ files each of paired-end 1st
mates (named <code class="docutils literal notranslate"><span class="pre">sample-1_R1.fq.gz</span></code> to <code class="docutils literal notranslate"><span class="pre">sample-63_R1.fq.gz</span></code>) and
2nd mates (named analogously but with <code class="docutils literal notranslate"><span class="pre">R2</span></code>), plus demultiplexed
single-end reads from three samples (I-III) and six references (A-F)
(named <code class="docutils literal notranslate"><span class="pre">sample-I/ref-A.fq.gz</span></code> to <code class="docutils literal notranslate"><span class="pre">sample-III/ref-F.fq.gz</span></code>).
Move the separate paired-end FASTQ files into their own directory,
and the demultiplexed single-end FASTQ files into another directory,
then run alignment by passing each directory of FASTQ files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>mkdir {paired}
mv sample-*_R?.fq.gz {paired}
mkdir {dm-single}
mv sample-I* {dm-single}
seismic align {refs.fa} -x {paired} -Z {dm-single}
</pre></div>
</div>
<p>This method is most useful when you have many FASTQ files.</p>
</li>
<li><p>Combine methods 1 and 2.
Suppose you are working on two projects, have generated a set of
many FASTQ files for each project, and want to process both sets.
Currently, the FASTQ files for projects 1 and 2 are in directories
<code class="docutils literal notranslate"><span class="pre">proj1</span></code> and <code class="docutils literal notranslate"><span class="pre">proj2</span></code>, and you want to keep them separate.
You can process both directories with one command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="n">proj1</span> <span class="o">-</span><span class="n">x</span> <span class="n">proj2</span>
</pre></div>
</div>
<p>This method is most useful when you have multiple directories of
FASTQ files that you would like to keep separate.</p>
</li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a directory is given for any of the FASTQ options, then it will
be searched for FASTQ files recursively, with no limit to the depth.
Thus, the given directory can have deeply nested subdirectories, and
SEISMIC-RNA will still find and process any FASTQ files within them.</p>
</div>
</section>
</section>
</section>
<section id="align-options">
<h2>Align: options<a class="headerlink" href="#align-options" title="Permalink to this headline"></a></h2>
<section id="options-for-phred-score-encoding">
<h3>Options for Phred score encoding<a class="headerlink" href="#options-for-phred-score-encoding" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA defaults to using Phred+33 encoding for FASTQ files, which
is standard on modern Illumina sequencers.
To change the Phred score encoding, use the option <code class="docutils literal notranslate"><span class="pre">--phred-enc</span></code>.
See <a class="reference internal" href="../../formats/data/index.html#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a> for more information.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If your FASTQ files do not use the Phred+33 encoding, then you must
specify the correct Phred score encoding, or else Cutadapt and/or
Bowtie 2 can produce incorrect output or fail outright.</p>
</div>
<p>If you do not know the encoding scheme of your FASTQ files, then you may
be able to determine it by using <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> or <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> (which
runs FastQC automatically).
In the HTML report file generated by FastQC, check the “Encoding” field
in the “Basic Statisics” section:</p>
<ul class="simple">
<li><p>If the Encoding field says <code class="docutils literal notranslate"><span class="pre">Illumina</span> <span class="pre">1.0</span></code> to <code class="docutils literal notranslate"><span class="pre">1.7</span></code>, then your
FASTQ files use Phred+64 encoding (<code class="docutils literal notranslate"><span class="pre">--phred-enc</span> <span class="pre">64</span></code>).</p></li>
<li><p>If the Encoding field says <code class="docutils literal notranslate"><span class="pre">Illumina</span> <span class="pre">1.8</span></code> or greater, then your
FASTQ files use Phred+33 encoding (<code class="docutils literal notranslate"><span class="pre">--phred-enc</span> <span class="pre">33</span></code>, the default).</p></li>
<li><p>Otherwise, you will need to search elsewhere for your encoding scheme
to determine the Phred score offset.</p></li>
</ul>
</section>
<section id="options-for-quality-assessment-with-fastqc">
<h3>Options for quality assessment with FastQC<a class="headerlink" href="#options-for-quality-assessment-with-fastqc" title="Permalink to this headline"></a></h3>
<p>By default, each FASTQ file is processed with <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, both before and
after trimming, in order to find any potential problems.
FastQC can be disabled with the flag <code class="docutils literal notranslate"><span class="pre">--no-fastqc</span></code>.
To enable automatic extraction of the zipped output files from FastQC,
add the flag <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code>.</p>
</section>
<section id="options-for-trimming-with-cutadapt">
<h3>Options for trimming with Cutadapt<a class="headerlink" href="#options-for-trimming-with-cutadapt" title="Permalink to this headline"></a></h3>
<p>By default, each FASTQ file and pair of mated FASTQ files is trimmed for
adapters and low-quality bases using <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">Cutadapt</a>. To disable adapter
trimming, add the flag <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code>.</p>
<section id="adapter-trimming">
<h4>Adapter trimming<a class="headerlink" href="#adapter-trimming" title="Permalink to this headline"></a></h4>
<p>By default, SEISMIC-RNA uses the standard, minimal adapter sequences for
Illumina sequencing runs for both read 1 and (if paired-end) read 2:</p>
<ul class="simple">
<li><p>5’: <code class="docutils literal notranslate"><span class="pre">GCTCTTCCGATCT</span></code></p></li>
<li><p>3’: <code class="docutils literal notranslate"><span class="pre">AGATCGGAAGAGC</span></code></p></li>
</ul>
<p>To use another adapter, type its sequence after the appropriate option:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 23%" />
<col style="width: 23%" />
<col style="width: 54%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Side</p></th>
<th class="head"><p>Read</p></th>
<th class="head"><p>Option</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>5’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>5’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g2</span></code></p></td>
</tr>
<tr class="row-even"><td><p>3’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>3’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a2</span></code></p></td>
</tr>
</tbody>
</table>
</section>
<section id="quality-trimming">
<h4>Quality trimming<a class="headerlink" href="#quality-trimming" title="Permalink to this headline"></a></h4>
<p>Base calls on either end of a read that fall below a minimum Phred score
quality are trimmed with Cutadapt.
The default minimum quality is 25, which corresponds to a probability of
1 - 10 <sup>-2.5</sup> = 0.997 that the base call is correct.
(See <a class="reference internal" href="../../formats/data/index.html#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a> for more details).
To change the quality threshold, use the option <code class="docutils literal notranslate"><span class="pre">--min-phred</span></code>.</p>
</section>
<section id="dark-cycle-trimming">
<h4>Dark cycle trimming<a class="headerlink" href="#dark-cycle-trimming" title="Permalink to this headline"></a></h4>
<p>On some Illumina sequencers (e.g. NextSeq, iSeq), the probes used to
detect G bases emit no light. Hence, these instruments will label a base
call as a G if it appears dark. If sequencing reaches the end of a read,
then there will be no more bases to sequence, so every cycle thereafter
will be dark, causing a string of Gs to be added to the 3’ end of the
read. Using the option <code class="docutils literal notranslate"><span class="pre">--cut-nextseq</span></code> tells Cutadapt to trim off any
high-quality G bases from the 3’ end of each read. This may improve the
alignment (especially in end-to-end mode) but also removes real G bases
from the 3’ ends of reads (since they cannot be distinguished from any
artefactual G bases).</p>
</section>
<section id="additional-options-for-cutadapt">
<h4>Additional options for Cutadapt<a class="headerlink" href="#additional-options-for-cutadapt" title="Permalink to this headline"></a></h4>
<p>See <a class="reference internal" href="../../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can
use with Cutadapt, and the <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/reference.html">Cutadapt reference guide</a> for details on
each of these options.
These options should suffice for most users.
If you require a more customized adapter trimming workflow, you can trim
your FASTQ files outside of SEISMIC-RNA, then perform alignment within
SEISMIC-RNA using the option <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code> to disable additional adapter
trimming.</p>
</section>
</section>
<section id="options-for-alignment-with-bowtie-2">
<h3>Options for alignment with Bowtie 2<a class="headerlink" href="#options-for-alignment-with-bowtie-2" title="Permalink to this headline"></a></h3>
<section id="bowtie-2-index-files-automatic-and-pre-built">
<h4>Bowtie 2 index files: automatic and pre-built<a class="headerlink" href="#bowtie-2-index-files-automatic-and-pre-built" title="Permalink to this headline"></a></h4>
<p>SEISMIC-RNA searches for a set of Bowtie 2 index files in the directory
where the FASTA file is.
A valid Bowtie 2 index comprises six files, all with the same name as
the FASTA file, with the extensions <code class="docutils literal notranslate"><span class="pre">.1.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.2.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.3.bt2</span></code>,
<code class="docutils literal notranslate"><span class="pre">.4.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.rev.1.bt2</span></code>, and <code class="docutils literal notranslate"><span class="pre">.rev.2.bt2</span></code>.
If this set of files exists, then SEISMIC-RNA uses it as the index.
Otherwise, it calls <code class="docutils literal notranslate"><span class="pre">bowtie2-build</span></code> to build an index in a temporary
directory, which is deleted after alignment finishes.
Indexing a small FASTA file takes several seconds; thus, it is generally
more convenient to build a temporary .
However, indexing a very large FASTA (e.g. a whole transcriptome) can
take hours, so it is advantageous to pre-build your index in the same
directory as the FASTA file, to save time in case you need to align more
than once.
You can pre-build an index with this command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bowtie2</span><span class="o">-</span><span class="n">build</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="p">{</span><span class="n">refs</span><span class="p">}</span>
</pre></div>
</div>
<p>replacing <code class="docutils literal notranslate"><span class="pre">{refs}</span></code> with the path to and name of your FASTA file.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer">Bowtie 2 Indexer manual</a> for more details.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If SEISMIC-RNA finds a pre-built Bowtie 2 index, then it does <em>not</em>
verify that the index was actually built from the FASTA file of the
same name.
You must verify this yourself if using a pre-built index.
You can assume the index is correct if you build it using the above
command and avoid modifying or replacing the FASTA and index files.</p>
</div>
</section>
<section id="alignment-modes-local-and-end-to-end">
<h4>Alignment modes: local and end-to-end<a class="headerlink" href="#alignment-modes-local-and-end-to-end" title="Permalink to this headline"></a></h4>
<p>During alignment, Bowtie 2 can either align the entirety of each read
(end-to-end mode) or find and align only the section of the read that
yields the best alignment score (local mode).
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#end-to-end-alignment-versus-local-alignment">description of alignment modes in Bowtie 2</a> for more details.</p>
<p>Generally, end-to-end mode yields spurious mutations (false positives)
at the ends of reads if the reads contain artifacts such as low-quality
base calls or untrimmed or improperly trimmed adapters.
Conversely, local mode misses real mutations (false negatives) within
several nucleotides of the ends of reads because such mutations are not,
by definition, part of the best local alignment.</p>
<p>Concerning RNA mutational profiling, false positives are generally much
more problematic than false negatives, so SEISMIC-RNA uses local mode
(<code class="docutils literal notranslate"><span class="pre">--bt2-local</span></code>) by default.
Use end-to-end mode (<code class="docutils literal notranslate"><span class="pre">--bt2-end-to-end</span></code>) only if you have a compelling
reason to do so (e.g. if it is essential to detect mutations at the ends
of reads) and only after carefully trimming any extraneous sequences
from the ends of the reads.</p>
</section>
<section id="options-for-aligning-paired-end-reads">
<h4>Options for aligning paired-end reads<a class="headerlink" href="#options-for-aligning-paired-end-reads" title="Permalink to this headline"></a></h4>
<p>Bowtie 2 considers paired-end reads to align “concordantly” when their
relative positions match expectations and “discordantly” otherwise. See
the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#concordant-pairs-match-pair-expectations-discordant-pairs-dont">Bowtie 2 manual for details on concordant/discordant alignments</a>.
By default, SEISMIC-RNA treats only concordantly aligning pairs as valid
alignments. To also treat discordant pairs as valid alignments, use the
flag <code class="docutils literal notranslate"><span class="pre">--bt2-discordant</span></code>.</p>
<p>Several options control which types of alignments are concordant. First,
the expected orientation of paired mates is set using <code class="docutils literal notranslate"><span class="pre">--bt2-orient</span></code>.
It can be <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the 5’-most mate is forward, the 3’-most is reversed),
<code class="docutils literal notranslate"><span class="pre">rf</span></code> (the 5’-most mate is reversed, the 3’-most is forward), or <code class="docutils literal notranslate"><span class="pre">ff</span></code>
(both mates are forward). The default is <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the most common type).
Second, the mates may <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mates-can-overlap-contain-or-dovetail-each-other">overlap partially or completely, or dovetail</a>.
By default, overlaps (partial and complete) are considered concordant,
and dovetailing is considered discordant. The flag <code class="docutils literal notranslate"><span class="pre">--bt2-no-contain</span></code>
treats as discordant pairs where one mate completely overlaps the other,
while <code class="docutils literal notranslate"><span class="pre">--bt2-dovetail</span></code> treats dovetailed pairs as concordant. Pairs
that overlap partially are always considered concordant in SEISMIC-RNA.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The flags <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]contain</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]dovetail</span></code> choose
whether to treat these types of overlaps as concordant (yes) or
discordant (no). If they are treated as discordant, then the flag
<code class="docutils literal notranslate"><span class="pre">--bt2-[no-]discordant</span></code> determines whether they are considered
valid alignments (yes) or invalid (no).</p>
</div>
<p>The option <code class="docutils literal notranslate"><span class="pre">--bt2-mixed</span></code> enables <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a> wherein, for pairs that
fail to produce a valid paired-end alignment, Bowtie 2 attempts to align
each mate individually (as if it were a single-end read).</p>
</section>
<section id="options-for-filtering-alignments">
<h4>Options for filtering alignments<a class="headerlink" href="#options-for-filtering-alignments" title="Permalink to this headline"></a></h4>
<p>Alignments can be filtered by <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">alignment score</a> and <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">mapping quality</a>,
which are distinct properties.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">Alignment score</a> measures how well a read aligns to a given location
in the reference.
It is calculated from the number of matches, substitutions, and gaps
using the score parameters.
The minimum alignment scores for local and end-to-end modes can be set
using <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-loc</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-e2e</span></code>, respectively.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#valid-alignments-meet-or-exceed-the-minimum-score-threshold">section of the Bowtie 2 manual on alignment scores</a> for advice
on setting this parameter.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">Mapping quality</a> measures how unique an alignment is: high quality if
the read aligns with a high score to exactly one location, low quality
if it aligns with similar scores to multiple locations in the reference.
The default minimum quality is 25, which corresponds to a confidence of
1 - 10 <sup>-2.5</sup> = 0.997 that the read has aligned correctly.
To change the quality threshold, use the option <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>.
For those searching for this option in Bowtie 2, you will not find it.
Instead, reads with insufficient mapping quality are filtered out after
alignment using the <a class="reference external" href="https://www.htslib.org/doc/samtools-view.html">view command in Samtools</a>.</p>
</section>
<section id="additional-options-for-bowtie-2">
<h4>Additional options for Bowtie 2<a class="headerlink" href="#additional-options-for-bowtie-2" title="Permalink to this headline"></a></h4>
<p>See <a class="reference internal" href="../../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can
use with Bowtie 2, and the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">Bowtie 2 manual</a> for details on each of
these options.
These options should suffice for most users.
If you require a more customized alignment workflow, you can align your
your FASTQ files outside of SEISMIC-RNA, then pass the resulting XAM
files into SEISMIC-RNA at the step <a class="reference internal" href="relate.html#step-relate"><span class="std std-ref">Relate each read to every reference position</span></a>.</p>
</section>
</section>
<section id="formatting-alignment-maps">
<span id="bam-vs-cram"></span><h3>Formatting alignment maps<a class="headerlink" href="#formatting-alignment-maps" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA can output alignment map files in either BAM or CRAM format.
For details on these file formats, see <a class="reference internal" href="../../formats/data/xam.html"><span class="doc">SAM, BAM, and CRAM: Alignment Maps</span></a>.
The default is CRAM format (option <code class="docutils literal notranslate"><span class="pre">--cram</span></code>); BAM format is enabled
using the option <code class="docutils literal notranslate"><span class="pre">--bam</span></code>.</p>
<p>Alignment maps in CRAM format are smaller than their BAM counterparts,
and hence better suited to long-term storage.
However, the better compression of CRAM files comes at three costs:</p>
<ul class="simple">
<li><p>A CRAM file must be accompanied by a FASTA file storing the sequence
of every reference that appears in the header of the CRAM file.
A CRAM file stores only the relative path to its FASTA file, not the
sequence information, which enables the CRAM file to be much smaller
than it would be if it did need to contain its own sequences.
Because the FASTA file existed before and during the alignment, having
this FASTA file accompany the CRAM file usually incurs no extra cost.
However, moving or deleting the FASTA will break the CRAM file.
As a safeguard against this fragility, SEISMIC-RNA keeps a copy of the
original FASTA file in the same directory as the output CRAM file.
Creating an actual copy would require more storage space and defeat
the purpose of CRAM’s smaller file size, so SEISMIC-RNA actually makes
a <a class="reference external" href="https://en.wikipedia.org/wiki/Hard_link">hard link</a> – not a copy – which requires minimal extra space.
In some circumstances, making a hard link can fail, in which case
SEISMIC-RNA will resort to copying the FASTA file instead.</p></li>
<li><p>Reading and writing CRAM files is slower than for BAM files due to the
extra effort needed for compressing and decompressing CRAM files.</p></li>
<li><p>In the CIGAR strings, distinction between reference matches (<code class="docutils literal notranslate"><span class="pre">=</span></code>)
and substitutions (<code class="docutils literal notranslate"><span class="pre">X</span></code>) is lost upon compressing to CRAM format.
Thus, <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">relate</span></code> must perform extra work to determine if each
non-gapped position is a match or substitution, which makes it run
more slowly than it would if the distinction had been preserved.</p></li>
</ul>
<p>In general, use CRAM format if minimizning the size of your alignment
map files is a priority, especially for long-term storage.
Use BAM format to make the <code class="docutils literal notranslate"><span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">relate</span></code> steps run faster,
and to increase the robustness of the output files (because BAM files
are self-contained, while CRAM files will break without the FASTA file
that accompanies them).</p>
</section>
</section>
<section id="align-output-files">
<h2>Align: output files<a class="headerlink" href="#align-output-files" title="Permalink to this headline"></a></h2>
<section id="output-files-fastqc-reports">
<h3>Output files: FastQC reports<a class="headerlink" href="#output-files-fastqc-reports" title="Permalink to this headline"></a></h3>
<p>If FastQC is run, then it outputs files to <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc</span></code>, where
<code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory (<code class="docutils literal notranslate"><span class="pre">--out-dir</span></code>) and <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is
the name of the sample.
The directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc/init</span></code> is always created and contains
FastQC reports of the initial FASTQ files.
If adapter/quality trimming was run, <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc/trim</span></code> is also
created for FastQC reports of the trimmed FASTQ files.</p>
<p>In each directory (<code class="docutils literal notranslate"><span class="pre">init</span></code> and <code class="docutils literal notranslate"><span class="pre">trim</span></code>), FastQC writes two files for
each FASTQ file: <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.html</span></code> and <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.zip</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{fq_name}</span></code> is the name of the original FASTQ file up to the
file extension.
If the option <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code> is given, then FastQC will also unzip
<code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.zip</span></code> to the directory <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc</span></code>.
For details on these outputs, see the documentation for <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>.</p>
</section>
<section id="output-files-alignment-maps">
<h3>Output files: alignment maps<a class="headerlink" href="#output-files-alignment-maps" title="Permalink to this headline"></a></h3>
<p>The most important outputs of <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> are alignment map files.
SEISMIC-RNA can output alignment maps in either BAM or CRAM format.
For a comparison of these formats, see <a class="reference internal" href="index.html#bam-vs-cram"><span class="std std-ref">Formatting alignment maps</span></a>.</p>
<p>Alignment maps store the location in the reference sequence to which
each read aligned, as well as the Phred quality scores, mapping quality,
and mutated positions.
SEISMIC-RNA outputs alignment maps where every read aligns to the same
reference (although this is not a restriction outside of SEISMIC-RNA).
Each alignment map is written to <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/align/{ref}.{xam}</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory (<code class="docutils literal notranslate"><span class="pre">--out-dir</span></code>), <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is
the name of the sample from which the reads came, <code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is the name
of the reference to which the reads aligned, and <code class="docutils literal notranslate"><span class="pre">{xam}</span></code> is the file
extension (<code class="docutils literal notranslate"><span class="pre">.bam</span></code> or <code class="docutils literal notranslate"><span class="pre">.cram</span></code>, depending on the selected format).</p>
</section>
<section id="examining-unaligned-reads-pairs">
<h3>Examining unaligned reads/pairs<a class="headerlink" href="#examining-unaligned-reads-pairs" title="Permalink to this headline"></a></h3>
<p>For some reasons, including troubleshooting low alignment rates, it can
be helpful to examine the reads that did not align. Depending on the</p>
</section>
</section>
<section id="troubleshooting-alignment">
<h2>Troubleshooting alignment<a class="headerlink" href="#troubleshooting-alignment" title="Permalink to this headline"></a></h2>
<section id="alignment-rate-is-low">
<span id="low-align-rate"></span><h3>Alignment rate is low<a class="headerlink" href="#alignment-rate-is-low" title="Permalink to this headline"></a></h3>
<p>If the percent of reads aligning to the reference is less than expected,
then try the following steps (in this order):</p>
<ol class="arabic simple">
<li><p>Ensure you are using Bowtie version 2.5.1 or later (version 2.5.0
has a bug that affects alignment rate). You can check the version by
running <code class="docutils literal notranslate"><span class="pre">bowtie2</span> <span class="pre">--version</span> <span class="pre">|</span> <span class="pre">head</span> <span class="pre">-n</span> <span class="pre">1</span></code>.</p></li>
<li><p>Double check that the FASTA has the correct reference sequence(s)
and that, if the Bowtie 2 index was pre-built before the align step,
that the correct FASTA file was used.</p></li>
<li><p>Reads that failed to align with Bowtie2 are written to the file
<code class="docutils literal notranslate"><span class="pre">out/{sample}/align/unaligned.</span></code> Open the SAM file,
process several unaligned reads randomly, and use <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&amp;PAGE_TYPE=BlastSearch&amp;LINK_LOC=blasthome">BLAST</a> to discern
their origins, which can help in deducing what went wrong.</p></li>
</ol>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>