<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Align: Trim FASTQ files and align them to reference sequences &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Relate: Compute relationships between references and aligned reads" href="relate.html" />
    <link rel="prev" title="Demultiplex: Split multiplexed FASTQ files by their barcodes" href="demult.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Align: Trim FASTQ files and align them to reference sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#align-input-files">Align: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#align-settings">Align: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#align-output-files">Align: Output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#align-troubleshoot-and-optimize">Align: Troubleshoot and optimize</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate: Compute relationships between references and aligned reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="pool.html">Pool: Merge samples (vertically) from the Relate step</a></li>
<li class="toctree-l3"><a class="reference internal" href="mask.html">Mask: Define mutations and sections to filter reads and positions</a></li>
<li class="toctree-l3"><a class="reference internal" href="cluster.html">Cluster: Infer alternative structures by clustering reads’ mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="join.html">Join: Merge sections (horizontally) from the Mask or Cluster step</a></li>
<li class="toctree-l3"><a class="reference internal" href="table.html">Table: Count mutations for each read and position</a></li>
<li class="toctree-l3"><a class="reference internal" href="fold.html">Fold: Predict RNA secondary structures using mutation rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="graph.html">Graph: Plot data from tables and/or structures and compare samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3"><a class="reference internal" href="wf.html">Workflow: Run all steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lists.html">List Positions Matching Criteria</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adjclust.html">Add/Delete Orders to/from an Already-Clustered Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Align: Trim FASTQ files and align them to reference sequences</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/align.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="align-trim-fastq-files-and-align-them-to-reference-sequences">
<h1>Align: Trim FASTQ files and align them to reference sequences<a class="headerlink" href="#align-trim-fastq-files-and-align-them-to-reference-sequences" title="Link to this heading"></a></h1>
<section id="align-input-files">
<h2>Align: Input files<a class="headerlink" href="#align-input-files" title="Link to this heading"></a></h2>
<section id="align-input-file-reference-sequences">
<h3>Align input file: Reference sequences<a class="headerlink" href="#align-input-file-reference-sequences" title="Link to this heading"></a></h3>
<p>You need one file of reference sequences in FASTA format (for details on this
format, see <a class="reference internal" href="../../formats/data/fasta.html"><span class="doc">FASTA: Reference sequences</span></a>).
If your file has characters or formatting incompatible with SEISMIC-RNA, then
you can fix it using the <a class="reference internal" href="../cleanfa.html"><span class="doc">Clean FASTA Files</span></a> tool.</p>
</section>
<section id="align-input-file-read-sequences">
<h3>Align input file: Read sequences<a class="headerlink" href="#align-input-file-read-sequences" title="Link to this heading"></a></h3>
<p>Your read sequences must be in FASTQ format (see <a class="reference internal" href="../../formats/data/fastq.html"><span class="doc">FASTQ: Sequencing Reads</span></a>
for details on this format).</p>
<section id="reads-can-be-single-end-or-paired-end">
<h4>Reads can be single-end or paired-end<a class="headerlink" href="#reads-can-be-single-end-or-paired-end" title="Link to this heading"></a></h4>
<p>You can align FASTQ files of single- and paired-end reads with SEISMIC-RNA.
(For definitions of single- and paired-end reads, see <a class="reference internal" href="../../formats/data/fastq.html#fastq-endedness"><span class="std std-ref">Endedness: single-end and paired-end reads</span></a>).
SEISMIC-RNA requires that single- and paired-end reads not be mixed within one
file, but it can accept different types of reads in separate FASTQ files.</p>
</section>
<section id="reads-can-come-from-whole-or-demultiplexed-samples">
<h4>Reads can come from whole or demultiplexed samples<a class="headerlink" href="#reads-can-come-from-whole-or-demultiplexed-samples" title="Link to this heading"></a></h4>
<p>You can align FASTQ files that come from a whole sample (possibly containing
multiple RNA sequences) or that have been demultiplexed before alignment so
that they contain reads from only one RNA sequence.
For more information on demultiplexing and how to perform it if needed, see
<a class="reference internal" href="demult.html"><span class="doc">Demultiplex: Split multiplexed FASTQ files by their barcodes</span></a>.</p>
</section>
<section id="how-to-specify-the-endedness-and-source-of-reads">
<h4>How to specify the endedness and source of reads<a class="headerlink" href="#how-to-specify-the-endedness-and-source-of-reads" title="Link to this heading"></a></h4>
<p>Specify the endedness (single-end, paired-end interleaved, or paired-end in
separate files) and source of the reads (whole or demultiplexed sample) in a
FASTQ file by giving the file path after the appropriate option:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Endedness</p></th>
<th class="head"><p>Whole Sample</p></th>
<th class="head"><p>Demultiplexed</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>paired-end, separate files</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqx</span></code> (<code class="docutils literal notranslate"><span class="pre">-x</span></code>)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqx</span></code> (<code class="docutils literal notranslate"><span class="pre">-X</span></code>)</p></td>
</tr>
<tr class="row-odd"><td><p>paired-end, interleaved</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqy</span></code> (<code class="docutils literal notranslate"><span class="pre">-y</span></code>)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqy</span></code> (<code class="docutils literal notranslate"><span class="pre">-Y</span></code>)</p></td>
</tr>
<tr class="row-even"><td><p>single-end</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqz</span></code> (<code class="docutils literal notranslate"><span class="pre">-z</span></code>)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqz</span></code> (<code class="docutils literal notranslate"><span class="pre">-Z</span></code>)</p></td>
</tr>
</tbody>
</table>
<p>For example, you could type <code class="docutils literal notranslate"><span class="pre">-z</span> <span class="pre">{myfile}</span></code> to input a FASTQ file of single-end
reads from a whole sample or <code class="docutils literal notranslate"><span class="pre">-X</span> <span class="pre">{myfile1}</span> <span class="pre">-X</span> <span class="pre">{myfile2}</span></code> to input two FASTQ
files of paired-end reads from a demultiplexed sample, replacing <code class="docutils literal notranslate"><span class="pre">{myfile}</span></code>
with the actual path(s) of your file(s).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you have paired-end reads in separate FASTQ files (<code class="docutils literal notranslate"><span class="pre">-x</span></code> or <code class="docutils literal notranslate"><span class="pre">-X</span></code>),
then you must give both files of 1st and 2nd mates.
See <a class="reference internal" href="#fastq-pair"><span class="std std-ref">How to align a pair of FASTQ files (paired-end reads in separate files)</span></a> for instructions.</p>
</div>
</section>
<section id="how-to-align-one-fastq-file-single-end-or-interleaved-paired-end-reads">
<h4>How to align one FASTQ file (single-end or interleaved paired-end reads)<a class="headerlink" href="#how-to-align-one-fastq-file-single-end-or-interleaved-paired-end-reads" title="Link to this heading"></a></h4>
<p>To align a FASTQ of single-end reads from a whole sample, use <code class="docutils literal notranslate"><span class="pre">-z</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs.fa}</span></code> is the path to your FASTA file of reference sequences and
<code class="docutils literal notranslate"><span class="pre">{sample.fq.gz}</span></code> is the path to your FASTQ file of the sample.</p>
<p>For a FASTQ of paired-end, interleaved reads that were demultiplexed, use <code class="docutils literal notranslate"><span class="pre">-Y</span></code>
instead:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">Y</span> <span class="p">{</span><span class="n">sample</span><span class="o">/</span><span class="n">ref</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample/ref.fq.gz}</span></code> is the path to your FASTQ file containing reads
from only one reference in the sample.</p>
</section>
<section id="how-to-align-a-pair-of-fastq-files-paired-end-reads-in-separate-files">
<span id="fastq-pair"></span><h4>How to align a pair of FASTQ files (paired-end reads in separate files)<a class="headerlink" href="#how-to-align-a-pair-of-fastq-files-paired-end-reads-in-separate-files" title="Link to this heading"></a></h4>
<p>If your reads are paired-end and you have one FASTQ file containing all 1st
mates and another containing all 2nd mates, then you will need to provide both
FASTQ files.
There are two methods:</p>
<ol class="arabic">
<li><p>Use <code class="docutils literal notranslate"><span class="pre">-x</span></code>/<code class="docutils literal notranslate"><span class="pre">-X</span></code> twice, once per FASTQ file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample_R1.fq.gz}</span></code> and <code class="docutils literal notranslate"><span class="pre">{sample_R2.fq.gz}</span></code> are the paths to your
FASTQ files of the 1st and 2nd mates, respectively.</p>
</li>
<li><p>Make a new directory, move both FASTQ files into that directory, and provide
the path to that directory with <code class="docutils literal notranslate"><span class="pre">-x</span></code>/<code class="docutils literal notranslate"><span class="pre">-X</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
<span class="n">mv</span> <span class="p">{</span><span class="n">sample_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="p">{</span><span class="n">sample_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
<span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the new directory for both FASTQ files.</p>
</li>
</ol>
</section>
<section id="how-to-align-multiple-fastq-files-or-pairs-of-paired-end-fastq-files">
<h4>How to align multiple FASTQ files or pairs of paired-end FASTQ files<a class="headerlink" href="#how-to-align-multiple-fastq-files-or-pairs-of-paired-end-fastq-files" title="Link to this heading"></a></h4>
<p>There are three ways to align multiple FASTQ files (or pairs thereof):</p>
<ul>
<li><p><strong>Use options more than once.</strong>
You can repeat any of <code class="docutils literal notranslate"><span class="pre">-x</span></code>/<code class="docutils literal notranslate"><span class="pre">-y</span></code>/<code class="docutils literal notranslate"><span class="pre">-z</span></code>/<code class="docutils literal notranslate"><span class="pre">-X</span></code>/<code class="docutils literal notranslate"><span class="pre">-Y</span></code>/<code class="docutils literal notranslate"><span class="pre">-Z</span></code>, as well as
mix them in one command.
For example, to align one pair of paired-end FASTQ files (sample 1), one
interleaved paired-end FASTQ file (sample 2), and two single-end FASTQ files
(samples 3 and 4), you could type</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample1_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample1_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">y</span> <span class="p">{</span><span class="n">sample2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample3</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample4</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>This method is most useful when you have a small number of FASTQ files.</p>
</li>
<li><p><strong>Group FASTQ files of the same type into a directory.</strong>
Suppose you have 63 pairs of FASTQ files, with the files of mate 1s named
<code class="docutils literal notranslate"><span class="pre">sample-1_R1.fq.gz</span></code> to <code class="docutils literal notranslate"><span class="pre">sample-63_R1.fq.gz</span></code> and the files of mate 2s named
<code class="docutils literal notranslate"><span class="pre">sample-1_R2.fq.gz</span></code> to <code class="docutils literal notranslate"><span class="pre">sample-63_R2.fq.gz</span></code>; plus demultiplexed single-end
reads from three samples (I-III) and six references (A-F), named
<code class="docutils literal notranslate"><span class="pre">sample-I/ref-A.fq.gz</span></code> to <code class="docutils literal notranslate"><span class="pre">sample-III/ref-F.fq.gz</span></code>).
You can align all of them with one command if you move the whole-sample,
paired-end FASTQ files into their own directory, and the demultiplexed,
single-end FASTQ files into another directory, and then give each directory
after the appropriate options (<code class="docutils literal notranslate"><span class="pre">-x</span></code> and <code class="docutils literal notranslate"><span class="pre">-Z</span></code>, respectively):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>mkdir {paired}
mv sample-*_R?.fq.gz {paired}
mkdir {dm-single}
mv sample-I* {dm-single}
seismic align {refs.fa} -x {paired} -Z {dm-single}
</pre></div>
</div>
<p>This method is most useful when you have many FASTQ files.</p>
</li>
<li><p><strong>Combine the first two methods.</strong>
Suppose you are working on two projects, have generated a set of many FASTQ
files for each project, and want to process both sets.
Currently, the FASTQ files for projects 1 and 2 are in directories <code class="docutils literal notranslate"><span class="pre">proj1</span></code>
and <code class="docutils literal notranslate"><span class="pre">proj2</span></code>, and you want to keep them separate.
You can process both directories with one command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="n">proj1</span> <span class="o">-</span><span class="n">x</span> <span class="n">proj2</span>
</pre></div>
</div>
<p>This method is most useful when you have multiple directories of FASTQ files
that you would like to keep separate.</p>
</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you give a directory for any of the FASTQ options, then SEISMIC-RNA will
search for FASTQ files recursively, with no limit to the depth.</p>
</div>
</section>
</section>
</section>
<section id="align-settings">
<h2>Align: Settings<a class="headerlink" href="#align-settings" title="Link to this heading"></a></h2>
<section id="align-setting-quality-score-encoding">
<h3>Align setting: Quality score encoding<a class="headerlink" href="#align-setting-quality-score-encoding" title="Link to this heading"></a></h3>
<p>Your FASTQ files may encode quality scores in several schemes (for details, see
<a class="reference internal" href="../../formats/data/fastq.html#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a>).
Modern Illumina sequencers use Phred+33 encoding, the default in SEISMIC-RNA.
To change the quality score encoding, use <code class="docutils literal notranslate"><span class="pre">--phred-enc</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If your FASTQ files do not use the Phred+33 encoding, then you must
specify the correct Phred score encoding, or else Cutadapt and/or
Bowtie 2 can produce incorrect output or fail outright.</p>
</div>
<p>If you do not know the encoding scheme of your FASTQ files, then you can process
them with <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> and check the “Encoding” field in the “Basic Statisics” part
of the FastQC report:</p>
<ul class="simple">
<li><p>If the Encoding field says <code class="docutils literal notranslate"><span class="pre">Illumina</span> <span class="pre">1.0</span></code> to <code class="docutils literal notranslate"><span class="pre">1.7</span></code>, then your FASTQ files
use Phred+64 encoding (<code class="docutils literal notranslate"><span class="pre">--phred-enc</span> <span class="pre">64</span></code>).</p></li>
<li><p>If the Encoding field says <code class="docutils literal notranslate"><span class="pre">Illumina</span> <span class="pre">1.8</span></code> or greater, then your FASTQ files
use Phred+33 encoding (<code class="docutils literal notranslate"><span class="pre">--phred-enc</span> <span class="pre">33</span></code>, the default).</p></li>
<li><p>Otherwise, you will need to search elsewhere for your encoding scheme.</p></li>
</ul>
</section>
<section id="align-setting-quality-assessment-with-fastqc">
<h3>Align setting: Quality assessment with FastQC<a class="headerlink" href="#align-setting-quality-assessment-with-fastqc" title="Link to this heading"></a></h3>
<p>To check the quality of your FASTQ files, SEISMIC-RNA runs <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> by default.
To disable FastQC, use <code class="docutils literal notranslate"><span class="pre">--no-fastqc</span></code>.
You can also enable automatic unzipping of the zipped output files from FastQC
with <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code>.</p>
</section>
<section id="align-setting-trimming-reads-with-cutadapt">
<h3>Align setting: Trimming reads with Cutadapt<a class="headerlink" href="#align-setting-trimming-reads-with-cutadapt" title="Link to this heading"></a></h3>
<p>To trim adapters and low-quality base calls before alignment, SEISMIC-RNA runs
<a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">Cutadapt</a> by default.
To disable trimming, use <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code>.</p>
<section id="how-to-trim-adapter-sequences">
<h4>How to trim adapter sequences<a class="headerlink" href="#how-to-trim-adapter-sequences" title="Link to this heading"></a></h4>
<p>Your reads may contain unwanted adapters (especially near their 3’ ends), which
can cause problems such as misalignment (alignment to the wrong location).
Your adapter sequences depend on how your samples were prepared for sequencing
(i.e. on your library prep kit) and on your sequencing platform.
Since Illumina sequencers are the most widely used for mutational profiling,
SEISMIC-RNA defaults to the standard, minimal adapter sequences for Illumina
for both read 1 and (if paired-end) read 2:</p>
<ul class="simple">
<li><p>5’: <code class="docutils literal notranslate"><span class="pre">GCTCTTCCGATCT</span></code></p></li>
<li><p>3’: <code class="docutils literal notranslate"><span class="pre">AGATCGGAAGAGC</span></code></p></li>
</ul>
<p>If your samples have other adapters, then you can specify their sequences using</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Side</p></th>
<th class="head"><p>Read</p></th>
<th class="head"><p>Option</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>5’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>5’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g2</span></code></p></td>
</tr>
<tr class="row-even"><td><p>3’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>3’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a2</span></code></p></td>
</tr>
</tbody>
</table>
</section>
<section id="how-to-trim-low-quality-base-calls">
<span id="quality-trimming"></span><h4>How to trim low-quality base calls<a class="headerlink" href="#how-to-trim-low-quality-base-calls" title="Link to this heading"></a></h4>
<p>Your reads may contain low-quality base calls (especially near their 3’ ends),
which can cause misalignment and excessive mutations.
By default, SEISMIC-RNA trims base calls with quality scores less than 25, which
corresponds to a probability of 10<sup>-2.5</sup> = 0.3% that the base call is
incorrect (for an explanation, see <a class="reference internal" href="../../formats/data/fastq.html#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a>).
You can set the quality threshold with <code class="docutils literal notranslate"><span class="pre">--min-phred</span></code>.
We discourage using a quality threshold less than 25 because doing so could lead
to a background error rate that is too high for accurate mutational profiling
(e.g. 1% with <code class="docutils literal notranslate"><span class="pre">--min-phred</span> <span class="pre">20</span></code>), especially if you want to cluster your reads.</p>
</section>
<section id="how-to-trim-extra-dark-cycles-for-illumina-two-channel-chemistry">
<h4>How to trim extra dark cycles (for Illumina two-channel chemistry)<a class="headerlink" href="#how-to-trim-extra-dark-cycles-for-illumina-two-channel-chemistry" title="Link to this heading"></a></h4>
<p>Illumina sequencers using <a class="reference external" href="https://www.illumina.com/science/technology/next-generation-sequencing/sequencing-technology/2-channel-sbs.html">two-channel chemistry</a> (e.g. NextSeq, NovaSeq, iSeq)
interpret the lack of color from either channel as G.
Consequently, if a DNA molecule is shorter than the read length, then the final
cycles of sequencing will produce no light and be <a class="reference external" href="https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/">called as a string Gs</a>.
Using <code class="docutils literal notranslate"><span class="pre">--cut-nextseq</span></code> tells Cutadapt to <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/guide.html#nextseq-trim">trim high-quality Gs</a>
from the 3’ end of every read.
Trimming dark cycles can improve alignment in end-to-end mode, but it also trims
real G bases (which cannot be distinguished from artifactual ones) from the 3’
ends of reads.</p>
</section>
<section id="how-to-further-customize-read-trimming">
<h4>How to further customize read trimming<a class="headerlink" href="#how-to-further-customize-read-trimming" title="Link to this heading"></a></h4>
<p>Refer to <a class="reference internal" href="../../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can use
with Cutadapt, and the <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/reference.html">Cutadapt reference guide</a> for details on each.
These options suffice for most users.
If you need more customization, then you can trim your FASTQ files externally
and then perform alignment within SEISMIC-RNA, using <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code> to disable
additional trimming.</p>
</section>
</section>
<section id="align-setting-mapping-reads-with-bowtie-2">
<h3>Align setting: Mapping reads with Bowtie 2<a class="headerlink" href="#align-setting-mapping-reads-with-bowtie-2" title="Link to this heading"></a></h3>
<p>SEISMIC-RNA uses <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie 2</a> to align your reads to your reference sequences.</p>
<section id="how-to-pre-build-a-bowtie-2-index-optional">
<h4>How to pre-build a Bowtie 2 index (optional)<a class="headerlink" href="#how-to-pre-build-a-bowtie-2-index-optional" title="Link to this heading"></a></h4>
<p>Bowtie 2 requires the FASTA file of reference sequences to be indexed.
You can have SEISMIC-RNA build the index for you automatically (the default) or
index your FASTA file manually.
With automatic indexing, SEISMIC-RNA builds the index in a temporary directory
and deletes it after alignment finishes.
This option is ideal for small sets of references (i.e. up to several hundred
sequences of several thousand nucleotides each) because building the index will
take on the order of seconds to minutes.
However, for large sets of references (e.g. an entire mammalian transcriptome),
building the index can take on the order of hours.
In this case, we recommend building the index yourself using the command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bowtie2</span><span class="o">-</span><span class="n">build</span> <span class="p">{</span><span class="n">refs</span><span class="p">}</span><span class="o">.</span><span class="n">fa</span> <span class="p">{</span><span class="n">refs</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs}.fa</span></code> is the path of your FASTA file and <code class="docutils literal notranslate"><span class="pre">{refs}</span></code> is the path
without the FASTA file extension.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer">Bowtie 2 Indexer manual</a> for more information on building an index.
Note that, while Bowtie 2 does not require the index to have the same name as
the FASTA file, SEISMIC-RNA does, so make sure that you use the same path for
the FASTA file and the index, except that the index path should not have the
FASTA file extension.</p>
<p>Indexing will generate six files with the extensions <code class="docutils literal notranslate"><span class="pre">.1.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.2.bt2</span></code>,
<code class="docutils literal notranslate"><span class="pre">.3.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.4.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.rev.1.bt2</span></code>, and <code class="docutils literal notranslate"><span class="pre">.rev.2.bt2</span></code>.
As long as all six files are in the same directory as and have the same name
(minus the file extension) as the FASTA file, SEISMIC-RNA will use the index.
Otherwise, SEISMIC-RNA will build and use a temporary index.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you use a pre-built Bowtie 2 index, then SEISMIC-RNA does <em>not</em> verify
that the index was actually built from the FASTA file of the same name.
Discrepancies between the FASTA file and the index files can crash the Align
and Relate steps or produce erroneous results.</p>
</div>
</section>
<section id="how-to-choose-between-local-and-end-to-end-alignment">
<h4>How to choose between local and end-to-end alignment<a class="headerlink" href="#how-to-choose-between-local-and-end-to-end-alignment" title="Link to this heading"></a></h4>
<p>You can align either whole reads (end-to-end mode) or the part of each read that
aligns best to the reference (local mode).
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#end-to-end-alignment-versus-local-alignment">description of alignment modes in Bowtie 2</a> for more details.</p>
<p>Generally, end-to-end mode yields spurious mutations (false positives) at the
ends of reads if the reads contain artifacts such as low-quality base calls or
untrimmed or improperly trimmed adapters.
Conversely, local mode misses real mutations (false negatives) within several
nucleotides of the ends of reads because such mutations cannot be part of the
best local alignment, which penalizes mutations and rewards matches.</p>
<p>For RNA mutational profiling, false positives generally cause more problems than
do false negatives, so SEISMIC-RNA uses local mode (<code class="docutils literal notranslate"><span class="pre">--bt2-local</span></code>) by default.
Use end-to-end mode (<code class="docutils literal notranslate"><span class="pre">--bt2-end-to-end</span></code>) only if you have a compelling reason
to do so (e.g. if you must quantify mutations at the ends of reads) and only
after carefully trimming any extraneous sequences from the ends of the reads.</p>
</section>
<section id="how-to-align-paired-end-reads">
<h4>How to align paired-end reads<a class="headerlink" href="#how-to-align-paired-end-reads" title="Link to this heading"></a></h4>
<p>If your reads are paired-end, then you have additional options for keeping or
discarding read pairs depending on how the two reads in the pair (called mates)
align relative to each other.
Bowtie 2 considers mates to align “concordantly” when their relative positions
match expectations and “discordantly” otherwise.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#concordant-pairs-match-pair-expectations-discordant-pairs-dont">Bowtie 2 manual for details on concordant/discordant alignments</a>.
By default, SEISMIC-RNA keeps only concordantly aligned pairs.
To include discordantly aligned pairs too, add <code class="docutils literal notranslate"><span class="pre">--bt2-discordant</span></code>.</p>
<p>Several options control which types of alignments are considered concordant
versus discordant.</p>
<p>You can specify where mates should align relative to each other: mates may
<a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mates-can-overlap-contain-or-dovetail-each-other">overlap partially or completely, or dovetail</a>.
By default, overlaps (partial and complete) are considered concordant, while
dovetailing is considered discordant.
You can treat complete overlaps as discordant with <code class="docutils literal notranslate"><span class="pre">--bt2-no-contain</span></code>, or
dovetailed mates as concordant with <code class="docutils literal notranslate"><span class="pre">--bt2-dovetail</span></code>.
Pairs that overlap partially (without dovetailing) are always concordant in
SEISMIC-RNA.</p>
<p>You can also specify the orientation of paired mates using <code class="docutils literal notranslate"><span class="pre">--bt2-orient</span></code>.
The choices are <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the 5’-most mate is forward, the 3’-most is reversed),
<code class="docutils literal notranslate"><span class="pre">rf</span></code> (the 5’-most mate is reversed, the 3’-most is forward), or <code class="docutils literal notranslate"><span class="pre">ff</span></code> (both
mates are forward).
The default is <code class="docutils literal notranslate"><span class="pre">fr</span></code> (and if you are not sure which orientation you need, then
you probably need the default).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>First, <code class="docutils literal notranslate"><span class="pre">--bt2[-no]-contain</span></code>, <code class="docutils literal notranslate"><span class="pre">--bt2[-no]-dovetail</span></code>, and <code class="docutils literal notranslate"><span class="pre">--bt2-orient</span></code>
choose which paired-end alignments count as concordant or discordant.
If discordant, then <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]discordant</span></code> choose whether to keep them.
Using <code class="docutils literal notranslate"><span class="pre">--bt2-no-contain</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-discordant</span></code>, for example, would make
alignments where one mate fully contains the other discordant (because of
<code class="docutils literal notranslate"><span class="pre">--bt2-no-contain</span></code>) but still kept (because of <code class="docutils literal notranslate"><span class="pre">--bt2-discordant</span></code>),
despite what the name “no-contain” would imply.</p>
</div>
<p>You can also enable <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a> with <code class="docutils literal notranslate"><span class="pre">--bt2-mixed</span></code>.
In mixed mode, if two mates fail to align as a pair, then Bowtie 2 will attempt
to align each mate individually, like a single-end read.
(It is possible in mixed mode for only one mate in a pair to align.)</p>
</section>
<section id="how-to-filter-aligned-reads">
<h4>How to filter aligned reads<a class="headerlink" href="#how-to-filter-aligned-reads" title="Link to this heading"></a></h4>
<p>You can filter alignments by <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">alignment score</a> and <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">mapping quality</a>.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">Alignment score</a> measures how <em>well</em> a read aligns to <em>one specific location</em>
in <em>one reference sequence</em> and depends on the number of matches, substitutions,
and gaps, using the score parameters.
You can specify the minimum alignment score for local and end-to-end modes using
<code class="docutils literal notranslate"><span class="pre">--bt2-score-min-loc</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-e2e</span></code>, respectively.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#valid-alignments-meet-or-exceed-the-minimum-score-threshold">section of the Bowtie 2 manual on alignment scores</a> for advice.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">Mapping quality</a> measures how <em>unique</em> an alignment is among <em>all locations</em>
in <em>all reference sequences</em>: high if the read aligns with a high alignment
score to exactly one location, low quality if it aligns with similar alignment
scores to multiple locations in the reference (and thus it is hard to determine
a single location where the read aligns).
The default minimum mapping quality is 25, meaning that the probability that the
chosen location is incorrect is 10<sup>-2.5</sup> = 0.3%.
You can change the minimum mapping quality using <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>.</p>
</section>
<section id="how-to-filter-by-number-of-aligned-reads">
<h4>How to filter by number of aligned reads<a class="headerlink" href="#how-to-filter-by-number-of-aligned-reads" title="Link to this heading"></a></h4>
<p>Alignment maps containing very few reads are not generally useful for mutational
profiling, due to low coverage per position.
When aligning to many references (e.g. an entire transcriptome), most references
will receive few reads, producing many output files that would be unusable for
further processing.
To prevent unusable files from cluttering your output directory, you can choose
to have alignment map files with insufficient reads deleted automatically.
The default minimum is 1000 reads, which you can change using <code class="docutils literal notranslate"><span class="pre">--min-reads</span></code>.
With no minimum (<code class="docutils literal notranslate"><span class="pre">--min-reads</span> <span class="pre">0</span></code>), no files are deleted automatically.</p>
</section>
<section id="how-to-further-customize-alignment">
<h4>How to further customize alignment<a class="headerlink" href="#how-to-further-customize-alignment" title="Link to this heading"></a></h4>
<p>See <a class="reference internal" href="../../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can use with
Bowtie 2, and the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">Bowtie 2 manual</a> for details on each of these options.
These options suffice for most users.
If you need more customization, then you can align your FASTQ files externally
and pass the alignment maps into SEISMIC-RNA during <a class="reference internal" href="relate.html"><span class="doc">Relate: Compute relationships between references and aligned reads</span></a>.</p>
</section>
</section>
<section id="align-setting-format-of-alignment-maps">
<span id="bam-vs-cram"></span><h3>Align setting: Format of alignment maps<a class="headerlink" href="#align-setting-format-of-alignment-maps" title="Link to this heading"></a></h3>
<p>You can choose to output alignment map files in either BAM or CRAM format.
For information on these file formats, see <a class="reference internal" href="../../formats/data/xam.html"><span class="doc">SAM, BAM, and CRAM: Alignment Maps</span></a>.
The default is CRAM (<code class="docutils literal notranslate"><span class="pre">--cram</span></code>); you can switch to BAM using <code class="docutils literal notranslate"><span class="pre">--bam</span></code>.</p>
<p>Alignment maps in CRAM format are smaller than their BAM counterparts, and hence
better suited to long-term storage.
However, the better compression of CRAM files comes at three costs:</p>
<ul class="simple">
<li><p>A CRAM file must be accompanied by a FASTA file storing the sequence of every
reference that appears in the header of the CRAM file.
A CRAM file stores only the relative path to its FASTA file, not the sequence
information, which enables the CRAM file to be much smaller than it would be
if it did need to contain its own sequences.
Because the FASTA file existed before and during the alignment, having this
FASTA file accompany the CRAM file usually incurs no extra cost.
However, moving or deleting the FASTA will break the CRAM file.
As a safeguard against this fragility, SEISMIC-RNA keeps a copy of the FASTA
file in the same directory as the output CRAM file.
Creating an actual copy would require more storage space and defeat the point
of CRAM’s smaller file size, so SEISMIC-RNA actually makes a <a class="reference external" href="https://en.wikipedia.org/wiki/Hard_link">hard link</a> –
not a copy – which requires minimal extra space.
In some circumstances, making a hard link can fail, in which case SEISMIC-RNA
will resort to copying the FASTA file instead.</p></li>
<li><p>Reading and writing CRAM files is slower than for BAM files due to the extra
effort needed for compressing and decompressing CRAM files.</p></li>
<li><p>In the <a class="reference external" href="https://samtools.github.io/hts-specs/">CIGAR strings</a>, distinction between reference matches (<code class="docutils literal notranslate"><span class="pre">=</span></code>) and
substitutions (<code class="docutils literal notranslate"><span class="pre">X</span></code>) is lost upon compressing to CRAM format.
Thus, the Relate step must perform extra work to determine if each non-indel
position is a match or substitution, which makes it run more slowly than it
would if the distinction had been preserved.</p></li>
</ul>
<p>In general, use CRAM format if minimizing the size of your alignment map files
is a priority, especially for long-term storage.
Use BAM format to make the <code class="docutils literal notranslate"><span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">relate</span></code> steps run faster, and to
make the output files more portable (since BAM files are self-contained, while
CRAM files will break without the FASTA file that accompanies them).</p>
</section>
</section>
<section id="align-output-files">
<h2>Align: Output files<a class="headerlink" href="#align-output-files" title="Link to this heading"></a></h2>
<p>All output files except FastQC reports are written to <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/align</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is your output directory and <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the sample name.</p>
<section id="align-output-file-fastqc-reports">
<h3>Align output file: FastQC reports<a class="headerlink" href="#align-output-file-fastqc-reports" title="Link to this heading"></a></h3>
<p>If you run FastQC, then its report files go into <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc</span></code>.
The directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc/initial</span></code> contains the FastQC reports for your
initial FASTQ files, before trimming.
If you also run trimming, then reports for the post-trimmed FASTQ files go into
<code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc/trimmed</span></code>.</p>
<p>In each directory, FastQC outputs <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.html</span></code> (the FastQC report)
and <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.zip</span></code> (extra information), where <code class="docutils literal notranslate"><span class="pre">{fq_name}</span></code> comes from
the original FASTQ file.
If you add <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code>, then each <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.zip</span></code> will be unzipped
to the directory <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc</span></code>.
For details on these outputs, see the documentation for <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>.</p>
</section>
<section id="align-output-file-alignment-maps">
<h3>Align output file: Alignment maps<a class="headerlink" href="#align-output-file-alignment-maps" title="Link to this heading"></a></h3>
<p>Alignment maps store the location in the reference sequence to which each read
aligned, plus the Phred quality scores, mapping quality, and mutated positions.
(For more information on alignment maps, see <a class="reference internal" href="../../formats/data/xam.html"><span class="doc">SAM, BAM, and CRAM: Alignment Maps</span></a>.)
SEISMIC-RNA outputs alignment maps where every read aligns to the same reference
(although this is not a restriction outside of SEISMIC-RNA).
Each alignment map is written to <code class="docutils literal notranslate"><span class="pre">{ref}.{xam}</span></code>, where <code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is the name of
the reference to which the reads aligned, and <code class="docutils literal notranslate"><span class="pre">{xam}</span></code> is the file extension
(depending on the selected format).
SEISMIC-RNA can output alignment maps in either BAM or CRAM format.
For a comparison of these formats, see <a class="reference internal" href="#bam-vs-cram"><span class="std std-ref">Align setting: Format of alignment maps</span></a>.</p>
</section>
<section id="align-output-file-reference-sequences">
<h3>Align output file: Reference sequences<a class="headerlink" href="#align-output-file-reference-sequences" title="Link to this heading"></a></h3>
<p>If you choose to output alignment maps in CRAM format, then you also get a FASTA
file(s) of the reference sequence(s) alongside the CRAM files.
If the reads came from a whole sample, then a single FASTA file with the same
name as the input FASTA file will be output.
The output file will be a <a class="reference external" href="https://en.wikipedia.org/wiki/Hard_link">hard link</a> to the input file, if possible, to avoid
consuming unnecessary storage space.
If the reads were demultiplexed before alignment, then for each CRAM file, a
FASTA file with the same name (up to the file extension) will be output.
In both cases, each FASTA file will be indexed using <a class="reference external" href="https://www.htslib.org/doc/samtools-faidx.html">samtools faidx</a> to speed
up reading the CRAM files.
If you choose to output alignment maps in BAM format, then you get (and need)
no FASTA files alongside them.</p>
</section>
<section id="align-output-file-unaligned-reads">
<span id="align-unaligned"></span><h3>Align output file: Unaligned reads<a class="headerlink" href="#align-output-file-unaligned-reads" title="Link to this heading"></a></h3>
<p>In addition to the alignment maps, SEISMIC-RNA outputs FASTQ file(s) of reads
that Bowtie 2 could not align:</p>
<ul class="simple">
<li><p>Each whole-sample FASTQ file of single-end (<code class="docutils literal notranslate"><span class="pre">-z</span></code>) or interleaved (<code class="docutils literal notranslate"><span class="pre">-y</span></code>)
reads yields one file: <code class="docutils literal notranslate"><span class="pre">unaligned.fq.gz</span></code></p></li>
<li><p>Each pair of whole-sample FASTQ files of 1st and 2nd mates (<code class="docutils literal notranslate"><span class="pre">-x</span></code>) yields two
files: <code class="docutils literal notranslate"><span class="pre">unaligned.fq.1.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">unaligned.fq.2.gz</span></code></p></li>
<li><p>Each demultiplexed FASTQ file of single-end (<code class="docutils literal notranslate"><span class="pre">-Z</span></code>) or interleaved (<code class="docutils literal notranslate"><span class="pre">-Y</span></code>)
reads yields one file: <code class="docutils literal notranslate"><span class="pre">{ref}__unaligned.fq.gz</span></code></p></li>
<li><p>Each pair of demultiplexed FASTQ files of 1st and 2nd mates (<code class="docutils literal notranslate"><span class="pre">-X</span></code>) yields
two files: <code class="docutils literal notranslate"><span class="pre">{ref}__unaligned.fq.1.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">{ref}__unaligned.fq.2.gz</span></code></p></li>
</ul>
<p>where <code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is the reference for demultiplexed FASTQ files.</p>
<p>You can disable outputting unaligned using <code class="docutils literal notranslate"><span class="pre">--bt2-no-un</span></code>.</p>
</section>
<section id="align-output-file-align-report">
<h3>Align output file: Align report<a class="headerlink" href="#align-output-file-align-report" title="Link to this heading"></a></h3>
<p>SEISMIC-RNA also writes a report file, <code class="docutils literal notranslate"><span class="pre">align-report.json</span></code>, that records the
settings you used for running the Align step and summarizes the results.
See <a class="reference internal" href="../../formats/report/align.html"><span class="doc">Align Report</span></a> for more information.</p>
<section id="check-the-number-of-reads-that-aligned-overall">
<h4>Check the number of reads that aligned overall<a class="headerlink" href="#check-the-number-of-reads-that-aligned-overall" title="Link to this heading"></a></h4>
<p>Under “Number of reads after alignment”, the report says how many single-end
and/or paired-end reads were in the FASTQ file(s), and how many reads aligned.
This information is copied verbatim from the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#alignment-summary">alignment summary</a> of Bowtie 2;
see its documentation for more details.
For paired-end reads, each pair counts as one read.</p>
</section>
<section id="check-the-number-of-reads-that-aligned-to-each-reference">
<h4>Check the number of reads that aligned to each reference<a class="headerlink" href="#check-the-number-of-reads-that-aligned-to-each-reference" title="Link to this heading"></a></h4>
<p>Under “Number of reads aligned by reference”, the report lists every reference
in your input FASTA file and the number of reads that aligned to it.
For paired-end reads, each pair counts as one read.</p>
</section>
</section>
</section>
<section id="align-troubleshoot-and-optimize">
<h2>Align: Troubleshoot and optimize<a class="headerlink" href="#align-troubleshoot-and-optimize" title="Link to this heading"></a></h2>
<section id="align-produces-alignment-map-files-too-slowly">
<h3>Align produces alignment map files too slowly<a class="headerlink" href="#align-produces-alignment-map-files-too-slowly" title="Link to this heading"></a></h3>
<p>First, try running the Align step using more processors (with <code class="docutils literal notranslate"><span class="pre">--max-procs</span></code>),
at the cost of using more memory.
If, as a result, <a class="reference internal" href="#align-crash-hang"><span class="std std-ref">Align crashes or hangs without producing alignment map files</span></a>, then try adjusting the settings of
Bowtie 2 to increase the speed, at the risk of overlooking valid alignments.
See <a class="reference internal" href="../../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the Bowtie 2 settings you can adjust in SEISMIC-RNA,
and the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">Bowtie 2 manual</a> for more detailed descriptions.</p>
</section>
<section id="align-crashes-or-hangs-without-producing-alignment-map-files">
<span id="align-crash-hang"></span><h3>Align crashes or hangs without producing alignment map files<a class="headerlink" href="#align-crashes-or-hangs-without-producing-alignment-map-files" title="Link to this heading"></a></h3>
<p>Most likely, your system has run out of memory.
You can confirm using a program that monitors memory usage (such as <code class="docutils literal notranslate"><span class="pre">top</span></code> in a
Linux/macOS terminal, Activity Monitor on macOS, or Task Manager on Windows).
If so, then rerun the Align step using fewer processors (with <code class="docutils literal notranslate"><span class="pre">--max-procs</span></code>)
to limit the memory usage, at the cost of slower alignment.</p>
</section>
<section id="fewer-reads-aligned-than-you-expected">
<h3>Fewer reads aligned than you expected<a class="headerlink" href="#fewer-reads-aligned-than-you-expected" title="Link to this heading"></a></h3>
<p>Try the following steps (in this order):</p>
<ol class="arabic simple">
<li><p>Ensure you are using Bowtie version 2.5.1 or later (version 2.5.0 has a bug
that affects alignment rate).
You can check the version with <code class="docutils literal notranslate"><span class="pre">bowtie2</span> <span class="pre">--version</span> <span class="pre">|</span> <span class="pre">head</span> <span class="pre">-n</span> <span class="pre">1</span></code>.</p></li>
<li><p>Double check that your FASTA file has the correct reference sequence(s) and
that, if you pre-built the Bowtie 2 index before running <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code>,
that you indexed the correct FASTA file.</p></li>
<li><p>Examine the reads that failed to align (see <a class="reference internal" href="#align-unaligned"><span class="std std-ref">Align output file: Unaligned reads</span></a>).
Choose several reads randomly, copy one or two 20 - 40 nt segments from the
middle of each read, and check if the segments come from any known sources
by querying <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&amp;PAGE_TYPE=BlastSearch&amp;LINK_LOC=blasthome">BLAST</a> (or similar tools).
Identifying the sources of unaligned reads can help determine the cause of
the problem (e.g. contamination with ribosomal RNA or foreign nucleic acids
such as from <em>Mycoplasma</em>) and whether the reads that did align are usable.</p></li>
</ol>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="demult.html" class="btn btn-neutral float-left" title="Demultiplex: Split multiplexed FASTQ files by their barcodes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="relate.html" class="btn btn-neutral float-right" title="Relate: Compute relationships between references and aligned reads" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>