<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Fold: Predict RNA secondary structures using mutation rates &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Graph: Plot data from tables and/or structures and compare samples" href="graph.html" />
    <link rel="prev" title="Table: Count mutations for each read and position" href="table.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../whatdoes/index.html">What it Does</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="align.html">Align: Trim FASTQ files and align them to reference sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate: Compute relationships between references and aligned reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="pool.html">Pool: Combine samples from the Relate step</a></li>
<li class="toctree-l3"><a class="reference internal" href="mask.html">Mask: Define mutations and sections to filter reads and positions</a></li>
<li class="toctree-l3"><a class="reference internal" href="cluster.html">Cluster: Infer alternative structures by clustering reads’ mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="table.html">Table: Count mutations for each read and position</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Fold: Predict RNA secondary structures using mutation rates</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fold-input-files">Fold: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fold-settings">Fold: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fold-output-files">Fold: Output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fold-visualize-structures-in-varna">Fold: Visualize structures in VARNA</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="graph.html">Graph: Plot data from tables and/or structures and compare samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3"><a class="reference internal" href="wf.html">Workflow: Run all steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Fold: Predict RNA secondary structures using mutation rates</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/fold.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="fold-predict-rna-secondary-structures-using-mutation-rates">
<h1>Fold: Predict RNA secondary structures using mutation rates<a class="headerlink" href="#fold-predict-rna-secondary-structures-using-mutation-rates" title="Permalink to this headline"></a></h1>
<section id="fold-input-files">
<h2>Fold: Input files<a class="headerlink" href="#fold-input-files" title="Permalink to this headline"></a></h2>
<section id="fold-input-file-mask-or-cluster-positional-table">
<h3>Fold input file: Mask or Cluster positional table<a class="headerlink" href="#fold-input-file-mask-or-cluster-positional-table" title="Permalink to this headline"></a></h3>
<p>You can give any number of positional table files of masked or clustered reads
(<code class="docutils literal notranslate"><span class="pre">mask-per-pos.csv</span></code> or <code class="docutils literal notranslate"><span class="pre">clust-per-pos.csv</span></code>, respectively) as inputs.
See <a class="reference internal" href="../inputs.html"><span class="doc">List Input Files</span></a> for ways to list multiple files.
(SEISMIC-RNA will not crash if you give other type of table files, such as a
<code class="docutils literal notranslate"><span class="pre">relate-per-pos.csv</span></code> or <code class="docutils literal notranslate"><span class="pre">mask-per-read.csv.gz</span></code> file, but will ignore them.)</p>
<p>To predict structures using the mutational profiles in all valid tables in the
directory <code class="docutils literal notranslate"><span class="pre">{out}</span></code>, you could use the command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">fold</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span>
</pre></div>
</div>
</section>
</section>
<section id="fold-settings">
<h2>Fold: Settings<a class="headerlink" href="#fold-settings" title="Permalink to this headline"></a></h2>
<section id="fold-setting-define-sections">
<h3>Fold setting: Define sections<a class="headerlink" href="#fold-setting-define-sections" title="Permalink to this headline"></a></h3>
<p>You can predict structures of the full reference sequences or specific sections.
See <a class="reference internal" href="../sections.html"><span class="doc">Define Sections</span></a> for ways to define sections.</p>
<p>Defining sections in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">fold</span></code> works identically to <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">mask</span></code> but
accomplishes a very different purpose.
Sections in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">fold</span></code> determine for which parts of the reference sequence
to predict structures.
Sections in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">mask</span></code> determine for which parts of the reference sequence
to use mutational data.
SEISMIC-RNA allows these sections to be different.
There are several common scenarios:</p>
<ul class="simple">
<li><p>The section you are folding matches the section for which you have data.
For example, you could have mutationally profiled a full transcript and now
want to predict the structure of the full transcript using the data from the
full mutational profile.</p></li>
<li><p>You are folding a section that contains and is longer than the section for
which you have data.
For example, you could have mutationally profiled a short amplicon from a much
longer transcript; and after clustering that amplicon, you want to model each
alternative structure of the long transcript while using the short mutational
profile of each cluster to guide the structure predictions.</p></li>
<li><p>You are folding a short section that is contained by a longer section for
which you have mutational profiling data.
For example, you could have mutationally profiled a full transcript and now
want to predict the structure of a small part of the transcript that you are
reasonably sure does not interact with any other part of the transcript.</p></li>
</ul>
</section>
<section id="fold-setting-quantile-for-normalization">
<h3>Fold setting: Quantile for normalization<a class="headerlink" href="#fold-setting-quantile-for-normalization" title="Permalink to this headline"></a></h3>
<p>Folding requires that the mutation rates be normalized to the interval [0, 1].
See <a class="reference internal" href="../normalize.html"><span class="doc">Normalize Mutation Rates</span></a> for ways to normalize mutation rates.</p>
</section>
<section id="fold-setting-rnastructure-parameters">
<h3>Fold setting: RNAstructure parameters<a class="headerlink" href="#fold-setting-rnastructure-parameters" title="Permalink to this headline"></a></h3>
<p><code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">fold</span></code> exposes several options for the RNAstructure Fold program (see
the <a class="reference external" href="https://rna.urmc.rochester.edu/Text/Fold.html">documentation for Fold</a> for details on each option):</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 18%" />
<col style="width: 18%" />
<col style="width: 64%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">fold</span></code></p></th>
<th class="head"><p>Option in RNAstructure Fold</p></th>
<th class="head"><p>Brief explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--fold-temp</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--temperature</span></code></p></td>
<td><p>temperature (K) of folding</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--fold-constraint</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--constraint</span></code></p></td>
<td><p>optional <a class="reference external" href="https://rna.urmc.rochester.edu/Text/File_Formats.html#Constraint">folding constraints file</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--fold-md</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--maxdistance</span></code></p></td>
<td><p>maximum distance between paired bases</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--fold-mfe</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--MFE</span></code></p></td>
<td><p>predict only the optimal structure (same result as <code class="docutils literal notranslate"><span class="pre">--fold-max</span> <span class="pre">1</span></code>, but about twice as fast)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--fold-max</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--maximum</span></code></p></td>
<td><p>maximum number of structures to predict (ignored if using <code class="docutils literal notranslate"><span class="pre">--fold-mfe</span></code>)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--fold-percent</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--percent</span></code></p></td>
<td><p>maximum % difference in free energy of predicted structures (ignored if using <code class="docutils literal notranslate"><span class="pre">--fold-mfe</span></code>)</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="fold-output-files">
<h2>Fold: Output files<a class="headerlink" href="#fold-output-files" title="Permalink to this headline"></a></h2>
<p>All output files go into the directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/fold/{ref}/{sect}</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory, <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the sample, <code class="docutils literal notranslate"><span class="pre">{ref}</span></code>
is the reference, and <code class="docutils literal notranslate"><span class="pre">{sect}</span></code> is the section you <em>folded</em> (<strong>not</strong> that from
which the data came).
The files for each predicted structure are named <code class="docutils literal notranslate"><span class="pre">{sect}__{profile}</span></code>, where
<code class="docutils literal notranslate"><span class="pre">{sect}</span></code> is the section from which the <em>data</em> came (<strong>not</strong> that which you
folded) and <code class="docutils literal notranslate"><span class="pre">{profile}</span></code> is the mutational profile of those data, which can be
<code class="docutils literal notranslate"><span class="pre">average</span></code> (ensemble average) or <code class="docutils literal notranslate"><span class="pre">cluster-{n}-{i}</span></code> (where <code class="docutils literal notranslate"><span class="pre">{n}</span></code> is the
number of clusters and <code class="docutils literal notranslate"><span class="pre">{i}</span></code> is the cluster number).</p>
<section id="fold-output-file-fold-report">
<h3>Fold output file: Fold report<a class="headerlink" href="#fold-output-file-fold-report" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA writes a report file, <code class="docutils literal notranslate"><span class="pre">fold-report.json</span></code>, to record the settings
you used for running the Fold step.
See <a class="reference internal" href="../../formats/report/fold.html"><span class="doc">Fold Report</span></a> for more information.</p>
</section>
<section id="fold-output-file-connectivity-table">
<h3>Fold output file: Connectivity table<a class="headerlink" href="#fold-output-file-connectivity-table" title="Permalink to this headline"></a></h3>
<p>The primary output is a connectivity table file.
For details on this format, see <a class="reference internal" href="../../formats/data/ct.html"><span class="doc">Connectivity Table (CT): RNA secondary structures</span></a>.</p>
</section>
<section id="fold-output-file-dot-bracket-structure">
<span id="fold-db"></span><h3>Fold output file: Dot-bracket structure<a class="headerlink" href="#fold-output-file-dot-bracket-structure" title="Permalink to this headline"></a></h3>
<p>The Fold step also outputs the structures in dot-bracket format, which you can
copy-paste into RNA drawing software such as <a class="reference external" href="https://varna.lisn.upsaclay.fr/">VARNA</a>.
For details on this format, see <a class="reference internal" href="../../formats/data/db.html"><span class="doc">Dot-bracket (DB): RNA secondary structures</span></a>.</p>
</section>
<section id="fold-output-file-varna-color-file">
<h3>Fold output file: VARNA color file<a class="headerlink" href="#fold-output-file-varna-color-file" title="Permalink to this headline"></a></h3>
<p>The Fold step also outputs the normalized mutation rates in VARNA color format,
which you can import into the RNA drawing software <a class="reference external" href="https://varna.lisn.upsaclay.fr/">VARNA</a>.
For details on this format, see <a class="reference internal" href="../../formats/data/varna-color.html"><span class="doc">VARNA Color: Color codes for VARNA</span></a>.</p>
</section>
</section>
<section id="fold-visualize-structures-in-varna">
<h2>Fold: Visualize structures in VARNA<a class="headerlink" href="#fold-visualize-structures-in-varna" title="Permalink to this headline"></a></h2>
<p><a class="reference external" href="https://varna.lisn.upsaclay.fr/">VARNA</a> is a third-party application for drawing RNA structures.
To draw a structure from SEISMIC-RNA in VARNA:</p>
<ol class="arabic">
<li><p>Install (if needed) and launch VARNA.</p></li>
<li><p>Open your dot-bracket file (see <a class="reference internal" href="#fold-db"><span class="std std-ref">Fold output file: Dot-bracket structure</span></a>) in a text editor.</p></li>
<li><p>Right-click the drawing canvas, select “File” &gt; “New…”, and copy-paste the
sequence and dot-bracket structure.</p></li>
<li><p>Adjust the layout of the structure by clicking and dragging.</p></li>
<li><p>To color the bases by their mutation rates, right-click the drawing canvas,
select “Display” &gt; “Color map” &gt; “Load values…”, copy-paste the path to
your VARNA color file into the box or click “Choose file” and navigate to
your VARNA color file, and click “OK” to load the file.</p></li>
<li><p>To customize the colors, select “Display” &gt; “Color map” &gt; “Style…”:</p>
<ul class="simple">
<li><p>Drag a color bar to adjust its location.</p></li>
<li><p>Click the square below a color bar to change its color.</p></li>
<li><p>Click the X below the square to delete the color.</p></li>
<li><p>Click anywhere on the color spectrum to create a new color bar.</p></li>
</ul>
<p>We recommend setting the color for missing data (-1) to white or light gray
and using a continuous (not discrete) color scale for the mutation data.</p>
</li>
</ol>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="table.html" class="btn btn-neutral float-left" title="Table: Count mutations for each read and position" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="graph.html" class="btn btn-neutral float-right" title="Graph: Plot data from tables and/or structures and compare samples" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>