<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Mask: Define mutations and sections to filter reads and positions &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Cluster: Infer alternative structures by clustering reads’ mutations" href="cluster.html" />
    <link rel="prev" title="Pool: Merge samples (vertically) from the Relate step" href="pool.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../whatdoes/index.html">What it Does</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="align.html">Align: Trim FASTQ files and align them to reference sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate: Compute relationships between references and aligned reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="pool.html">Pool: Merge samples (vertically) from the Relate step</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Mask: Define mutations and sections to filter reads and positions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#mask-input-files">Mask: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mask-settings">Mask: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mask-output-files">Mask: Output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mask-troubleshoot-and-optimize">Mask: Troubleshoot and optimize</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="cluster.html">Cluster: Infer alternative structures by clustering reads’ mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="join.html">Join: Merge sections (horizontally) from the Mask or Cluster step</a></li>
<li class="toctree-l3"><a class="reference internal" href="table.html">Table: Count mutations for each read and position</a></li>
<li class="toctree-l3"><a class="reference internal" href="fold.html">Fold: Predict RNA secondary structures using mutation rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="graph.html">Graph: Plot data from tables and/or structures and compare samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3"><a class="reference internal" href="wf.html">Workflow: Run all steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adjclust.html">Add/Delete Orders to/from an Already-Clustered Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Mask: Define mutations and sections to filter reads and positions</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/mask.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="mask-define-mutations-and-sections-to-filter-reads-and-positions">
<h1>Mask: Define mutations and sections to filter reads and positions<a class="headerlink" href="#mask-define-mutations-and-sections-to-filter-reads-and-positions" title="Permalink to this headline"></a></h1>
<section id="mask-input-files">
<h2>Mask: Input files<a class="headerlink" href="#mask-input-files" title="Permalink to this headline"></a></h2>
<section id="mask-input-file-relate-report">
<h3>Mask input file: Relate report<a class="headerlink" href="#mask-input-file-relate-report" title="Permalink to this headline"></a></h3>
<p>You can give any number of Relate report files as inputs for the Mask step.
See <a class="reference internal" href="../inputs.html"><span class="doc">List Input Files</span></a> for ways to list multiple files.</p>
<p>For example, to mask relation vectors of reads from <code class="docutils literal notranslate"><span class="pre">sample-1</span></code> related to
references <code class="docutils literal notranslate"><span class="pre">ref-1</span></code> and <code class="docutils literal notranslate"><span class="pre">ref-2</span></code>, and from <code class="docutils literal notranslate"><span class="pre">sample-2</span></code> related to reference
<code class="docutils literal notranslate"><span class="pre">ref-1</span></code>, use the command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">mask</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/</span><span class="n">sample</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">relate</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mi">1</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/</span><span class="n">sample</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">relate</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mi">2</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/</span><span class="n">sample</span><span class="o">-</span><span class="mi">2</span><span class="o">/</span><span class="n">relate</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mi">1</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the path of your output directory from the Relate step.</p>
<p>To mask all relation vectors in <code class="docutils literal notranslate"><span class="pre">{out}</span></code>, you can use the command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">mask</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span>
</pre></div>
</div>
</section>
</section>
<section id="mask-settings">
<h2>Mask: Settings<a class="headerlink" href="#mask-settings" title="Permalink to this headline"></a></h2>
<section id="mask-setting-define-sections">
<h3>Mask setting: Define sections<a class="headerlink" href="#mask-setting-define-sections" title="Permalink to this headline"></a></h3>
<p>You can mask the full reference sequences or select specific sections.
The latter is useful for investigating small elements of longer sequences, such
as a 350 nt <a class="reference external" href="https://en.wikipedia.org/wiki/Internal_ribosome_entry_site">IRES</a> within a 9,600 nt viral genome.
See <a class="reference internal" href="../sections.html"><span class="doc">Define Sections</span></a> for ways to define sections.</p>
</section>
<section id="mask-setting-define-mutations">
<h3>Mask setting: Define mutations<a class="headerlink" href="#mask-setting-define-mutations" title="Permalink to this headline"></a></h3>
<p>The Mask step takes in relation vectors – which encode relationships including
ambiguous mutations – and outputs bit vectors, wherein each position in each
read has a binary, mutated/matched status.
For more information on relation vectors, see <a class="reference internal" href="../../data/relate/codes.html"><span class="doc">Relation Vectors</span></a>.</p>
<p>Producing bit vectors requires deciding which types of relationships count as
mutations, which count as matches, and which count as neither.
You can choose which types of relationships to count as matches and mutations.
The default is to count all 4 types of matches (A→A, C→C, G→G, T→T) as matches
and all 12 types of substitutions (A→C, A→G, A→T, C→A, C→G, C→T, G→A, G→C, G→T,
T→A, T→C, T→G) as mutations, but not to count deletions and insertions (indels).
To count deletions and insertions as mutations, add <code class="docutils literal notranslate"><span class="pre">--count-del</span></code> and
<code class="docutils literal notranslate"><span class="pre">--count-ins</span></code>, respectively.</p>
<p>You can also choose to not count individual types of relationships, such as
substitutions from A to G (but still count every other type of substitution).
To ignore one type of relationship, add <code class="docutils literal notranslate"><span class="pre">--discount-mut</span></code> followed by a code
of two lowercase letters:</p>
<ul class="simple">
<li><p>The first letter is the base in the reference (<code class="docutils literal notranslate"><span class="pre">a</span></code>/<code class="docutils literal notranslate"><span class="pre">c</span></code>/<code class="docutils literal notranslate"><span class="pre">g</span></code>/<code class="docutils literal notranslate"><span class="pre">t</span></code>)</p></li>
<li><p>The second letter is the base in the read (for substitutions) or <code class="docutils literal notranslate"><span class="pre">d</span></code>/<code class="docutils literal notranslate"><span class="pre">i</span></code>
(for deletions and insertions, respectively).</p></li>
</ul>
<p>For example, to count all substitutions except A→G and all deletions except
of C, use <code class="docutils literal notranslate"><span class="pre">--count-del</span> <span class="pre">--discount-mut</span> <span class="pre">ag</span> <span class="pre">--discount-mut</span> <span class="pre">cd</span></code>.</p>
</section>
<section id="mask-setting-exclude-positions">
<span id="mask-exclude"></span><h3>Mask setting: Exclude positions<a class="headerlink" href="#mask-setting-exclude-positions" title="Permalink to this headline"></a></h3>
<p>The first substep of masking is excluding pre-specified positions.
You can specify three types of positions to exclude.</p>
<section id="exclude-positions-with-g-and-u-bases">
<h4>Exclude positions with G and U bases<a class="headerlink" href="#exclude-positions-with-g-and-u-bases" title="Permalink to this headline"></a></h4>
<p>DMS methylates G and U much less than A and C residues under physiological
conditions [<a class="reference external" href="https://doi.org/10.1038/nmeth.4057">Zubradt et al. (2017)</a>], so positions with G or U bases are
generally excluded when DMS is the chemical probe.
Use <code class="docutils literal notranslate"><span class="pre">--exclude-gu</span></code> (default) and <code class="docutils literal notranslate"><span class="pre">--include-gu</span></code> to choose whether to use
G and U bases.</p>
</section>
<section id="exclude-positions-with-poly-a-sequences">
<h4>Exclude positions with poly(A) sequences<a class="headerlink" href="#exclude-positions-with-poly-a-sequences" title="Permalink to this headline"></a></h4>
<p>Although DMS and SHAPE reagents do modify A residues that are not immobilized
by base pairing, stretches of consecutive A residues tend to have very low
mutation rates due to an artifact from the reverse transcriptases that are used
commonly for mutational profiling (including TGIRT-III and SuperScript II)
[<a class="reference external" href="https://doi.org/10.1021/acs.biochem.0c00020">Kladwang et al. (2020)</a>].
Thus, using poly(A) sequences in structural analyses can give erroneous results.
SEISMIC-RNA automatically excludes all positions within stretches of 5 or more
consecutive A residues.
You can customize this behavior with <code class="docutils literal notranslate"><span class="pre">--exclude-polya</span></code> followed by the minimum
length of poly(A) sequences to exclude.
To disable poly(A) exclusion, use <code class="docutils literal notranslate"><span class="pre">--exclude-polya</span> <span class="pre">0</span></code>.</p>
</section>
<section id="exclude-arbitary-positions">
<h4>Exclude arbitary positions<a class="headerlink" href="#exclude-arbitary-positions" title="Permalink to this headline"></a></h4>
<p>You can also exclude any arbitary positions from any reference sequence.
A common reason to exclude a position is if the base is modified endogenously
in a way that causes mutations during reverse transcription.
To exclude an arbitrary position, use <code class="docutils literal notranslate"><span class="pre">--exclude-file</span></code> followed by a file of
all positions to exclude, in <a class="reference internal" href="../../formats/list/listpos.html"><span class="doc">List of Positions</span></a> format.</p>
</section>
</section>
<section id="mask-setting-filter-reads">
<h3>Mask setting: Filter reads<a class="headerlink" href="#mask-setting-filter-reads" title="Permalink to this headline"></a></h3>
<p>The second substep of masking is filtering reads.
You can filter reads based on three criteria, in this order:</p>
<section id="filter-reads-by-fraction-of-informative-positions">
<h4>Filter reads by fraction of informative positions<a class="headerlink" href="#filter-reads-by-fraction-of-informative-positions" title="Permalink to this headline"></a></h4>
<p>For some applications, such as finding alternative structures, every read must
span the vast majority of positions in the section of the reference.
You can set a limit on the minimum number of informative bases in the read,
as a fraction of the number of non-excluded positions in the section, using
<code class="docutils literal notranslate"><span class="pre">--min-finfo-read</span> <span class="pre">{f}</span></code>.
For example, to require 95% of the non-excluded positions in the section to be
informative, use <code class="docutils literal notranslate"><span class="pre">--min-finfo-read</span> <span class="pre">0.95</span></code>.
If the section had 296 positions, and 141 remained after excluding positions
(see <a class="reference internal" href="#mask-exclude"><span class="std std-ref">Mask setting: Exclude positions</span></a>), then a read with 137 informative positions would
have an informed fraction of 97% and be kept, but a read with 133 informative
positions would have an informed fraction of 94% and be discarded.</p>
</section>
<section id="filter-reads-by-fraction-of-mutated-positions">
<h4>Filter reads by fraction of mutated positions<a class="headerlink" href="#filter-reads-by-fraction-of-mutated-positions" title="Permalink to this headline"></a></h4>
<p>Rarely, a read may have an excessive number of mutations, possibly because it
underwent template switching during reverse transcription or misaligned during
the Align step.
You can set a limit to the fraction of mutated positions in the read using
<code class="docutils literal notranslate"><span class="pre">--max-fmut-read</span></code>.
For example, using the default limit of 10%, a read with 121 informative and
15 mutated positions would have a mutated fraction of 15 / 121 = 12% and be
discarded, but a read with 121 informative and 10 mutated positions would have
a mutated fraction of 8% and be kept.
Using <code class="docutils literal notranslate"><span class="pre">--max-fmut-read</span> <span class="pre">1.0</span></code> disables filtering by fraction mutated.</p>
</section>
<section id="filter-reads-by-space-between-mutations">
<h4>Filter reads by space between mutations<a class="headerlink" href="#filter-reads-by-space-between-mutations" title="Permalink to this headline"></a></h4>
<p>Reads with closely spaced mutations are very underrepresented in mutational
profiling data, presumably because reverse transcripases struggle to read
through closely spaced pairs of modifications [<a class="reference external" href="https://doi.org/10.1038/s41586-020-2253-5">Tomezsko et al. (2020)</a>].
Therefore, the data are biased towards reads without closely spaced mutations,
which would skew the mutation rates.
However, SEISMIC-RNA can correct the bias: first by removing any reads that
did happen to have mutations close together, then calculating the mutation
rates without such reads, and inferring what the mutation rates would have
been if no reads had dropped out.</p>
<p>The correction for observer bias is most important for finding alternative
structures and (to minimize surprises) does not run by default.
You can correct observer bias using <code class="docutils literal notranslate"><span class="pre">--min-mut-gap</span></code> followed by the minimum
number of non-mutated bases that must separate two mutations; reads with any
pair of mutations closer than this gap are discarded.
If you correct for observer bias, then we recommend using <code class="docutils literal notranslate"><span class="pre">--min-mut-gap</span> <span class="pre">3</span></code>,
based on our previous findings in <a class="reference external" href="https://doi.org/10.1038/s41586-020-2253-5">Tomezsko et al. (2020)</a>.</p>
</section>
</section>
<section id="mask-setting-filter-positions">
<h3>Mask setting: Filter positions<a class="headerlink" href="#mask-setting-filter-positions" title="Permalink to this headline"></a></h3>
<p>The third substep of masking is filtering positions.
You can filter positions based on two criteria, in this order:</p>
<section id="filter-positions-by-number-of-informative-reads">
<h4>Filter positions by number of informative reads<a class="headerlink" href="#filter-positions-by-number-of-informative-reads" title="Permalink to this headline"></a></h4>
<p>Estimating the fraction of mutated reads at a given position requires a large
number of reads so that the uncertainty (i.e. error bars) is much smaller than
the fraction of mutated reads.
The default minimum number of informative reads is 1000, which we have found
to yield a reasonably small uncertainties in the mutation fraction.
You can specify the minimum number of informative reads at each position using
<code class="docutils literal notranslate"><span class="pre">--min-ninfo-pos</span></code>.
We discourage going below 1000 reads unless you have multiple replicates, the
total number of informative reads at the position among all replicates is at
least 1000, and the mutation rates of the replicates correlate with a Pearson
or Spearman coefficient of at least 0.95.</p>
</section>
<section id="filter-positions-by-fraction-of-mutated-reads">
<h4>Filter positions by fraction of mutated reads<a class="headerlink" href="#filter-positions-by-fraction-of-mutated-reads" title="Permalink to this headline"></a></h4>
<p>Mutational profiling generally yields fractions of mutated reads up to 0.3.
Positions with fractions of mutated reads that exceed 0.5 are likely to be
mutated for some reason other than chemcial probing, such as misalignment
(especially when two or more reference sequences are very similar), an
endogenous RNA modification (if the RNA came from cells), a mistake in the
template DNA (if the RNA was transcribed <em>in vitro</em>), or a mistake in the
reference sequence.
Thus, SEISMIC-RNA discards positions with a fraction of mutated reads greater
than 0.5, by default.
You can set the maximum fraction of mutated reads using <code class="docutils literal notranslate"><span class="pre">--max-fmut-pos</span> <span class="pre">{f}</span></code>.</p>
</section>
</section>
</section>
<section id="mask-output-files">
<h2>Mask: Output files<a class="headerlink" href="#mask-output-files" title="Permalink to this headline"></a></h2>
<p>All output files go into the directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/mask/{ref}/{sect}</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory, <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the sample, <code class="docutils literal notranslate"><span class="pre">{ref}</span></code>
is the reference, and <code class="docutils literal notranslate"><span class="pre">{sect}</span></code> is the section.</p>
<section id="mask-output-file-batch-of-masked-reads">
<h3>Mask output file: Batch of masked reads<a class="headerlink" href="#mask-output-file-batch-of-masked-reads" title="Permalink to this headline"></a></h3>
<p>Each batch of masked reads contains a <code class="docutils literal notranslate"><span class="pre">MaskBatchIO</span></code> object and is saved to the
file <code class="docutils literal notranslate"><span class="pre">mask-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <span class="xref std std-doc">../../data/mask/mask</span> for more information on the data structure.
See <a class="reference internal" href="../../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for more information on brickle files.</p>
</section>
<section id="mask-output-file-mask-report">
<h3>Mask output file: Mask report<a class="headerlink" href="#mask-output-file-mask-report" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA also writes a report file, <code class="docutils literal notranslate"><span class="pre">mask-report.json</span></code>, that records the
settings you used for running the Mask step and summarizes the results, such as
which and how many positions and reads were filtered out for each reason.
See <a class="reference internal" href="../../formats/report/mask.html"><span class="doc">Mask Report</span></a> for more information.</p>
</section>
</section>
<section id="mask-troubleshoot-and-optimize">
<h2>Mask: Troubleshoot and optimize<a class="headerlink" href="#mask-troubleshoot-and-optimize" title="Permalink to this headline"></a></h2>
<section id="too-many-reads-are-filtered-out">
<span id="mask-too-many-reads"></span><h3>Too many reads are filtered out<a class="headerlink" href="#too-many-reads-are-filtered-out" title="Permalink to this headline"></a></h3>
<p>In the Mask report file, check the settings for filtering reads and the number
of reads removed by each filter.</p>
<ul class="simple">
<li><p>If the settings appear too strict, then rerun the Mask step using new settings
that would keep more reads, such as a lower value for <code class="docutils literal notranslate"><span class="pre">--min-finfo-read</span></code> or
<code class="docutils literal notranslate"><span class="pre">--min-mut-gap</span></code> or a higher value for <code class="docutils literal notranslate"><span class="pre">--max-fmut-read</span></code>.</p></li>
<li><p>If you are losing too many reads for having too few informative positions,
then also double check the 5’ and 3’ ends of the section over which you are
masking and ensure that the section is not too long compared to your reads.</p></li>
<li><p>If you are losing too many reads for having too many mutations, or mutations
that are too close together, then there may be a problem with the data quality
that is causing excessive mutations, such as</p>
<ul>
<li><p>Your RNA was low-quality, contained many endogenous modififications that
caused mutations during RT, or did not have the sequence you expected.</p></li>
<li><p>Your sequencing run gave low-quality base calls (check the FastQC reports)
that you did not trim (in Align) or flag as ambiguous (in Relate).</p></li>
<li><p>You aligned to reference sequences that differ from the actual RNA.</p></li>
<li><p>Many reads misaligned (possibly because your FASTA file has several similar
sequences), and your mapping quality filter did not remove misaligned reads.</p></li>
<li><p>In the Mask step, you did not pre-exclude problematic positions, such as
sites of endogenous RNA modifications.</p></li>
</ul>
</li>
</ul>
</section>
<section id="too-many-positions-are-filtered-out">
<h3>Too many positions are filtered out<a class="headerlink" href="#too-many-positions-are-filtered-out" title="Permalink to this headline"></a></h3>
<p>In the Mask report file, check the settings for filtering positions and the
number of positions removed by each filter.</p>
<ul class="simple">
<li><p>If the settings appear too strict, then rerun the Mask step using new settings
that would keep more positions, such as a lower value for <code class="docutils literal notranslate"><span class="pre">--min-ninfo-pos</span></code>
or a higher value for <code class="docutils literal notranslate"><span class="pre">--max-fmut-pos</span></code>.</p></li>
<li><p>If you are losing too many positions for having too few informative reads,
then there are three likely reasons:</p>
<ul>
<li><p>Your sample was sequenced with insufficient depth or quality.</p></li>
<li><p>Your sample contained insufficient RNAs from this reference/section.</p></li>
<li><p>You lost too many reads during filtering; see <a class="reference internal" href="#mask-too-many-reads"><span class="std std-ref">Too many reads are filtered out</span></a>.</p></li>
</ul>
</li>
<li><p>If you are losing too many positions for having too many mutations, then there
may be a problem with the data quality that is causing excessive mutations,
such as</p>
<ul>
<li><p>Your RNA was low-quality, contained many endogenous modififications that
caused mutations during RT, or did not have the sequence you expected.</p></li>
<li><p>Your sequencing run gave low-quality base calls (check the FastQC reports)
that you did not trim (in Align) or flag as ambiguous (in Relate).</p></li>
<li><p>You aligned to reference sequences that differ from the actual RNA.</p></li>
<li><p>Many reads misaligned (possibly because your FASTA file has several similar
sequences), and your mapping quality filter did not remove misaligned reads.</p></li>
</ul>
</li>
</ul>
</section>
<section id="mask-crashes-or-hangs-while-producing-few-or-no-batch-files">
<h3>Mask crashes or hangs while producing few or no batch files<a class="headerlink" href="#mask-crashes-or-hangs-while-producing-few-or-no-batch-files" title="Permalink to this headline"></a></h3>
<p>Most likely, your system has run out of memory.
You can confirm using a program that monitors memory usage (such as <code class="docutils literal notranslate"><span class="pre">top</span></code> in a
Linux/macOS terminal, Activity Monitor on macOS, or Task Manager on Windows).
If so, then you can either</p>
<ul class="simple">
<li><p>Use fewer processors (with <code class="docutils literal notranslate"><span class="pre">--max-procs</span></code>) to limit the memory usage, at the
cost of slower processing.</p></li>
<li><p>Rerun Relate with smaller batches (with <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code>) to limit the size of
each batch, at the cost of having more files with a larger total size.</p></li>
</ul>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pool.html" class="btn btn-neutral float-left" title="Pool: Merge samples (vertically) from the Relate step" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="cluster.html" class="btn btn-neutral float-right" title="Cluster: Infer alternative structures by clustering reads’ mutations" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>