<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Table: Count mutations for each read and position &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Fold: Predict RNA secondary structures using mutation rates" href="fold.html" />
    <link rel="prev" title="Join: Merge sections (horizontally) from the Mask or Cluster step" href="join.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="align.html">Align: Trim FASTQ files and align them to reference sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate: Compute relationships between references and aligned reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="pool.html">Pool: Merge samples (vertically) from the Relate step</a></li>
<li class="toctree-l3"><a class="reference internal" href="mask.html">Mask: Define mutations and sections to filter reads and positions</a></li>
<li class="toctree-l3"><a class="reference internal" href="cluster.html">Cluster: Infer alternative structures by clustering reads’ mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="join.html">Join: Merge sections (horizontally) from the Mask or Cluster step</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Table: Count mutations for each read and position</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#table-input-files">Table: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#table-settings">Table: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#table-output-files">Table: Output files</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="fold.html">Fold: Predict RNA secondary structures using mutation rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="graph.html">Graph: Plot data from tables and/or structures and compare samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3"><a class="reference internal" href="wf.html">Workflow: Run all steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lists.html">List Positions Matching Criteria</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adjclust.html">Add/Delete Orders to/from an Already-Clustered Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Table: Count mutations for each read and position</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/table.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="table-count-mutations-for-each-read-and-position">
<h1>Table: Count mutations for each read and position<a class="headerlink" href="#table-count-mutations-for-each-read-and-position" title="Link to this heading"></a></h1>
<section id="table-input-files">
<h2>Table: Input files<a class="headerlink" href="#table-input-files" title="Link to this heading"></a></h2>
<section id="table-input-file-report">
<h3>Table input file: Report<a class="headerlink" href="#table-input-file-report" title="Link to this heading"></a></h3>
<p>You can give any number of report files from the Relate, Mask, and Cluster steps
as inputs for the Table step.
See <a class="reference internal" href="../inputs.html"><span class="doc">List Input Files</span></a> for ways to list multiple files.</p>
<p>To tabulate all results in <code class="docutils literal notranslate"><span class="pre">{out}</span></code>, you can use the command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">table</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span>
</pre></div>
</div>
</section>
</section>
<section id="table-settings">
<h2>Table: Settings<a class="headerlink" href="#table-settings" title="Link to this heading"></a></h2>
<section id="table-setting-suppress-per-position-and-per-read-tables">
<h3>Table setting: Suppress per-position and per-read tables<a class="headerlink" href="#table-setting-suppress-per-position-and-per-read-tables" title="Link to this heading"></a></h3>
<p>SEISMIC-RNA outputs all possible tables by default.
To suppress per-position and per-read tables, use <code class="docutils literal notranslate"><span class="pre">--no-table-pos</span></code> and
<code class="docutils literal notranslate"><span class="pre">--no-table-read</span></code>, respectively.</p>
</section>
</section>
<section id="table-output-files">
<h2>Table: Output files<a class="headerlink" href="#table-output-files" title="Link to this heading"></a></h2>
<p>All output files go into the directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/table/{ref}/{sect}</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory, <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the sample, <code class="docutils literal notranslate"><span class="pre">{ref}</span></code>
is the reference, and <code class="docutils literal notranslate"><span class="pre">{sect}</span></code> is the section.
Each output file from Relate, Mask, and Cluster reports are prefixed with
<code class="docutils literal notranslate"><span class="pre">relate-</span></code>, <code class="docutils literal notranslate"><span class="pre">mask-</span></code>, and <code class="docutils literal notranslate"><span class="pre">clust-</span></code>, respectively.</p>
<section id="table-output-file-relationships-per-position">
<h3>Table output file: Relationships per position<a class="headerlink" href="#table-output-file-relationships-per-position" title="Link to this heading"></a></h3>
<p>SEISMIC-RNA outputs the number of relationships at each position to a CSV file
named <code class="docutils literal notranslate"><span class="pre">{step}-per-pos.csv</span></code>, where <code class="docutils literal notranslate"><span class="pre">{step}</span></code> is <code class="docutils literal notranslate"><span class="pre">relate</span></code>/<code class="docutils literal notranslate"><span class="pre">mask</span></code>/<code class="docutils literal notranslate"><span class="pre">clust</span></code>.</p>
</section>
<section id="table-output-file-relationships-per-read">
<h3>Table output file: Relationships per read<a class="headerlink" href="#table-output-file-relationships-per-read" title="Link to this heading"></a></h3>
<p>SEISMIC-RNA outputs the number of relationships in each read to a CSV file named
<code class="docutils literal notranslate"><span class="pre">{step}-per-read.csv.gz</span></code>, where <code class="docutils literal notranslate"><span class="pre">{step}</span></code> is <code class="docutils literal notranslate"><span class="pre">relate</span></code>/<code class="docutils literal notranslate"><span class="pre">mask</span></code>/<code class="docutils literal notranslate"><span class="pre">clust</span></code>.</p>
</section>
<section id="table-output-file-reads-per-cluster">
<h3>Table output file: Reads per cluster<a class="headerlink" href="#table-output-file-reads-per-cluster" title="Link to this heading"></a></h3>
<p>For Cluster reports, SEISMIC-RNA outputs the number of reads in each cluster to
a CSV file named <code class="docutils literal notranslate"><span class="pre">clust-freq.csv</span></code>.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="join.html" class="btn btn-neutral float-left" title="Join: Merge sections (horizontally) from the Mask or Cluster step" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="fold.html" class="btn btn-neutral float-right" title="Fold: Predict RNA secondary structures using mutation rates" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>