<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Workflow: Run all steps &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Graph Results" href="../graph/index.html" />
    <link rel="prev" title="Export: Export a file of each sample for the seismic-graph web app" href="export.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../whatdoes/index.html">What it Does</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="align.html">Align: Trim FASTQ files and align them to reference sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate: Compute relationships between references and aligned reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="mask.html">Mask: Define mutations and sections to filter reads and positions</a></li>
<li class="toctree-l3"><a class="reference internal" href="cluster.html">Cluster: Infer alternative structures by clustering reads’ mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="table.html">Table: Count mutations for each read and position</a></li>
<li class="toctree-l3"><a class="reference internal" href="fold.html">Fold: Predict RNA secondary structures using mutation rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="graph.html">Graph: Plot data from tables and/or structures and compare samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Workflow: Run all steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#workflow-input-files">Workflow: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#workflow-settings">Workflow: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#workflow-output-files">Workflow: Output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#workflow-troubleshoot-and-optimize">Workflow: Troubleshoot and optimize</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Workflow: Run all steps</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/wf.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="workflow-run-all-steps">
<h1>Workflow: Run all steps<a class="headerlink" href="#workflow-run-all-steps" title="Permalink to this headline"></a></h1>
<p>For convenience, <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code> runs all steps of the main workflow:</p>
<img alt="../../_images/wf.png" src="../../_images/wf.png" />
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When you run the workflow with <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>, the steps Cluster, Fold, and
Export do not run by default, but you can turn them on with options:</p>
<ul class="simple">
<li><p>To cluster, use <code class="docutils literal notranslate"><span class="pre">--max-clusters</span></code> (<code class="docutils literal notranslate"><span class="pre">-k</span></code>) followed by the maximum number
of clusters (in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>, <code class="docutils literal notranslate"><span class="pre">-k</span></code> defaults to 0 – no clustering).</p></li>
<li><p>To fold, use <code class="docutils literal notranslate"><span class="pre">--fold</span></code>.</p></li>
<li><p>To export, use <code class="docutils literal notranslate"><span class="pre">--export</span></code>.</p></li>
</ul>
</div>
<p>SEISMIC-RNA will process every input file from the point that it enters the
workflow, following the gray arrays, until the end of the entire workflow.
For example, for each file of read sequences that you input with <code class="docutils literal notranslate"><span class="pre">-x</span></code>, <code class="docutils literal notranslate"><span class="pre">-y</span></code>,
or <code class="docutils literal notranslate"><span class="pre">-z</span></code>, it will</p>
<ol class="arabic simple">
<li><p>run Align on the FASTQ files of reads to yield alignment maps</p></li>
<li><p>run Relate on the alignment maps to yield Relate reports</p></li>
<li><p>run Mask on the Relate reports to yield Mask reports</p></li>
<li><p>(optionally) run Cluster on the Mask reports to make Cluster reports</p></li>
<li><p>run Table on the Relate, Mask, and (optionally) Cluster reports to yield
table files</p></li>
<li><p>(optionally) run Fold on the Mask and (optionally) Cluster table files to
yield predicted RNA secondary structure models</p></li>
<li><p>(optionally) run Export on the Relate, Mask, and (optionally) Cluster table
files and (optionally) structure models to yield sample results</p></li>
<li><p>run Graph on the Relate, Mask, and (optionally) Cluster tables and
(optionally) structure models to yield graph files</p></li>
</ol>
<p>By contrast, for each Mask table file you input, SEISMIC-RNA will only</p>
<ol class="arabic simple">
<li><p>(optionally) run Fold on the table file to yield predicted RNA secondary
structure models</p></li>
<li><p>(optionally) run Export on the table file and (optionally) structure models
to yield sample results</p></li>
<li><p>run Graph on the table file and (optionally) structure models to yield graph
files</p></li>
</ol>
<section id="workflow-input-files">
<h2>Workflow: Input files<a class="headerlink" href="#workflow-input-files" title="Permalink to this headline"></a></h2>
<p>You can give any input file to <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code> if you could give it to any
individual command in the workflow.</p>
<p>The only required input file is of the reference sequences, which is used by the
Align and Relate steps; see <a class="reference internal" href="relate.html#relate-refs"><span class="std std-ref">Relate input file: Reference sequences</span></a> for more information.
This file must be the first positional argument.</p>
<p>You can list additional input files (e.g. alignment maps, report files, table
files) as positional arguments after the reference sequences, using any of the
ways to list input files (see <a class="reference internal" href="../inputs.html"><span class="doc">List Input Files</span></a>), for example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">wf</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="n">sample1</span><span class="o">/</span><span class="n">refA</span><span class="o">.</span><span class="n">bam</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/*/</span><span class="n">relate</span><span class="o">/</span><span class="n">refB</span><span class="o">/</span><span class="n">relate</span><span class="o">-</span><span class="n">report</span><span class="o">.</span><span class="n">json</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/*/</span><span class="n">table</span><span class="o">/</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs.fa}</span></code> is the path of your FASTA file of reference sequences and
<code class="docutils literal notranslate"><span class="pre">{out}</span></code> is your top-level output directory.</p>
<p>You can put keyword input files (<code class="docutils literal notranslate"><span class="pre">-x</span></code>/<code class="docutils literal notranslate"><span class="pre">-y</span></code>/<code class="docutils literal notranslate"><span class="pre">-z</span></code>/<code class="docutils literal notranslate"><span class="pre">-X</span></code>/<code class="docutils literal notranslate"><span class="pre">-Y</span></code>/<code class="docutils literal notranslate"><span class="pre">-Z</span></code>) in
any locations amid positional arguments, for example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">wf</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample4</span><span class="o">/</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample5</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample6</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">sample1</span><span class="o">/</span><span class="n">refA</span><span class="o">.</span><span class="n">bam</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/*/</span><span class="n">relate</span><span class="o">/</span><span class="n">refB</span><span class="o">/</span><span class="n">relate</span><span class="o">-</span><span class="n">report</span><span class="o">.</span><span class="n">json</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/*/</span><span class="n">table</span><span class="o">/</span>
</pre></div>
</div>
</section>
<section id="workflow-settings">
<h2>Workflow: Settings<a class="headerlink" href="#workflow-settings" title="Permalink to this headline"></a></h2>
<p>You can use any option in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code> if you could use it in any individual
command in the workflow.
Additionally, <code class="docutils literal notranslate"><span class="pre">--fold</span></code>/<code class="docutils literal notranslate"><span class="pre">--no-fold</span></code> and <code class="docutils literal notranslate"><span class="pre">--export</span></code>/<code class="docutils literal notranslate"><span class="pre">--no-export</span></code> control
whether the Fold and Export steps, respectively, will run in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>
(both are “no” by default).
Options can be mixed in any locations amid positional arguments, for example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">wf</span> <span class="o">--</span><span class="n">force</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample4</span><span class="o">/</span> <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">finfo</span><span class="o">-</span><span class="n">read</span> <span class="mf">0.95</span> <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">ninfo</span><span class="o">-</span><span class="n">pos</span> <span class="mi">100000</span> <span class="o">-</span><span class="n">k</span> <span class="mi">3</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span><span class="o">/*/</span><span class="n">table</span><span class="o">/</span> <span class="o">--</span><span class="n">fold</span>
</pre></div>
</div>
</section>
<section id="workflow-output-files">
<h2>Workflow: Output files<a class="headerlink" href="#workflow-output-files" title="Permalink to this headline"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code> command creates all output files that would have been created
by running every step of the workflow individually.
All output files go into the directory that you specify with <code class="docutils literal notranslate"><span class="pre">--out-dir</span></code>,
which is <code class="docutils literal notranslate"><span class="pre">./out</span></code> by default.</p>
</section>
<section id="workflow-troubleshoot-and-optimize">
<h2>Workflow: Troubleshoot and optimize<a class="headerlink" href="#workflow-troubleshoot-and-optimize" title="Permalink to this headline"></a></h2>
<p>Refer to the advice for the individual step you are troubleshooting/optimizing.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="export.html" class="btn btn-neutral float-left" title="Export: Export a file of each sample for the seismic-graph web app" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../graph/index.html" class="btn btn-neutral float-right" title="Graph Results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>