<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Running the Workflow &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Cleaning FASTA Files" href="cleanfa.html" />
    <link rel="prev" title="Manuals" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            seismic-rna
              <img src="../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install/index.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Manuals</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Running the Workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#align-trim-fastq-files-and-align-them-to-reference-sequences">Align: Trim FASTQ files and align them to reference sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#align-input-files">Align: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#align-settings">Align: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#align-output-files">Align: Output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#align-troubleshooting">Align: Troubleshooting</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#relate-compute-relationships-between-references-and-aligned-reads">Relate: Compute relationships between references and aligned reads</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#relate-input-files">Relate: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#relate-settings">Relate: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#relate-output-files">Relate: Output files</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#mask-define-mutations-and-sections-to-filter-reads-and-positions">Mask: Define mutations and sections to filter reads and positions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#mask-input-files">Mask: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mask-settings">Mask: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id4">Relate: Output files</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#run-the-entire-workflow-from-demult-align-to-graph-export">Run the entire workflow, from demult/align to graph/export.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="cleanfa.html">Cleaning FASTA Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="inputs.html">Listing Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="parallel.html">Parallelizing Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="logging.html">Logging Messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Manuals</a></li>
      <li class="breadcrumb-item active">Running the Workflow</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/manuals/workflow.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="running-the-workflow">
<h1>Running the Workflow<a class="headerlink" href="#running-the-workflow" title="Permalink to this headline"></a></h1>
<p>There are three points from which you can begin the main workflow:</p>
<ul class="simple">
<li><p>If you are starting from files of raw sequencing reads (FASTQ format)
that must be demultiplexed (i.e. split into subfiles using barcodes),
then begin at the step <a class="reference internal" href="#wf-demult"><span class="std std-ref">Align: Trim FASTQ files and align them to reference sequences</span></a>.
(If you do not know whether you need this feature, then you probably
do not need it.)</p></li>
<li><p>If you are starting from files of raw sequencing reads (FASTQ format)
that do not need to be demultiplexed (i.e. split into subfiles), then
begin at the step <a class="reference internal" href="#wf-align"><span class="std std-ref">Align: Trim FASTQ files and align them to reference sequences</span></a>.
<strong>Most users will begin at this step.</strong>.</p></li>
<li><p>If you are starting from files of aligned sequencing reads (SAM, BAM,
or CRAM format), then begin at the step <a class="reference internal" href="#wf-relate"><span class="std std-ref">Relate: Compute relationships between references and aligned reads</span></a>.</p></li>
</ul>
<p>Each step of the workflow can be run individually by typing the command
to which it corresponds; for example <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> runs alignment.
For convenience, SEISMIC-RNA also provides the command <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>,
which runs every step of the main workflow automatically.
See <a class="reference internal" href="#wf-wf"><span class="std std-ref">Run the entire workflow, from demult/align to graph/export.</span></a> for more information.</p>
<section id="align-trim-fastq-files-and-align-them-to-reference-sequences">
<span id="wf-align"></span><span id="wf-demult"></span><h2>Align: Trim FASTQ files and align them to reference sequences<a class="headerlink" href="#align-trim-fastq-files-and-align-them-to-reference-sequences" title="Permalink to this headline"></a></h2>
<section id="align-input-files">
<h3>Align: Input files<a class="headerlink" href="#align-input-files" title="Permalink to this headline"></a></h3>
<section id="align-input-file-reference-sequences">
<h4>Align input file: Reference sequences<a class="headerlink" href="#align-input-file-reference-sequences" title="Permalink to this headline"></a></h4>
<p>Alignment requires exactly one file of reference sequences, which must
be DNA sequences (only A, C, G, T, and N) in FASTA format.
If needed, you may clean the FASTA file with the <a class="reference internal" href="cleanfa.html"><span class="doc">Cleaning FASTA Files</span></a> tool.
See <a class="reference internal" href="../formats/data/fasta.html"><span class="doc">FASTA: Reference sequences</span></a> for details.</p>
</section>
<section id="align-input-file-sequencing-reads">
<h4>Align input file: Sequencing reads<a class="headerlink" href="#align-input-file-sequencing-reads" title="Permalink to this headline"></a></h4>
<p>Alignment requires one or more files of sequencing reads, which must be
DNA sequences (only A, C, G, T, and N) in FASTQ format.
See <a class="reference internal" href="../formats/data/fastq.html"><span class="doc">FASTQ: Sequencing Reads</span></a> for details.</p>
<section id="sequencing-reads-can-be-single-end-or-paired-end">
<h5>Sequencing reads can be single-end or paired-end<a class="headerlink" href="#sequencing-reads-can-be-single-end-or-paired-end" title="Permalink to this headline"></a></h5>
<p>Sequencing reads can be single-end or paired-end (for more details, see
<a class="reference internal" href="../formats/data/fastq.html#fastq-endedness"><span class="std std-ref">Endedness: single-end and paired-end reads</span></a>).
The SEISMIC-RNA workflow can handle both single- and paired-end reads,
provided that no FASTQ file contains a mixture of both types of reads.</p>
</section>
<section id="sequencing-reads-can-come-from-whole-or-demultiplexed-samples">
<h5>Sequencing reads can come from whole or demultiplexed samples<a class="headerlink" href="#sequencing-reads-can-come-from-whole-or-demultiplexed-samples" title="Permalink to this headline"></a></h5>
<p>FASTQ files can contain reads from a whole sample (which can possibly
come from many different reference sequences) or can be demultiplexed
before alignment so that they contain reads from one reference sequence.</p>
</section>
<section id="how-to-specify-the-endedness-and-source-of-sequencing-reads">
<h5>How to specify the endedness and source of sequencing reads<a class="headerlink" href="#how-to-specify-the-endedness-and-source-of-sequencing-reads" title="Permalink to this headline"></a></h5>
<p>Specify the endedness (single-end, paired-end interleaved, or paired-end
separated) and source of the reads (whole or demultiplexed sample) in a
FASTQ file by giving the file name using the appropriate option below.
Each option has a “long form” (starting with <code class="docutils literal notranslate"><span class="pre">--</span></code>) and a “short form”
(starting with <code class="docutils literal notranslate"><span class="pre">-</span></code> and comprising a single character), which are both
indicated in this table, separated by a <code class="docutils literal notranslate"><span class="pre">/</span></code>:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 45%" />
<col style="width: 26%" />
<col style="width: 29%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Endedness</p></th>
<th class="head"><p>Whole Sample</p></th>
<th class="head"><p>Demultiplexed</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>paired-end, separate files</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqx/-x</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqx/-X</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>paired-end, interleaved</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqy/-y</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqy/-Y</span></code></p></td>
</tr>
<tr class="row-even"><td><p>single-end</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--fastqz/-z</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--dmfastqz/-Z</span></code></p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As with all options, the long and short forms are interchangable.
For brevity, the rest of this manual uses only the short forms.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If using paired-end reads in separate FASTQ files, both files of 1st
and 2nd mates are required. See <a class="reference internal" href="#fastq-pair"><span class="std std-ref">How to align a pair of FASTQ files (paired-end reads in separate files)</span></a> for instructions.</p>
</div>
</section>
<section id="how-to-align-one-fastq-file-single-end-or-interleaved-paired-end-reads">
<h5>How to align one FASTQ file (single-end or interleaved paired-end reads)<a class="headerlink" href="#how-to-align-one-fastq-file-single-end-or-interleaved-paired-end-reads" title="Permalink to this headline"></a></h5>
<p>To align a FASTQ of single-end reads from a whole sample, use <code class="docutils literal notranslate"><span class="pre">-z</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs.fa}</span></code> is the path to the FASTA file of reference sequences
and <code class="docutils literal notranslate"><span class="pre">{sample.fq.gz}</span></code> is the path to the FASTQ file of the sample.</p>
<p>For a FASTQ of paired-end, interleaved reads that were demultiplexed,
use <code class="docutils literal notranslate"><span class="pre">-Y</span></code> instead:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">Y</span> <span class="p">{</span><span class="n">sample</span><span class="o">/</span><span class="n">ref</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample/ref.fq.gz}</span></code> is the path to the FASTQ file containing
reads from only one reference in the sample.</p>
</section>
<section id="how-to-align-a-pair-of-fastq-files-paired-end-reads-in-separate-files">
<span id="fastq-pair"></span><h5>How to align a pair of FASTQ files (paired-end reads in separate files)<a class="headerlink" href="#how-to-align-a-pair-of-fastq-files-paired-end-reads-in-separate-files" title="Permalink to this headline"></a></h5>
<p>If your reads are paired-end and you have one FASTQ file containing all
1st mates and another containing all 2nd mates, then you will need to
provide both FASTQ files.
There are two methods:</p>
<ol class="arabic">
<li><p>Use the option <code class="docutils literal notranslate"><span class="pre">-x</span></code> (or <code class="docutils literal notranslate"><span class="pre">-X</span></code>) twice, once per FASTQ file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample_R1.fq.gz}</span></code> and <code class="docutils literal notranslate"><span class="pre">{sample_R2.fq.gz}</span></code> are the paths
to the FASTQ files of the 1st and 2nd mates, respectively.</p>
</li>
<li><p>Make a new directory, move both FASTQ files into that directory, and
provide the path to that directory with <code class="docutils literal notranslate"><span class="pre">-x</span></code> (or <code class="docutils literal notranslate"><span class="pre">-X</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
<span class="n">mv</span> <span class="p">{</span><span class="n">sample_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="p">{</span><span class="n">sample_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
<span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the new directory for both FASTQ files.</p>
</li>
</ol>
</section>
<section id="how-to-align-multiple-fastq-files-or-pairs-of-paired-end-fastq-files">
<h5>How to align multiple FASTQ files or pairs of paired-end FASTQ files<a class="headerlink" href="#how-to-align-multiple-fastq-files-or-pairs-of-paired-end-fastq-files" title="Permalink to this headline"></a></h5>
<p>There are three ways to align multiple FASTQ files (or pairs thereof):</p>
<ol class="arabic">
<li><p><strong>Use options more than once.</strong>
The options for FASTQ files can all be given multiple times, and can
even be mixed in one command.
For example, to align a pair of paired-end FASTQ files (sample 1),
an interleaved paired-end FASTQ file (sample 2), and two single-end
FASTQ files (samples 3 and 4), use the following options:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample1_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="p">{</span><span class="n">sample1_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">y</span> <span class="p">{</span><span class="n">sample2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample3</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span> <span class="o">-</span><span class="n">z</span> <span class="p">{</span><span class="n">sample4</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span><span class="p">}</span>
</pre></div>
</div>
<p>This method is most useful when you have a few FASTQ files.</p>
</li>
<li><p><strong>Group FASTQ files of the same type into a directory.</strong>
For example, suppose you have 63 FASTQ files each of paired-end 1st
mates (named <code class="docutils literal notranslate"><span class="pre">sample-1_R1.fq.gz</span></code> to <code class="docutils literal notranslate"><span class="pre">sample-63_R1.fq.gz</span></code>) and
2nd mates (named analogously but with <code class="docutils literal notranslate"><span class="pre">R2</span></code>), plus demultiplexed
single-end reads from three samples (I-III) and six references (A-F)
(named <code class="docutils literal notranslate"><span class="pre">sample-I/ref-A.fq.gz</span></code> to <code class="docutils literal notranslate"><span class="pre">sample-III/ref-F.fq.gz</span></code>).
Move the separate paired-end FASTQ files into their own directory,
and the demultiplexed single-end FASTQ files into another directory,
then run alignment by passing each directory of FASTQ files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>mkdir {paired}
mv sample-*_R?.fq.gz {paired}
mkdir {dm-single}
mv sample-I* {dm-single}
seismic align {refs.fa} -x {paired} -Z {dm-single}
</pre></div>
</div>
<p>This method is most useful when you have many FASTQ files.</p>
</li>
<li><p><strong>Combine methods 1 and 2.</strong>
Suppose you are working on two projects, have generated a set of
many FASTQ files for each project, and want to process both sets.
Currently, the FASTQ files for projects 1 and 2 are in directories
<code class="docutils literal notranslate"><span class="pre">proj1</span></code> and <code class="docutils literal notranslate"><span class="pre">proj2</span></code>, and you want to keep them separate.
You can process both directories with one command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">align</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="o">-</span><span class="n">x</span> <span class="n">proj1</span> <span class="o">-</span><span class="n">x</span> <span class="n">proj2</span>
</pre></div>
</div>
<p>This method is most useful when you have multiple directories of
FASTQ files that you would like to keep separate.</p>
</li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a directory is given for any of the FASTQ options, then it will
be searched for FASTQ files recursively, with no limit to the depth.
Thus, the given directory can have deeply nested subdirectories, and
SEISMIC-RNA will still find and process any FASTQ files within them.</p>
</div>
</section>
</section>
</section>
<section id="align-settings">
<h3>Align: Settings<a class="headerlink" href="#align-settings" title="Permalink to this headline"></a></h3>
<section id="align-setting-phred-score-encoding">
<h4>Align setting: Phred score encoding<a class="headerlink" href="#align-setting-phred-score-encoding" title="Permalink to this headline"></a></h4>
<p>SEISMIC-RNA defaults to using Phred+33 encoding for FASTQ files, which
is standard on modern Illumina sequencers.
To change the Phred score encoding, use the option <code class="docutils literal notranslate"><span class="pre">--phred-enc</span></code>.
See <a class="reference internal" href="../formats/data/fastq.html#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a> for more information.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If your FASTQ files do not use the Phred+33 encoding, then you must
specify the correct Phred score encoding, or else Cutadapt and/or
Bowtie 2 can produce incorrect output or fail outright.</p>
</div>
<p>If you do not know the encoding scheme of your FASTQ files, then you may
be able to determine it by using <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> or <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> (which
runs FastQC automatically).
In the HTML report file generated by FastQC, check the “Encoding” field
in the “Basic Statisics” section:</p>
<ul class="simple">
<li><p>If the Encoding field says <code class="docutils literal notranslate"><span class="pre">Illumina</span> <span class="pre">1.0</span></code> to <code class="docutils literal notranslate"><span class="pre">1.7</span></code>, then your
FASTQ files use Phred+64 encoding (<code class="docutils literal notranslate"><span class="pre">--phred-enc</span> <span class="pre">64</span></code>).</p></li>
<li><p>If the Encoding field says <code class="docutils literal notranslate"><span class="pre">Illumina</span> <span class="pre">1.8</span></code> or greater, then your
FASTQ files use Phred+33 encoding (<code class="docutils literal notranslate"><span class="pre">--phred-enc</span> <span class="pre">33</span></code>, the default).</p></li>
<li><p>Otherwise, you will need to search elsewhere for your encoding scheme
to determine the Phred score offset.</p></li>
</ul>
</section>
<section id="align-setting-quality-assessment-with-fastqc">
<h4>Align setting: Quality assessment with FastQC<a class="headerlink" href="#align-setting-quality-assessment-with-fastqc" title="Permalink to this headline"></a></h4>
<p>By default, each FASTQ file is processed with <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, both before and
after trimming, in order to find any potential problems.
FastQC can be disabled with the flag <code class="docutils literal notranslate"><span class="pre">--no-fastqc</span></code>.
To enable automatic extraction of the zipped output files from FastQC,
add the flag <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code>.</p>
</section>
<section id="align-setting-trimming-reads-with-cutadapt">
<h4>Align setting: Trimming reads with Cutadapt<a class="headerlink" href="#align-setting-trimming-reads-with-cutadapt" title="Permalink to this headline"></a></h4>
<p>By default, each FASTQ file and pair of mated FASTQ files is trimmed for
adapters and low-quality bases using <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">Cutadapt</a>. To disable trimming,
add the flag <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code>.</p>
<section id="how-to-trim-adapter-sequences">
<h5>How to trim adapter sequences<a class="headerlink" href="#how-to-trim-adapter-sequences" title="Permalink to this headline"></a></h5>
<p>By default, SEISMIC-RNA uses the standard, minimal adapter sequences for
Illumina sequencing runs for both read 1 and (if paired-end) read 2:</p>
<ul class="simple">
<li><p>5’: <code class="docutils literal notranslate"><span class="pre">GCTCTTCCGATCT</span></code></p></li>
<li><p>3’: <code class="docutils literal notranslate"><span class="pre">AGATCGGAAGAGC</span></code></p></li>
</ul>
<p>To use another adapter, type its sequence after the appropriate option:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 23%" />
<col style="width: 23%" />
<col style="width: 54%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Side</p></th>
<th class="head"><p>Read</p></th>
<th class="head"><p>Option</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>5’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>5’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g2</span></code></p></td>
</tr>
<tr class="row-even"><td><p>3’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>3’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a2</span></code></p></td>
</tr>
</tbody>
</table>
</section>
<section id="how-to-trim-low-quality-base-calls">
<h5>How to trim low-quality base calls<a class="headerlink" href="#how-to-trim-low-quality-base-calls" title="Permalink to this headline"></a></h5>
<p>Base calls on either end of a read that fall below a minimum Phred score
quality are trimmed with Cutadapt.
The default minimum quality is 25, which corresponds to a probability of
1 - 10<sup>-2.5</sup> = 0.997 that the base call is correct.
(See <a class="reference internal" href="../formats/data/fastq.html#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a> for more details).
To change the quality threshold, use the option <code class="docutils literal notranslate"><span class="pre">--min-phred</span></code>.</p>
</section>
<section id="how-to-use-cutadapt-to-trim-dark-cycles-for-illumina-nextseq-or-iseq">
<h5>How to use Cutadapt to trim dark cycles (for Illumina NextSeq or iSeq)<a class="headerlink" href="#how-to-use-cutadapt-to-trim-dark-cycles-for-illumina-nextseq-or-iseq" title="Permalink to this headline"></a></h5>
<p>On some Illumina sequencers (e.g. NextSeq, iSeq), the probes used to
detect G bases emit no light. Hence, these instruments will label a base
call as a G if it appears dark. If sequencing reaches the end of a read,
then there will be no more bases to sequence, so every cycle thereafter
will be dark, causing a string of Gs to be added to the 3’ end of the
read. Using the option <code class="docutils literal notranslate"><span class="pre">--cut-nextseq</span></code> tells Cutadapt to trim off any
high-quality G bases from the 3’ end of each read. This may improve the
alignment (especially in end-to-end mode) but also removes real G bases
from the 3’ ends of reads (since they cannot be distinguished from any
artefactual G bases).</p>
</section>
<section id="how-to-further-customize-read-trimming">
<h5>How to further customize read trimming<a class="headerlink" href="#how-to-further-customize-read-trimming" title="Permalink to this headline"></a></h5>
<p>See <a class="reference internal" href="../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can
use with Cutadapt, and the <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/reference.html">Cutadapt reference guide</a> for details on
each of these options.
These options should suffice for most users.
If you require a more customized adapter trimming workflow, you can trim
your FASTQ files outside of SEISMIC-RNA, then perform alignment within
SEISMIC-RNA, using the option <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code> to disable additional adapter
trimming.</p>
</section>
</section>
<section id="align-setting-mapping-reads-with-bowtie-2">
<h4>Align setting: Mapping reads with Bowtie 2<a class="headerlink" href="#align-setting-mapping-reads-with-bowtie-2" title="Permalink to this headline"></a></h4>
<section id="how-to-pre-build-a-bowtie-2-index-optional">
<h5>How to pre-build a Bowtie 2 index (optional)<a class="headerlink" href="#how-to-pre-build-a-bowtie-2-index-optional" title="Permalink to this headline"></a></h5>
<p>Bowtie 2 requires the FASTA file of reference sequences to be indexed.
A Bowtie 2 index comprises six files, all in the same directory as and
with the same name as the FASTA file, with the extensions <code class="docutils literal notranslate"><span class="pre">.1.bt2</span></code>,
<code class="docutils literal notranslate"><span class="pre">.2.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.3.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.4.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.rev.1.bt2</span></code>, and <code class="docutils literal notranslate"><span class="pre">.rev.2.bt2</span></code>.</p>
<p>If the index is missing, then SEISMIC-RNA will create a temporary index
automatically using <code class="docutils literal notranslate"><span class="pre">bowtie2-build</span></code> each time you run alignment.
Automatic indexing is efficient when the FASTA file is small: several
hundred reference sequences or fewer.
For larger FASTA files (e.g. a whole eukaryotic transcriptome), building
a temporary index each time alignment is run becomes costly.
In this case, it is more efficient to pre-build the index, which you can
do with this command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bowtie2</span><span class="o">-</span><span class="n">build</span> <span class="p">{</span><span class="n">refs</span><span class="p">}</span><span class="o">.</span><span class="n">fa</span> <span class="p">{</span><span class="n">refs</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs}</span></code> is the path to and name of your FASTA file.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer">Bowtie 2 Indexer manual</a> for more details.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you use a pre-built Bowtie 2 index, then SEISMIC-RNA does <em>not</em>
verify that the index was actually built from the FASTA file of the
same name.
You can assume the index is correct if you build it using the above
command and avoid modifying or replacing the FASTA and index files.
Discrepancies between the FASTA file and the index files can crash
the <code class="docutils literal notranslate"><span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">relate</span></code> steps or produce erroneous results.</p>
</div>
</section>
<section id="how-to-choose-between-local-and-end-to-end-alignment">
<h5>How to choose between local and end-to-end alignment<a class="headerlink" href="#how-to-choose-between-local-and-end-to-end-alignment" title="Permalink to this headline"></a></h5>
<p>During alignment, Bowtie 2 can either align the entirety of each read
(end-to-end mode) or find and align only the section of the read that
yields the best alignment score (local mode).
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#end-to-end-alignment-versus-local-alignment">description of alignment modes in Bowtie 2</a> for more details.</p>
<p>Generally, end-to-end mode yields spurious mutations (false positives)
at the ends of reads if the reads contain artifacts such as low-quality
base calls or untrimmed or improperly trimmed adapters.
Conversely, local mode misses real mutations (false negatives) within
several nucleotides of the ends of reads because such mutations are not,
by definition, part of the best local alignment.</p>
<p>Concerning RNA mutational profiling, false positives are generally much
more problematic than false negatives, so SEISMIC-RNA uses local mode
(<code class="docutils literal notranslate"><span class="pre">--bt2-local</span></code>) by default.
Use end-to-end mode (<code class="docutils literal notranslate"><span class="pre">--bt2-end-to-end</span></code>) only if you have a compelling
reason to do so (e.g. if it is essential to detect mutations at the ends
of reads) and only after carefully trimming any extraneous sequences
from the ends of the reads.</p>
</section>
<section id="how-to-align-paired-end-reads">
<h5>How to align paired-end reads<a class="headerlink" href="#how-to-align-paired-end-reads" title="Permalink to this headline"></a></h5>
<p>Bowtie 2 considers paired-end reads to align “concordantly” when their
relative positions match expectations and “discordantly” otherwise. See
the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#concordant-pairs-match-pair-expectations-discordant-pairs-dont">Bowtie 2 manual for details on concordant/discordant alignments</a>.
By default, SEISMIC-RNA treats only concordantly aligning pairs as valid
alignments. To also treat discordant pairs as valid alignments, use the
flag <code class="docutils literal notranslate"><span class="pre">--bt2-discordant</span></code>.</p>
<p>Several options control which types of alignments are concordant. First,
the expected orientation of paired mates is set using <code class="docutils literal notranslate"><span class="pre">--bt2-orient</span></code>.
It can be <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the 5’-most mate is forward, the 3’-most is reversed),
<code class="docutils literal notranslate"><span class="pre">rf</span></code> (the 5’-most mate is reversed, the 3’-most is forward), or <code class="docutils literal notranslate"><span class="pre">ff</span></code>
(both mates are forward). The default is <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the most common type).
Second, the mates may <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mates-can-overlap-contain-or-dovetail-each-other">overlap partially or completely, or dovetail</a>.
By default, overlaps (partial and complete) are considered concordant,
and dovetailing is considered discordant. The flag <code class="docutils literal notranslate"><span class="pre">--bt2-no-contain</span></code>
treats as discordant pairs where one mate completely overlaps the other,
while <code class="docutils literal notranslate"><span class="pre">--bt2-dovetail</span></code> treats dovetailed pairs as concordant. Pairs
that overlap partially are always considered concordant in SEISMIC-RNA.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The flags <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]contain</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]dovetail</span></code> choose
whether to treat these types of overlaps as concordant (yes) or
discordant (no). If they are treated as discordant, then the flag
<code class="docutils literal notranslate"><span class="pre">--bt2-[no-]discordant</span></code> determines whether they are considered
valid alignments (yes) or invalid (no).</p>
</div>
<p>The option <code class="docutils literal notranslate"><span class="pre">--bt2-mixed</span></code> enables <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a> wherein, for pairs that
fail to produce a valid paired-end alignment, Bowtie 2 attempts to align
each mate individually (as if it were a single-end read).</p>
</section>
<section id="how-to-filter-aligned-reads">
<h5>How to filter aligned reads<a class="headerlink" href="#how-to-filter-aligned-reads" title="Permalink to this headline"></a></h5>
<p>Alignments can be filtered by <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">alignment score</a> and <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">mapping quality</a>,
which are distinct properties.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">Alignment score</a> measures how well a read aligns to a given location
in the reference.
It is calculated from the number of matches, substitutions, and gaps
using the score parameters.
The minimum alignment scores for local and end-to-end modes can be set
using <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-loc</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-e2e</span></code>, respectively.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#valid-alignments-meet-or-exceed-the-minimum-score-threshold">section of the Bowtie 2 manual on alignment scores</a> for advice
on setting this parameter.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">Mapping quality</a> measures how unique an alignment is: high quality if
the read aligns with a high score to exactly one location, low quality
if it aligns with similar scores to multiple locations in the reference.
The default minimum quality is 25, which corresponds to a confidence of
1 - 10<sup>-2.5</sup> = 0.997 that the read has aligned correctly.
To change the quality threshold, use the option <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>.
For those searching for this option in Bowtie 2, you will not find it.
Instead, reads with insufficient mapping quality are filtered out after
alignment using the <a class="reference external" href="https://www.htslib.org/doc/samtools-view.html">view command in Samtools</a>.</p>
</section>
<section id="how-to-filter-by-number-of-aligned-reads">
<h5>How to filter by number of aligned reads<a class="headerlink" href="#how-to-filter-by-number-of-aligned-reads" title="Permalink to this headline"></a></h5>
<p>In general, alignment maps containing very few reads are not useful for
mutational profiling, due to their inherently low coverage per position.
Worse, if aligning to a very large number of references (e.g. an entire
transcriptome), most of the references would likely receive insufficient
reads, so most of the (many) output XAM files would be useless clutter.</p>
<p>To remedy this inconvenience, after alignment has finished, XAM files
with fewer than a minimum number of reads are automatically deleted.
The default is 1000, which can be set using the option <code class="docutils literal notranslate"><span class="pre">--min-reads</span></code>.
Setting <code class="docutils literal notranslate"><span class="pre">--min-reads</span></code> to 0 disables automatically deleting XAM files.</p>
</section>
<section id="how-to-further-customize-alignment">
<h5>How to further customize alignment<a class="headerlink" href="#how-to-further-customize-alignment" title="Permalink to this headline"></a></h5>
<p>See <a class="reference internal" href="../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can
use with Bowtie 2, and the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">Bowtie 2 manual</a> for details on each of
these options.
These options should suffice for most users.
If you require a more customized alignment workflow, then you can align
your FASTQ files outside of SEISMIC-RNA, then pass the resulting XAM
files into SEISMIC-RNA at the step <a class="reference internal" href="#wf-relate"><span class="std std-ref">Relate: Compute relationships between references and aligned reads</span></a>.</p>
</section>
</section>
<section id="align-setting-format-of-alignment-maps">
<span id="bam-vs-cram"></span><h4>Align setting: Format of alignment maps<a class="headerlink" href="#align-setting-format-of-alignment-maps" title="Permalink to this headline"></a></h4>
<p>SEISMIC-RNA can output alignment map files in either BAM or CRAM format.
For details on these file formats, see <a class="reference internal" href="../formats/data/xam.html"><span class="doc">SAM, BAM, and CRAM: Alignment Maps</span></a>.
The default is CRAM format (option <code class="docutils literal notranslate"><span class="pre">--cram</span></code>); BAM format is enabled
using the option <code class="docutils literal notranslate"><span class="pre">--bam</span></code>.</p>
<p>Alignment maps in CRAM format are smaller than their BAM counterparts,
and hence better suited to long-term storage.
However, the better compression of CRAM files comes at three costs:</p>
<ul class="simple">
<li><p>A CRAM file must be accompanied by a FASTA file storing the sequence
of every reference that appears in the header of the CRAM file.
A CRAM file stores only the relative path to its FASTA file, not the
sequence information, which enables the CRAM file to be much smaller
than it would be if it did need to contain its own sequences.
Because the FASTA file existed before and during the alignment, having
this FASTA file accompany the CRAM file usually incurs no extra cost.
However, moving or deleting the FASTA will break the CRAM file.
As a safeguard against this fragility, SEISMIC-RNA keeps a copy of the
original FASTA file in the same directory as the output CRAM file.
Creating an actual copy would require more storage space and defeat
the purpose of CRAM’s smaller file size, so SEISMIC-RNA actually makes
a <a class="reference external" href="https://en.wikipedia.org/wiki/Hard_link">hard link</a> – not a copy – which requires minimal extra space.
In some circumstances, making a hard link can fail, in which case
SEISMIC-RNA will resort to copying the FASTA file instead.</p></li>
<li><p>Reading and writing CRAM files is slower than for BAM files due to the
extra effort needed for compressing and decompressing CRAM files.</p></li>
<li><p>In the <a class="reference external" href="https://samtools.github.io/hts-specs/">CIGAR strings</a>, distinction between reference matches (<code class="docutils literal notranslate"><span class="pre">=</span></code>)
and substitutions (<code class="docutils literal notranslate"><span class="pre">X</span></code>) is lost upon compressing to CRAM format.
Thus, <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">relate</span></code> must perform extra work to determine if each
non-gapped position is a match or substitution, which makes it run
more slowly than it would if the distinction had been preserved.</p></li>
</ul>
<p>In general, use CRAM format if minimizing the size of your alignment
map files is a priority, especially for long-term storage.
Use BAM format to make the <code class="docutils literal notranslate"><span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">relate</span></code> steps run faster,
and to increase the robustness of the output files (because BAM files
are self-contained, while CRAM files will break without the FASTA file
that accompanies them).</p>
</section>
</section>
<section id="align-output-files">
<h3>Align: Output files<a class="headerlink" href="#align-output-files" title="Permalink to this headline"></a></h3>
<p>All output files, except FastQC reports, are written into the directory
<code class="docutils literal notranslate"><span class="pre">{out}/{sample}/align</span></code>, where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory and
<code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the name of the sample.</p>
<section id="align-output-file-fastqc-reports">
<h4>Align output file: FastQC reports<a class="headerlink" href="#align-output-file-fastqc-reports" title="Permalink to this headline"></a></h4>
<p>If FastQC is run, then it outputs files to <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc</span></code>, where
<code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory (<code class="docutils literal notranslate"><span class="pre">--out-dir</span></code>) and <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is
the name of the sample.
The directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc/init</span></code> is always created and contains
FastQC reports of the initial FASTQ files.
If adapter/quality trimming was run, <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/qc/trim</span></code> is also
created for FastQC reports of the trimmed FASTQ files.</p>
<p>In each directory (<code class="docutils literal notranslate"><span class="pre">init</span></code> and <code class="docutils literal notranslate"><span class="pre">trim</span></code>), FastQC writes two files for
each FASTQ file: <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.html</span></code> and <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.zip</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{fq_name}</span></code> is the name of the original FASTQ file up to the
file extension.
If the option <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code> is given, then FastQC will also unzip
<code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc.zip</span></code> to the directory <code class="docutils literal notranslate"><span class="pre">{fq_name}_fastqc</span></code>.
For details on these outputs, see the documentation for <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>.</p>
</section>
<section id="align-output-file-alignment-maps">
<h4>Align output file: Alignment maps<a class="headerlink" href="#align-output-file-alignment-maps" title="Permalink to this headline"></a></h4>
<p>The most important outputs of <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> are alignment map files.
Alignment maps store the location in the reference sequence to which
each read aligned, as well as the Phred quality scores, mapping quality,
and mutated positions.
SEISMIC-RNA outputs alignment maps where every read aligns to the same
reference (although this is not a restriction outside of SEISMIC-RNA).
Each alignment map is written to <code class="docutils literal notranslate"><span class="pre">{ref}.{xam}</span></code>, where <code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is the
name of the reference to which the reads aligned, and <code class="docutils literal notranslate"><span class="pre">{xam}</span></code> is the
file extension (depending on the selected format).
SEISMIC-RNA can output alignment maps in either BAM or CRAM format.
For a comparison of these formats, see <a class="reference internal" href="#bam-vs-cram"><span class="std std-ref">Align setting: Format of alignment maps</span></a>.</p>
</section>
<section id="align-output-file-reference-sequences">
<h4>Align output file: Reference sequences<a class="headerlink" href="#align-output-file-reference-sequences" title="Permalink to this headline"></a></h4>
<p>If the alignment maps are output in CRAM format, then FASTA file(s) of
the reference sequence(s) are also output alongside the CRAM files.
If the sequencing reads came from a whole sample, then a single FASTA
file, bearing the same name as the input FASTA file, will be output.
The output file will be a <a class="reference external" href="https://en.wikipedia.org/wiki/Hard_link">hard link</a> to the input file, if possible,
to avoid consuming unnecessary storage space.
If the sequencing reads were demultiplexed before alignment, then for
each output CRAM file, a FASTA file with the same name (up to the file
extension) will be written to the same directory.
In both cases, each output FASTA will be indexed using <a class="reference external" href="https://www.htslib.org/doc/samtools-faidx.html">samtools faidx</a>
to speed up reading the CRAM files.
If the alignment maps are output in BAM format, then FASTA files are not
output alongside them.</p>
</section>
<section id="align-output-file-unaligned-reads">
<span id="wf-unaligned"></span><h4>Align output file: Unaligned reads<a class="headerlink" href="#align-output-file-unaligned-reads" title="Permalink to this headline"></a></h4>
<p>In addition to the alignment maps, SEISMIC-RNA outputs FASTQ file(s) of
reads that Bowtie 2 could not align:</p>
<ul class="simple">
<li><p>Each whole-sample FASTQ file of single-end (<code class="docutils literal notranslate"><span class="pre">-z</span></code>) or interleaved
(<code class="docutils literal notranslate"><span class="pre">-y</span></code>) reads yields one file: <code class="docutils literal notranslate"><span class="pre">unaligned.fq.gz</span></code></p></li>
<li><p>Each pair of whole-sample FASTQ files of 1st and 2nd mates (<code class="docutils literal notranslate"><span class="pre">-x</span></code>)
yields two files: <code class="docutils literal notranslate"><span class="pre">unaligned.fq.1.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">unaligned.fq.2.gz</span></code></p></li>
<li><p>Each demultiplexed FASTQ file of single-end (<code class="docutils literal notranslate"><span class="pre">-Z</span></code>) or interleaved
(<code class="docutils literal notranslate"><span class="pre">-Y</span></code>) reads yields one file: <code class="docutils literal notranslate"><span class="pre">{ref}__unaligned.fq.gz</span></code></p></li>
<li><p>Each pair of demultiplexed FASTQ files of 1st and 2nd mates (<code class="docutils literal notranslate"><span class="pre">-X</span></code>)
yields two files:
<code class="docutils literal notranslate"><span class="pre">{ref}__unaligned.fq.1.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">{ref}__unaligned.fq.2.gz</span></code></p></li>
</ul>
<p>where <code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is the reference for demultiplexed FASTQ files.</p>
<p>Outputting these files of unaligned reads can be disabled using the
option <code class="docutils literal notranslate"><span class="pre">--bt2-no-un</span></code>.</p>
</section>
<section id="align-output-file-report">
<h4>Align output file: Report<a class="headerlink" href="#align-output-file-report" title="Permalink to this headline"></a></h4>
<p>A report file, <code class="docutils literal notranslate"><span class="pre">align-report.json</span></code>, is also written that records the
settings used to run alignment and summarizes the results of alignment.
See <a class="reference internal" href="../formats/report/align.html"><span class="doc">Align Report</span></a> for more information.</p>
</section>
</section>
<section id="align-troubleshooting">
<h3>Align: Troubleshooting<a class="headerlink" href="#align-troubleshooting" title="Permalink to this headline"></a></h3>
<section id="troubleshooting-a-lower-than-expected-alignment-rate">
<h4>Troubleshooting a lower-than-expected alignment rate<a class="headerlink" href="#troubleshooting-a-lower-than-expected-alignment-rate" title="Permalink to this headline"></a></h4>
<p>If the percent of reads aligning to the reference is less than expected,
then try the following steps (in this order):</p>
<ol class="arabic simple">
<li><p>Ensure you are using Bowtie version 2.5.1 or later (version 2.5.0
has a bug that affects alignment rate).
You can check the version with <code class="docutils literal notranslate"><span class="pre">bowtie2</span> <span class="pre">--version</span> <span class="pre">|</span> <span class="pre">head</span> <span class="pre">-n</span> <span class="pre">1</span></code>.</p></li>
<li><p>Double check that the FASTA has the correct reference sequence(s)
and that, if the Bowtie 2 index was pre-built before the align step,
that the correct FASTA file was used.</p></li>
<li><p>Examine the reads that failed to align (see <a class="reference internal" href="#wf-unaligned"><span class="std std-ref">Align output file: Unaligned reads</span></a>).
Choose several reads randomly and check if they could come from any
known sources by querying <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&amp;PAGE_TYPE=BlastSearch&amp;LINK_LOC=blasthome">BLAST</a> (or similar tools) for short
(20 - 40 nt) segments of each read.
Identifying the sources of unaligned reads can help determine the
cause of the problem (e.g. contamination with ribosomal or foreign
RNA such as from <em>Mycoplasma</em>, incorrect indexes used during FASTQ
generation) and whether the reads that did align are still usable.</p></li>
</ol>
</section>
</section>
</section>
<section id="relate-compute-relationships-between-references-and-aligned-reads">
<span id="wf-relate"></span><h2>Relate: Compute relationships between references and aligned reads<a class="headerlink" href="#relate-compute-relationships-between-references-and-aligned-reads" title="Permalink to this headline"></a></h2>
<section id="relate-input-files">
<h3>Relate: Input files<a class="headerlink" href="#relate-input-files" title="Permalink to this headline"></a></h3>
<section id="relate-input-file-reference-sequences">
<h4>Relate input file: Reference sequences<a class="headerlink" href="#relate-input-file-reference-sequences" title="Permalink to this headline"></a></h4>
<p>Relate requires exactly one file of reference sequences, which must be
DNA sequences (only A, C, G, T, and N) in FASTA format.
If needed, you may clean the FASTA file with the <a class="reference internal" href="cleanfa.html"><span class="doc">Cleaning FASTA Files</span></a> tool.
See <a class="reference internal" href="../formats/data/fasta.html"><span class="doc">FASTA: Reference sequences</span></a> for details.</p>
</section>
<section id="relate-input-file-alignment-maps">
<h4>Relate input file: Alignment maps<a class="headerlink" href="#relate-input-file-alignment-maps" title="Permalink to this headline"></a></h4>
<p>Relate accepts any number of alignment map files, each of which must be
in SAM, BAM, or CRAM format (collectively, “XAM” format).
See <a class="reference internal" href="../formats/data/xam.html"><span class="doc">SAM, BAM, and CRAM: Alignment Maps</span></a> for details.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The references in the FASTA file must match those to which the reads
in the alignment map were aligned.
Discrepancies can cause the <code class="docutils literal notranslate"><span class="pre">relate</span></code> command to fail or produce
erroneous relation vectors.
This problem will not occur if you use the same (unaltered) FASTA
file for both the <code class="docutils literal notranslate"><span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">relate</span></code> commands, or run both
at once using the command <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>.</p>
</div>
<p>List every alignment map file after the FASTA file.
Refer to <a class="reference internal" href="inputs.html"><span class="doc">Listing Input Files</span></a> for details on how to list multiple files.
For example, to compute relation vectors for reads from <code class="docutils literal notranslate"><span class="pre">sample-1</span></code>
aligned to references <code class="docutils literal notranslate"><span class="pre">ref-1</span></code> and <code class="docutils literal notranslate"><span class="pre">ref-2</span></code>, and from <code class="docutils literal notranslate"><span class="pre">sample-2</span></code>
aligned to reference <code class="docutils literal notranslate"><span class="pre">ref-1</span></code>, use the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">relate</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="n">sample</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mf">1.</span><span class="n">cram</span> <span class="n">sample</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mf">2.</span><span class="n">cram</span> <span class="n">sample</span><span class="o">-</span><span class="mi">2</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mf">1.</span><span class="n">cram</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs.fa}</span></code> is the path to the file of reference sequences.</p>
</section>
</section>
<section id="relate-settings">
<h3>Relate: Settings<a class="headerlink" href="#relate-settings" title="Permalink to this headline"></a></h3>
<section id="relate-options-shared-with-alignment">
<h4>Relate options shared with alignment<a class="headerlink" href="#relate-options-shared-with-alignment" title="Permalink to this headline"></a></h4>
<p>Because this workflow can be started from the <code class="docutils literal notranslate"><span class="pre">align</span></code> or <code class="docutils literal notranslate"><span class="pre">relate</span></code>
commands, the latter duplicates some of the options of the former:
<code class="docutils literal notranslate"><span class="pre">--phred-enc</span></code>, <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>, <code class="docutils literal notranslate"><span class="pre">--min-reads</span></code>, and <code class="docutils literal notranslate"><span class="pre">--out-dir</span></code> have
the same functions in <code class="docutils literal notranslate"><span class="pre">relate</span></code> and <code class="docutils literal notranslate"><span class="pre">align</span></code>.</p>
</section>
<section id="relate-setting-minimum-phred-score">
<h4>Relate setting: Minimum Phred score<a class="headerlink" href="#relate-setting-minimum-phred-score" title="Permalink to this headline"></a></h4>
<p>Base calls with Phred scores below <code class="docutils literal notranslate"><span class="pre">--min-phred</span></code> are labeled ambiguous
matches or substitutions, as if they were <code class="docutils literal notranslate"><span class="pre">N</span></code>s.
For example, if the minimum Phred score is 25 (the default) and a base
<code class="docutils literal notranslate"><span class="pre">T</span></code> is called as a match with a Phred score of 20, then it would be
marked as possibly a match and possibly a subsitution to A, C, or G.
See <a class="reference internal" href="../data/relate/codes.html#relate-low-qual"><span class="std std-ref">Encoding low-quality base calls</span></a> for more information.</p>
</section>
<section id="relate-setting-ambiguous-insertions-and-deletions">
<h4>Relate setting: Ambiguous insertions and deletions<a class="headerlink" href="#relate-setting-ambiguous-insertions-and-deletions" title="Permalink to this headline"></a></h4>
<p>The most tricky problem in computing relation vectors is that insertions
and deletions (“indels”) in repetitive regions cause ambiguities.
SEISMIC-RNA introduces a new algorithm for identifying ambiguous indels
(see <span class="xref std std-doc">../algos/ambrel</span> for more information).
This algorithm is enabled by default.
If it is not necessary to identify ambiguous indels, then the algorithm
can be disabled with <code class="docutils literal notranslate"><span class="pre">--no-ambrel</span></code>, which will speed up <code class="docutils literal notranslate"><span class="pre">relate</span></code> at
the cost of reducing its accuracy on indels.</p>
</section>
<section id="relate-setting-batch-size">
<h4>Relate setting: Batch size<a class="headerlink" href="#relate-setting-batch-size" title="Permalink to this headline"></a></h4>
<p>For an explanation of batching and how to use it, see <a class="reference internal" href="parallel.html#batches"><span class="std std-ref">Batches and Benefits</span></a>.</p>
<p>The dataset is partitioned into batches by the <code class="docutils literal notranslate"><span class="pre">relate</span></code> command.
The option <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code> sets a target amount of data for each batch,
in millions of base calls (megabases).
This calculation considers the total number of relationships per read,
which equals the length of the reference sequence.
Thus, the number of base calls <em>B</em> is the product of the number of reads
<em>N</em> and the length of the reference sequence <em>L</em>:</p>
<p><em>B</em> = <em>NL</em></p>
<p>Since <em>L</em> is known and <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code> specifies a target size for <em>B</em>,
<em>N</em> can be solved for:</p>
<p><em>N</em> = <em>B</em>/<em>L</em></p>
<p>SEISMIC-RNA will aim to put exactly <em>N</em> reads in each batch but the last
(the last batch can be smaller because it has just the leftover reads).
If the reads are single-ended or were not aligned in <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a>, then
every batch but the last will contain exactly <em>N</em> reads.
If the reads are paired-ended and were aligned in <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a>, then
batches may contain more than <em>N</em> reads, up to a maximum of 2<em>N</em> in
the extreme case that only one read aligned in every mate pair.</p>
</section>
</section>
<section id="relate-output-files">
<h3>Relate: Output files<a class="headerlink" href="#relate-output-files" title="Permalink to this headline"></a></h3>
<p>All output files go into the directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/relate/{ref}</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory, <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the sample, and
<code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is the name of the reference.</p>
<section id="relate-output-file-batch-of-relation-vectors">
<h4>Relate output file: Batch of relation vectors<a class="headerlink" href="#relate-output-file-batch-of-relation-vectors" title="Permalink to this headline"></a></h4>
<p>The data of relationships is written in batches.
Each batch contains a <code class="docutils literal notranslate"><span class="pre">RelateBatchIO</span></code> object and is saved to the file
<code class="docutils literal notranslate"><span class="pre">relate-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <a class="reference internal" href="../data/relate/relate.html"><span class="doc">Relate Batch</span></a> for details on the data structure.
See <a class="reference internal" href="../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for details on brickle files.</p>
</section>
<section id="relate-output-file-batch-of-read-names">
<h4>Relate output file: Batch of read names<a class="headerlink" href="#relate-output-file-batch-of-read-names" title="Permalink to this headline"></a></h4>
<p>For each batch, the relate step assigns an index (a nonnegative integer)
to each read and writes a file mapping the indexes to the read names.
Each batch contains a <code class="docutils literal notranslate"><span class="pre">QnamesBatchIO</span></code> object and is saved to the file
<code class="docutils literal notranslate"><span class="pre">qnames-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <a class="reference internal" href="../data/relate/qnames.html"><span class="doc">Read Names Batch</span></a> for details on the data structure.
See <a class="reference internal" href="../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for details on brickle files.</p>
</section>
<section id="relate-output-file-reference-sequence">
<h4>Relate output file: Reference sequence<a class="headerlink" href="#relate-output-file-reference-sequence" title="Permalink to this headline"></a></h4>
<p>The relate step writes the reference sequence as a <code class="docutils literal notranslate"><span class="pre">RefseqIO</span></code> object
to the file <code class="docutils literal notranslate"><span class="pre">refseq.brickle</span></code>.
See <a class="reference internal" href="../data/relate/refseq.html"><span class="doc">Reference Sequence</span></a> for details on the data structure.
See <a class="reference internal" href="../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for details on brickle files.</p>
</section>
<section id="relate-output-file-report">
<h4>Relate output file: Report<a class="headerlink" href="#relate-output-file-report" title="Permalink to this headline"></a></h4>
<p>A report file is written that records the settings used to generate
relation vectors summarizes the results.
See <a class="reference internal" href="../formats/report/relate.html"><span class="doc">Relate Report</span></a> for more information.</p>
</section>
</section>
</section>
<section id="mask-define-mutations-and-sections-to-filter-reads-and-positions">
<span id="wf-mask"></span><h2>Mask: Define mutations and sections to filter reads and positions<a class="headerlink" href="#mask-define-mutations-and-sections-to-filter-reads-and-positions" title="Permalink to this headline"></a></h2>
<section id="mask-input-files">
<h3>Mask: Input files<a class="headerlink" href="#mask-input-files" title="Permalink to this headline"></a></h3>
<section id="mask-input-file-relate-report">
<h4>Mask input file: Relate report<a class="headerlink" href="#mask-input-file-relate-report" title="Permalink to this headline"></a></h4>
<p>Mask accepts any number of report files from the Relate step.
As with all input files, each report file can be given individually, or
directories containing any number of report files (alongside other types
of files) can be given, or both.
Refer to <a class="reference internal" href="inputs.html"><span class="doc">Listing Input Files</span></a> for details on how to list multiple files.</p>
<p>For example, the following command will run the Mask step (with default
settings) on every set of reads from the Relate step:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">mask</span> <span class="p">{</span><span class="n">out</span><span class="p">}</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the path to the output directory.</p>
</section>
</section>
<section id="mask-settings">
<h3>Mask: Settings<a class="headerlink" href="#mask-settings" title="Permalink to this headline"></a></h3>
<section id="mask-defining-sections">
<h4>Mask: Defining sections<a class="headerlink" href="#mask-defining-sections" title="Permalink to this headline"></a></h4>
<p>Using the Mask step, sections of reference sequences can be selected for
analysis, ignoring other parts of the sequences.
This feature is useful for analyzing small elements of longer sequences,
such as an <a class="reference external" href="https://en.wikipedia.org/wiki/Internal_ribosome_entry_site">IRES</a> of several hundred nucleotides within a viral genome
of several thousand.</p>
<section id="mask-defining-sections-by-coordinates-or-primer-sequences">
<h5>Mask: Defining sections by coordinates or primer sequences<a class="headerlink" href="#mask-defining-sections-by-coordinates-or-primer-sequences" title="Permalink to this headline"></a></h5>
<p>Sections can be specified by their coordinates or, if the sample was
prepared as an amplicon (using <a class="reference external" href="https://en.wikipedia.org/wiki/Reverse_transcription_polymerase_chain_reaction">RT-PCR</a> with a pair of primers flanking
a specific sequence), by their forward and reverse primer sequences.</p>
<p>If using coordinates, then note that the first position in the reference
sequence is always numbered 1, and that the section will contain both
the first and last coordinates.
For example, if a reference sequence is 10 bases, then a section defined
by the first and last coordinates of 4 and 9, respectively, will include
bases 4, 5, 6, 7, 8, and 9; but not bases 1, 2, 3, or 10.
The first coordinate must be a positive integer, and the last coordinate
a non-negative integer greater than or equal to the first coordinate
minus one, and no greater than the length of the reference sequence.
If the first and last coordinates are equal, then the section will have
exactly one position: the position of the equal coordinates.
If the last coordinate is one less than the first, then the section will
have zero positions; this behavior is permitted in order to handle the
edge case of a length-zero section smoothly, but is probably of no use.</p>
<p>If using primers, then the entire sequence of the forward primer and the
reverse complement of the reverse primer must both occur exactly once in
the reference sequence (no mismatches or gaps are permitted).
Primers will cover up any mutations in the cDNA during PCR, so the sites
bound by the primers provide no information for mutational profiling.
Thus, the first coordinate of the section is defined as the position one
after the last position in the forward primer binding site, and the last
coordinate defined as the position one before the first position in the
reverse primer binding site (the reverse complement of the primer).
For example:</p>
<ul class="simple">
<li><p>reference sequence: <code class="docutils literal notranslate"><span class="pre">TTTCGCTATGTGTTAC</span></code></p></li>
<li><p>forward primer: <code class="docutils literal notranslate"><span class="pre">TCG</span></code></p></li>
<li><p>reverse primer: <code class="docutils literal notranslate"><span class="pre">AAC</span></code></p></li>
</ul>
<p>These forward and reverse primers bind coordinates 3-5 and 12-14 of the
reference sequence, respectively; thus, they define the section 6-11,
with the sequence <code class="docutils literal notranslate"><span class="pre">CTATGT</span></code>.</p>
<p>Note that artifacts may occur near the ends of the primers, so ignoring
one or two positions nearest the primer may help reduce artifacts.
The option <code class="docutils literal notranslate"><span class="pre">--primer-gap</span></code> controls the number of positions between the
ends of the primers and the ends of the section.
In the example above, using <code class="docutils literal notranslate"><span class="pre">--primer-gap</span> <span class="pre">1</span></code> would remove one position
from each side of the section, now coordinates 7-10 (<code class="docutils literal notranslate"><span class="pre">TATG</span></code>).</p>
</section>
<section id="mask-defining-sections-using-command-line-options">
<h5>Mask: Defining sections using command line options<a class="headerlink" href="#mask-defining-sections-using-command-line-options" title="Permalink to this headline"></a></h5>
<p>Coordinates and primers can be given as options on the command line.</p>
</section>
</section>
<section id="id1">
<h4>Relate setting: Minimum Phred score<a class="headerlink" href="#id1" title="Permalink to this headline"></a></h4>
<p>Base calls with Phred scores below <code class="docutils literal notranslate"><span class="pre">--min-phred</span></code> are labeled ambiguous
matches or substitutions, as if they were <code class="docutils literal notranslate"><span class="pre">N</span></code>s.
For example, if the minimum Phred score is 25 (the default) and a base
<code class="docutils literal notranslate"><span class="pre">T</span></code> is called as a match with a Phred score of 20, then it would be
marked as possibly a match and possibly a subsitution to A, C, or G.
See <a class="reference internal" href="../data/relate/codes.html#relate-low-qual"><span class="std std-ref">Encoding low-quality base calls</span></a> for more information.</p>
</section>
<section id="id2">
<h4>Relate setting: Ambiguous insertions and deletions<a class="headerlink" href="#id2" title="Permalink to this headline"></a></h4>
<p>The most tricky problem in computing relation vectors is that insertions
and deletions (“indels”) in repetitive regions cause ambiguities.
SEISMIC-RNA introduces a new algorithm for identifying ambiguous indels
(see <span class="xref std std-doc">../algos/ambrel</span> for more information).
This algorithm is enabled by default.
If it is not necessary to identify ambiguous indels, then the algorithm
can be disabled with <code class="docutils literal notranslate"><span class="pre">--no-ambrel</span></code>, which will speed up <code class="docutils literal notranslate"><span class="pre">relate</span></code> at
the cost of reducing its accuracy on indels.</p>
</section>
<section id="id3">
<h4>Relate setting: Batch size<a class="headerlink" href="#id3" title="Permalink to this headline"></a></h4>
<p>For an explanation of batching and how to use it, see <a class="reference internal" href="parallel.html#batches"><span class="std std-ref">Batches and Benefits</span></a>.</p>
<p>The dataset is partitioned into batches by the <code class="docutils literal notranslate"><span class="pre">relate</span></code> command.
The option <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code> sets a target amount of data for each batch,
in millions of base calls (megabases).
This calculation considers the total number of relationships per read,
which equals the length of the reference sequence.
Thus, the number of base calls <em>B</em> is the product of the number of reads
<em>N</em> and the length of the reference sequence <em>L</em>:</p>
<p><em>B</em> = <em>NL</em></p>
<p>Since <em>L</em> is known and <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code> specifies a target size for <em>B</em>,
<em>N</em> can be solved for:</p>
<p><em>N</em> = <em>B</em>/<em>L</em></p>
<p>SEISMIC-RNA will aim to put exactly <em>N</em> reads in each batch but the last
(the last batch can be smaller because it has just the leftover reads).
If the reads are single-ended or were not aligned in <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a>, then
every batch but the last will contain exactly <em>N</em> reads.
If the reads are paired-ended and were aligned in <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a>, then
batches may contain more than <em>N</em> reads, up to a maximum of 2<em>N</em> in
the extreme case that only one read aligned in every mate pair.</p>
</section>
</section>
<section id="id4">
<h3>Relate: Output files<a class="headerlink" href="#id4" title="Permalink to this headline"></a></h3>
<p>All output files go into the directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/relate/{ref}</span></code>,
where <code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory, <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the sample, and
<code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is the name of the reference.</p>
<section id="id5">
<h4>Relate output file: Batch of relation vectors<a class="headerlink" href="#id5" title="Permalink to this headline"></a></h4>
<p>The data of relationships is written in batches.
Each batch contains a <code class="docutils literal notranslate"><span class="pre">RelateBatchIO</span></code> object and is saved to the file
<code class="docutils literal notranslate"><span class="pre">relate-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <a class="reference internal" href="../data/relate/relate.html"><span class="doc">Relate Batch</span></a> for details on the data structure.
See <a class="reference internal" href="../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for details on brickle files.</p>
</section>
<section id="id6">
<h4>Relate output file: Batch of read names<a class="headerlink" href="#id6" title="Permalink to this headline"></a></h4>
<p>For each batch, the relate step assigns an index (a nonnegative integer)
to each read and writes a file mapping the indexes to the read names.
Each batch contains a <code class="docutils literal notranslate"><span class="pre">QnamesBatchIO</span></code> object and is saved to the file
<code class="docutils literal notranslate"><span class="pre">qnames-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <a class="reference internal" href="../data/relate/qnames.html"><span class="doc">Read Names Batch</span></a> for details on the data structure.
See <a class="reference internal" href="../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for details on brickle files.</p>
</section>
<section id="id7">
<h4>Relate output file: Reference sequence<a class="headerlink" href="#id7" title="Permalink to this headline"></a></h4>
<p>The relate step writes the reference sequence as a <code class="docutils literal notranslate"><span class="pre">RefseqIO</span></code> object
to the file <code class="docutils literal notranslate"><span class="pre">refseq.brickle</span></code>.
See <a class="reference internal" href="../data/relate/refseq.html"><span class="doc">Reference Sequence</span></a> for details on the data structure.
See <a class="reference internal" href="../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for details on brickle files.</p>
</section>
<section id="id8">
<h4>Relate output file: Report<a class="headerlink" href="#id8" title="Permalink to this headline"></a></h4>
<p>A report file is written that records the settings used to generate
relation vectors summarizes the results.
See <a class="reference internal" href="../formats/report/relate.html"><span class="doc">Relate Report</span></a> for more information.</p>
</section>
</section>
</section>
<section id="run-the-entire-workflow-from-demult-align-to-graph-export">
<span id="wf-wf"></span><h2>Run the entire workflow, from demult/align to graph/export.<a class="headerlink" href="#run-the-entire-workflow-from-demult-align-to-graph-export" title="Permalink to this headline"></a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code> accepts FASTQ, SAM/BAM/CRAM, relate/mask/cluster report, and
table files and directories as inputs.</p>
</div>
<p>From BAM, report, and/or table file(s):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">wf</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="n">out</span><span class="o">/</span><span class="n">sample</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">Ref</span><span class="o">.</span><span class="n">bam</span> <span class="n">out</span><span class="o">/</span><span class="n">sample</span><span class="o">/*/*-</span><span class="n">report</span><span class="o">.</span><span class="n">json</span> <span class="n">out</span><span class="o">/</span><span class="n">sample</span><span class="o">/</span><span class="n">table</span><span class="o">/*/*.</span><span class="n">csv</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only the align, relate, mask, and table steps run by default.
Enable clustering by specifying <code class="docutils literal notranslate"><span class="pre">--max-clusters</span></code> (<code class="docutils literal notranslate"><span class="pre">-k</span></code>) followed by the
maximum number of clusters to attempt. Enable structure prediction
with the flag <code class="docutils literal notranslate"><span class="pre">--fold</span></code>.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Manuals" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="cleanfa.html" class="btn btn-neutral float-right" title="Cleaning FASTA Files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>