<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>FASTQ: Sequencing Reads &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="SAM, BAM, and CRAM: Alignment Maps" href="xam.html" />
    <link rel="prev" title="FASTA: Reference sequences" href="fasta.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../manuals/index.html">Manuals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">File Formats</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../report/index.html">Report Formats</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Data Formats</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="fasta.html">FASTA: Reference sequences</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">FASTQ: Sequencing Reads</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fastq-file-content-format">FASTQ file: Content format</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fastq-file-path-format">FASTQ file: Path format</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fastq-file-uses">FASTQ file: Uses</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="xam.html">SAM, BAM, and CRAM: Alignment Maps</a></li>
<li class="toctree-l3"><a class="reference internal" href="brickle.html">Brickle: Compressed Python Objects</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../meta/index.html">Metadata Formats</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">File Formats</a></li>
          <li class="breadcrumb-item"><a href="index.html">Data Formats</a></li>
      <li class="breadcrumb-item active">FASTQ: Sequencing Reads</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/formats/data/fastq.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="fastq-sequencing-reads">
<h1>FASTQ: Sequencing Reads<a class="headerlink" href="#fastq-sequencing-reads" title="Permalink to this headline"></a></h1>
<section id="fastq-file-content-format">
<h2>FASTQ file: Content format<a class="headerlink" href="#fastq-file-content-format" title="Permalink to this headline"></a></h2>
<p>In a FASTQ file, each sequencing read record comprises four lines:</p>
<ol class="arabic simple">
<li><p>A header that begins with <code class="docutils literal notranslate"><span class="pre">&#64;</span></code> and contains the name of the read.</p></li>
<li><p>The sequence of the read.</p></li>
<li><p>A second header that begins with <code class="docutils literal notranslate"><span class="pre">+</span></code> and may repeat the read name.</p></li>
<li><p>The quality string of the same length as the read, indicating the
quality of each base in the read using <a class="reference internal" href="#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a>.</p></li>
</ol>
<p>See <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ format</a> for more information.</p>
<section id="fastq-character-sets">
<h3>FASTQ character sets<a class="headerlink" href="#fastq-character-sets" title="Permalink to this headline"></a></h3>
<section id="fastq-dna-alphabet">
<h4>FASTQ DNA alphabet<a class="headerlink" href="#fastq-dna-alphabet" title="Permalink to this headline"></a></h4>
<p>Read sequences may contain only <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">C</span></code>, <code class="docutils literal notranslate"><span class="pre">G</span></code>, <code class="docutils literal notranslate"><span class="pre">T</span></code>, and <code class="docutils literal notranslate"><span class="pre">N</span></code>.</p>
</section>
<section id="phred-quality-score-encodings">
<span id="phred-encodings"></span><h4>Phred quality score encodings<a class="headerlink" href="#phred-quality-score-encodings" title="Permalink to this headline"></a></h4>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Phred_quality_score">Phred quality scores</a> represent the confidence that a base in a FASTQ
file was called correctly during sequencing.
The probability <em>p</em> that a base was called incorrectly is 10 raised to
the power of the quotient of the Phred score <em>s</em> and -10:</p>
<p><em>p</em> = 10<sup>-s/10</sup></p>
<p>For example, if a base call has a Phred score of 30, the probability
that the base call is incorrect is 10<sup>-30/10</sup> = 0.001.</p>
<p>In FASTQ files, each phred quality score (a non-negative integer) is
encoded as one character of text by adding another integer <em>N</em> to the
Phred score (<a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format#Encoding">Phred+N</a>) and then converting the number to the character
with the corresponding <a class="reference external" href="https://en.wikipedia.org/wiki/ASCII">ASCII code</a>.
For example, if <em>N</em> is 33, then the Phred score 25 would be encoded by
adding 33 to 25 (obtaining 58), then writing the character whose ASCII
code is 58 (which is <code class="docutils literal notranslate"><span class="pre">:</span></code>).</p>
</section>
</section>
<section id="endedness-single-end-and-paired-end-reads">
<span id="fastq-endedness"></span><h3>Endedness: single-end and paired-end reads<a class="headerlink" href="#endedness-single-end-and-paired-end-reads" title="Permalink to this headline"></a></h3>
<p>Illumina sequencers can be run in two main modes: single-end mode, in
which each molecule is read starting from one end; and paired-end mode,
in which each molecule is read from both ends.
For more details on endedness, see this <a class="reference external" href="https://www.illumina.com/science/technology/next-generation-sequencing/plan-experiments/paired-end-vs-single-read.html">guide from Illumina</a>.</p>
<p>Each paired-end read comprises a pair of so-called “mates”: mate 1 and
mate 2.
Each mate is similar to a single-end read, the difference being that it
is paired with and shares its name with its mate.
There are two options for paired-end reads:</p>
<ul class="simple">
<li><p>A single “interleaved” file in which mates 1 and 2 alternate, each
mate 2 coming directly after the mate 1 to which it is paired.</p></li>
<li><p>Two separate files containing all mate 1 reads and all mate 2 reads,
respectively, in the same order in both files.</p></li>
</ul>
<p>In SEISMIC-RNA (as well as many other pieces of software), each FASTQ
file must contain only one type of reads:</p>
<ul class="simple">
<li><p>single-end reads</p></li>
<li><p>paired-end reads, with interleaved 1st and 2nd mates</p></li>
<li><p>paired-end reads, with 1st mates only</p></li>
<li><p>paired-end reads, with 2nd mates only</p></li>
</ul>
<p>Thus, we refer to an entire FASTQ file as “single-end” if it contains
only single-end reads, “interleaved paired-end” if it contains only
interleaved 1st and 2nd mates, and “separate paired-end” if it contains
either 1st or 2nd mates of paired-end reads.</p>
<p>The following diagrams illustrate single-end, separate paired-end, and
interleaved paired-end FASTQ files.</p>
<section id="fastq-file-with-single-end-reads">
<h4>FASTQ file with single-end reads<a class="headerlink" href="#fastq-file-with-single-end-reads" title="Permalink to this headline"></a></h4>
<p>1 file, 4 lines per read</p>
<p><code class="docutils literal notranslate"><span class="pre">name.fq</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@Read_1_ID</span>
<span class="n">GCATGCTAGCCA</span>
<span class="o">+</span>
<span class="n">FFFFFFFFF</span><span class="p">:</span><span class="n">F</span><span class="p">:</span>
<span class="nd">@Read_2_ID</span>
<span class="n">ATCGTCATGTGT</span>
<span class="o">+</span>
<span class="n">FFFFFFF</span><span class="p">:</span><span class="n">FFFF</span>
</pre></div>
</div>
</section>
<section id="fastq-file-with-interleaved-paired-end-reads">
<h4>FASTQ file with interleaved paired-end reads<a class="headerlink" href="#fastq-file-with-interleaved-paired-end-reads" title="Permalink to this headline"></a></h4>
<p>1 file, 8 lines per paired read (4 for mate 1, then 4 for mate 2)</p>
<p><code class="docutils literal notranslate"><span class="pre">name.fq</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@Read_1_ID</span><span class="o">/</span><span class="mi">1</span>
<span class="n">GCATGCTAGCCA</span>
<span class="o">+</span>
<span class="n">FFFFFFFFF</span><span class="p">:</span><span class="n">F</span><span class="p">:</span>
<span class="nd">@Read_1_ID</span><span class="o">/</span><span class="mi">2</span>
<span class="n">TACGTCGTCGTC</span>
<span class="o">+</span>
<span class="n">FFFFF</span><span class="p">:</span><span class="n">FF</span><span class="p">:</span><span class="n">F</span><span class="p">::</span>
<span class="nd">@Read_2_ID</span><span class="o">/</span><span class="mi">1</span>
<span class="n">ATCGTCATGTGT</span>
<span class="o">+</span>
<span class="n">FFFFFFF</span><span class="p">:</span><span class="n">FFFF</span>
<span class="nd">@Read_2_ID</span><span class="o">/</span><span class="mi">2</span>
<span class="n">CACGAGCGATAG</span>
<span class="o">+</span>
<span class="n">FFFF</span><span class="p">:</span><span class="n">FF</span><span class="p">:::</span><span class="n">F</span><span class="p">:</span>
</pre></div>
</div>
</section>
<section id="pair-of-fastq-files-with-1st-and-2nd-mates-in-separate-files">
<h4>Pair of FASTQ files with 1st and 2nd mates in separate files<a class="headerlink" href="#pair-of-fastq-files-with-1st-and-2nd-mates-in-separate-files" title="Permalink to this headline"></a></h4>
<p>2 files, 4 lines per mate in each file</p>
<p><code class="docutils literal notranslate"><span class="pre">name_R1.fq</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@Read_1_ID</span><span class="o">/</span><span class="mi">1</span>
<span class="n">GCATGCTAGCCA</span>
<span class="o">+</span>
<span class="n">FFFFFFFFF</span><span class="p">:</span><span class="n">F</span><span class="p">:</span>
<span class="nd">@Read_2_ID</span><span class="o">/</span><span class="mi">1</span>
<span class="n">ATCGTCATGTGT</span>
<span class="o">+</span>
<span class="n">FFFFFFF</span><span class="p">:</span><span class="n">FFFF</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">name_R2.fq</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@Read_1_ID</span><span class="o">/</span><span class="mi">2</span>
<span class="n">TACGTCGTCGTC</span>
<span class="o">+</span>
<span class="n">FFFFF</span><span class="p">:</span><span class="n">FF</span><span class="p">:</span><span class="n">F</span><span class="p">::</span>
<span class="nd">@Read_2_ID</span><span class="o">/</span><span class="mi">2</span>
<span class="n">CACGAGCGATAG</span>
<span class="o">+</span>
<span class="n">FFFF</span><span class="p">:</span><span class="n">FF</span><span class="p">:::</span><span class="n">F</span><span class="p">:</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="fastq-file-path-format">
<h2>FASTQ file: Path format<a class="headerlink" href="#fastq-file-path-format" title="Permalink to this headline"></a></h2>
<section id="fastq-file-extensions">
<h3>FASTQ file extensions<a class="headerlink" href="#fastq-file-extensions" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA accepts the following extensions for FASTQ files:</p>
<ul class="simple">
<li><p>Compressed (with <a class="reference external" href="https://www.gnu.org/software/gzip/">gzip</a>):</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">.fq.gz</span></code> (default)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">.fastq.gz</span></code></p></li>
</ul>
</li>
<li><p>Uncompressed:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">.fq</span></code> (default)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">.fastq</span></code></p></li>
</ul>
</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>SEISMIC-RNA accepts FASTQ files that are compressed with <a class="reference external" href="https://www.gnu.org/software/gzip/">gzip</a>.
It is recommended to always use compressed FASTQ files because FASTQ
files are typically very large without compression, on the order of
100 Mb to 10 Gb.
The file extension will be preserved through the workflow, i.e. if
an input FASTQ file has the extension <code class="docutils literal notranslate"><span class="pre">.fq.gz</span></code>, then the trimmed
FASTQ file (if any) will also have that extension and be compressed.</p>
</div>
</section>
<section id="fastq-mate-1-and-2-labels">
<h3>FASTQ mate 1 and 2 labels<a class="headerlink" href="#fastq-mate-1-and-2-labels" title="Permalink to this headline"></a></h3>
<p>For paired-end reads whose 1st and 2nd mates are in separate files, the
file names must have one of the following labels before the extension:</p>
<ul class="simple">
<li><p>Mate 1: <code class="docutils literal notranslate"><span class="pre">_R1</span></code>, <code class="docutils literal notranslate"><span class="pre">_mate1</span></code>, <code class="docutils literal notranslate"><span class="pre">_1_sequence</span></code>, <code class="docutils literal notranslate"><span class="pre">_R1_001</span></code>, <code class="docutils literal notranslate"><span class="pre">_mate1_001</span></code>, <code class="docutils literal notranslate"><span class="pre">_1_sequence_001</span></code></p></li>
<li><p>Mate 2: <code class="docutils literal notranslate"><span class="pre">_R2</span></code>, <code class="docutils literal notranslate"><span class="pre">_mate2</span></code>, <code class="docutils literal notranslate"><span class="pre">_2_sequence</span></code>, <code class="docutils literal notranslate"><span class="pre">_R2_001</span></code>, <code class="docutils literal notranslate"><span class="pre">_mate2_001</span></code>, <code class="docutils literal notranslate"><span class="pre">_2_sequence_001</span></code></p></li>
</ul>
<p>For example, a sample named <code class="docutils literal notranslate"><span class="pre">sample-26</span></code> consisting of paired-end reads
could have the FASTQ files <code class="docutils literal notranslate"><span class="pre">sample-26_R1.fq</span></code> and <code class="docutils literal notranslate"><span class="pre">sample-26_R2.fq</span></code>.</p>
<p>If you would like future versions to support additional file extensions,
then please request so by creating an issue (see <a class="reference internal" href="../../issues.html"><span class="doc">Bugs and Requests</span></a>).</p>
</section>
<section id="fastq-path-parsing">
<h3>FASTQ path parsing<a class="headerlink" href="#fastq-path-parsing" title="Permalink to this headline"></a></h3>
<p>For FASTQ files from whole samples (<code class="docutils literal notranslate"><span class="pre">-x</span></code>, <code class="docutils literal notranslate"><span class="pre">-y</span></code>, <code class="docutils literal notranslate"><span class="pre">-z</span></code>), the sample
name is taken from the file name (dropping the mate number, if any).
For example, the single-end FASTQ <code class="docutils literal notranslate"><span class="pre">-z</span> <span class="pre">project/sienna.fq</span></code>
would be parsed to have sample name <code class="docutils literal notranslate"><span class="pre">sienna</span></code>.
And the separate paired-end FASTQ <code class="docutils literal notranslate"><span class="pre">-x</span> <span class="pre">project/lavender_R1.fq</span></code>
would be parsed to have sample name <code class="docutils literal notranslate"><span class="pre">lavender</span></code>.</p>
<p>For demultiplexed FASTQ files (<code class="docutils literal notranslate"><span class="pre">-X</span></code>, <code class="docutils literal notranslate"><span class="pre">-Y</span></code>, <code class="docutils literal notranslate"><span class="pre">-Z</span></code>), the reference
name is taken from the file name (dropping the mate number, if any),
and the sample name is taken from the directory of the FASTQ file.
For example, the single-end FASTQ <code class="docutils literal notranslate"><span class="pre">-Z</span> <span class="pre">project/azure/ochre.fq</span></code>
would be parsed to have sample <code class="docutils literal notranslate"><span class="pre">azure</span></code> and reference <code class="docutils literal notranslate"><span class="pre">ochre</span></code>.
And the separate paired-end FASTQ <code class="docutils literal notranslate"><span class="pre">-X</span> <span class="pre">project/lilac/teal_R2.fq</span></code>
would be parsed to have sample <code class="docutils literal notranslate"><span class="pre">lilac</span></code> and reference <code class="docutils literal notranslate"><span class="pre">teal</span></code>.</p>
</section>
</section>
<section id="fastq-file-uses">
<h2>FASTQ file: Uses<a class="headerlink" href="#fastq-file-uses" title="Permalink to this headline"></a></h2>
<section id="fastq-as-input-file">
<h3>FASTQ as input file<a class="headerlink" href="#fastq-as-input-file" title="Permalink to this headline"></a></h3>
<p>Sequencing reads for these commands must be input as FASTQ files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">all</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">align</span></code></p></li>
</ul>
</section>
<section id="fastq-as-output-file">
<h3>FASTQ as output file<a class="headerlink" href="#fastq-as-output-file" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">align</span></code> command outputs a file in FASTQ format containing the
unaligned reads from each input FASTQ (with option <code class="docutils literal notranslate"><span class="pre">--bt2-un</span></code>).</p></li>
</ul>
</section>
<section id="fastq-as-temporary-file">
<h3>FASTQ as temporary file<a class="headerlink" href="#fastq-as-temporary-file" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">align</span></code> command writes a temporary FASTQ file for each input
FASTQ that it trims with cutadapt (with option <code class="docutils literal notranslate"><span class="pre">--cut</span></code>).</p></li>
</ul>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="fasta.html" class="btn btn-neutral float-left" title="FASTA: Reference sequences" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="xam.html" class="btn btn-neutral float-right" title="SAM, BAM, and CRAM: Alignment Maps" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>