<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Align the sequencing reads &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index.html">User Manuals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Align the sequencing reads</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/manuals/steps/align.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="align-the-sequencing-reads">
<h1>Align the sequencing reads<a class="headerlink" href="#align-the-sequencing-reads" title="Permalink to this heading"></a></h1>
<section id="options-for-fastq-input">
<h2>Options for FASTQ input<a class="headerlink" href="#options-for-fastq-input" title="Permalink to this heading"></a></h2>
<p>SEISMIC-RNA can align single-end reads and paired-end reads where mate 1
and mate 2 reads are either interleaved in one FASTQ file or in separate
FASTQ files labeled 1 and 2.</p>
<section id="fastq-s-of-one-sample">
<h3>FASTQ(s) of one sample<a class="headerlink" href="#fastq-s-of-one-sample" title="Permalink to this heading"></a></h3>
<p>Give the path to a FASTQ file with the <code class="docutils literal notranslate"><span class="pre">-x</span></code>, <code class="docutils literal notranslate"><span class="pre">-y</span></code>, or <code class="docutils literal notranslate"><span class="pre">-z</span></code> option,
depending on the contents of the FASTQ file.</p>
<p>FASTQ files of paired-end reads, with mates 1 and 2 in separate files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ file of paired-end reads, with mates 1 and 2 interleaved:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ file of single-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="fastqs-of-multiple-samples">
<h3>FASTQs of multiple samples<a class="headerlink" href="#fastqs-of-multiple-samples" title="Permalink to this heading"></a></h3>
<p>Give the <code class="docutils literal notranslate"><span class="pre">-x</span></code>, <code class="docutils literal notranslate"><span class="pre">-y</span></code>, or <code class="docutils literal notranslate"><span class="pre">-z</span></code> option multiple times to process more
than one sample simultaneously.</p>
<p>FASTQ files of paired-end reads, with mates 1 and 2 in separate files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_2_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_2_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ files of paired-end reads, with mates 1 and 2 interleaved:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_3</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ files of single-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_3</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="directory-of-fastq-files">
<h3>Directory of FASTQ files<a class="headerlink" href="#directory-of-fastq-files" title="Permalink to this heading"></a></h3>
<p>If any given path is a directory instead of a FASTQ file, then it will
be searched for FASTQ files recursively (any subdirectories will also be
searched, any of their subdirectories will be searched, and so on).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Any file that does not have a valid FASTQ extension will be ignored,
so if the directory contains a mixture of FASTQ and non-FASTQ files,
then the program will still work.</p>
</div>
<p>Directory containing FASTQ files of paired-end reads, with mates 1 and 2
in separate files; note that if a directory contains both the mate 1 and
the mate 2 FASTQ files, then you only need to give it once:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">samples</span><span class="o">/</span>
</pre></div>
</div>
<p>Directory containing FASTQ files of paired-end reads, with mates 1 and 2
interleaved:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">y</span> <span class="n">samples</span><span class="o">/</span>
</pre></div>
</div>
<p>Directory containing FASTQ files of single-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">z</span> <span class="n">samples</span><span class="o">/</span>
</pre></div>
</div>
</section>
<section id="align-any-combination-of-the-above">
<h3>Align any combination of the above<a class="headerlink" href="#align-any-combination-of-the-above" title="Permalink to this heading"></a></h3>
<p>You may any combination of the above options for specifying FASTQ files
in a single command. For example, to align</p>
<ul class="simple">
<li><p>two directories of paired-end reads with mates 1 and 2 separated</p></li>
<li><p>one sample of paired-end reads with mates 1 and 2 interleaved</p></li>
<li><p>two samples of single-end reads</p></li>
</ul>
<p>one could type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">samples_1</span><span class="o">-</span><span class="mi">10</span><span class="o">/</span> <span class="o">-</span><span class="n">x</span> <span class="n">samples_11</span><span class="o">-</span><span class="mi">20</span><span class="o">/</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_21</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_22</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_23</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
</section>
<section id="options-for-phred-score-encoding">
<h2>Options for Phred score encoding<a class="headerlink" href="#options-for-phred-score-encoding" title="Permalink to this heading"></a></h2>
<p>SEISMIC-RNA defaults to using Phred+33 encoding for FASTQ files, which
is standard on modern Illumina sequencers. To change the Phred score
encoding, use the option <code class="docutils literal notranslate"><span class="pre">--phred-enc</span></code>.</p>
</section>
<section id="options-for-quality-assessment-with-fastqc">
<h2>Options for quality assessment with FastQC<a class="headerlink" href="#options-for-quality-assessment-with-fastqc" title="Permalink to this heading"></a></h2>
<p>By default, each FASTQ file is processed with <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, both before and
after trimming, in order to find any potential problems. FastQC can be
disabled with the flag <code class="docutils literal notranslate"><span class="pre">--no-fastqc</span></code>. To enable automatic extraction
of the zipped output files from FastQC, add the flag <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code>.</p>
</section>
<section id="options-for-trimming-with-cutadapt">
<h2>Options for trimming with Cutadapt<a class="headerlink" href="#options-for-trimming-with-cutadapt" title="Permalink to this heading"></a></h2>
<p>By default, each FASTQ file and pair of mated FASTQ files is trimmed for
adapters and low-quality bases using <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">Cutadapt</a>. To disable adapter
trimming, add the flag <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code>.</p>
<section id="adapter-trimming">
<h3>Adapter trimming<a class="headerlink" href="#adapter-trimming" title="Permalink to this heading"></a></h3>
<p>By default, SEISMIC-RNA uses the standard, minimal adapter sequences for
Illumina sequencing runs for both read 1 and (if paired-end) read 2:</p>
<ul class="simple">
<li><p>5’: <code class="docutils literal notranslate"><span class="pre">GCTCTTCCGATCT</span></code></p></li>
<li><p>3’: <code class="docutils literal notranslate"><span class="pre">AGATCGGAAGAGC</span></code></p></li>
</ul>
<p>To use another adapter, type its sequence after the appropriate option:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Side</p></th>
<th class="head"><p>Read</p></th>
<th class="head"><p>Option</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>5’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>5’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g2</span></code></p></td>
</tr>
<tr class="row-even"><td><p>3’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>3’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a2</span></code></p></td>
</tr>
</tbody>
</table>
</section>
<section id="quality-trimming">
<h3>Quality trimming<a class="headerlink" href="#quality-trimming" title="Permalink to this heading"></a></h3>
<p>Base calls on either end of a read that fall below a minimum Phred score
quality are trimmed with Cutadapt. The default minimum quality is 25,
which corresponds to a confidence of 1 - 10 <sup>-2.5</sup> = 0.997 that the
base call is correct. To change the quality threshold, use the option
<code class="docutils literal notranslate"><span class="pre">--min-phred</span></code>.</p>
</section>
<section id="dark-cycle-trimming">
<h3>Dark cycle trimming<a class="headerlink" href="#dark-cycle-trimming" title="Permalink to this heading"></a></h3>
<p>On some Illumina sequencers (e.g. NextSeq, iSeq), the probes used to
detect G bases emit no light. Hence, these instruments will label a base
call as a G if it appears dark. If sequencing reaches the end of a read,
then there will be no more bases to sequence, so every cycle thereafter
will be dark, causing a string of Gs to be added to the 3’ end of the
read. Using the option <code class="docutils literal notranslate"><span class="pre">--cut-nextseq</span></code> tells Cutadapt to trim off any
high-quality G bases from the 3’ end of each read. This may improve the
alignment (especially in end-to-end mode) but also removes real G bases
from the 3’ ends of reads (since they cannot be distinguished from any
artefactual G bases).</p>
</section>
<section id="additional-options-for-cutadapt">
<h3>Additional options for Cutadapt<a class="headerlink" href="#additional-options-for-cutadapt" title="Permalink to this heading"></a></h3>
<p>See the command line interface reference for the full list of options
that SEISMIC-RNA can use with Cutadapt. These options should suffice for
most users. If you require a more customized adapter trimming workflow,
we recommend that you perform adapter trimming and alignment outside of
SEISMIC-RNA and pass your BAM file(s) into the relate step.</p>
</section>
</section>
<section id="options-for-alignment-with-bowtie-2">
<h2>Options for alignment with Bowtie 2<a class="headerlink" href="#options-for-alignment-with-bowtie-2" title="Permalink to this heading"></a></h2>
<section id="bowtie-2-index-files-automatic-and-pre-built">
<h3>Bowtie 2 index files: automatic and pre-built<a class="headerlink" href="#bowtie-2-index-files-automatic-and-pre-built" title="Permalink to this heading"></a></h3>
<p>SEISMIC-RNA searches for a set of Bowtie 2 index files in the directory
where the FASTA file is. A valid Bowtie 2 index comprises six files, all
with the same name as the FASTA file, with the extensions <code class="docutils literal notranslate"><span class="pre">.1.bt2</span></code>,
<code class="docutils literal notranslate"><span class="pre">.2.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.3.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.4.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.rev.1.bt2</span></code>, and <code class="docutils literal notranslate"><span class="pre">.rev.2.bt2</span></code>.
If this set of files exists, then SEISMIC-RNA uses it as the index.
Otherwise, it calls <code class="docutils literal notranslate"><span class="pre">bowtie2-build</span></code> to build an index in a temporary
directory, which is deleted after alignment finishes. Indexing a small
FASTA file takes several seconds. However, indexing a very large FASTA
(e.g. a whole transcriptome) can take hours, so it is advantageous to
pre-build your index in the same directory as the FASTA file, to save
time in case you need to align more than once. You can pre-build an
index with the command</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bowtie2</span><span class="o">-</span><span class="n">build</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="n">refs</span>
</pre></div>
</div>
<p>replacing <cite>refs</cite> with the path to and name of your FASTA file. See the
<a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer">Bowtie 2 Indexer manual</a> for more details.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If SEISMIC-RNA finds a pre-built Bowtie 2 index, then it does <em>not</em>
verify that the index was actually built from the FASTA file of the
same name. You must verify this yourself if using a pre-built index.
You can assume the index is correct if you build it using the above
command and avoid modifying or replacing the FASTA and index files.</p>
</div>
</section>
<section id="alignment-modes-local-and-end-to-end">
<h3>Alignment modes: local and end-to-end<a class="headerlink" href="#alignment-modes-local-and-end-to-end" title="Permalink to this heading"></a></h3>
<p>During alignment, Bowtie 2 can either align the entirety of each read
(end-to-end mode) or find and align only the section of the read that
yields the best alignment score (local mode). Generally, end-to-end mode
produces more spurious mutations (false positives) caused by artefacts
such as untrimmed adapters at the ends of reads; while local mode drops
more real mutations (false negatives) within several nucleotides of the
ends of reads. For mutational profiling, false positives are much more
deleterious than false negatives, so SEISMIC-RNA defaults to local mode
(<code class="docutils literal notranslate"><span class="pre">--bt2-local</span></code>). End-to-end mode (<code class="docutils literal notranslate"><span class="pre">--bt2-end-to-end</span></code>) should be used
only with amplicon-based samples and careful adapter trimming; and even
then, local mode works well enough, except for counting mutations at the
very ends of reads.</p>
</section>
<section id="filtering-alignments">
<h3>Filtering alignments<a class="headerlink" href="#filtering-alignments" title="Permalink to this heading"></a></h3>
<p>Alignments can be filtered by <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">alignment score</a> and <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">mapping quality</a>,
which are distinct properties. Alignment score measures how well a read
aligns to a given location in the reference. It is calculated from the
number of matches, substitutions, and gaps using the score parameters.
The minimum alignment scores for local and end-to-end modes can be set
using <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-loc</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-e2e</span></code>, respectively.
Mapping quality measures how unique an alignment is: high quality if the
read aligns with a high score to exactly one location, low quality if it
aligns with similar scores to multiple locations in the reference. The
minimum mapping quality can be set with the option <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>.</p>
</section>
<section id="outputting-unaligned-reads-pairs">
<h3>Outputting unaligned reads/pairs<a class="headerlink" href="#outputting-unaligned-reads-pairs" title="Permalink to this heading"></a></h3>
<p>For some reasons, including troubleshooting low alignment rates, it can
be helpful to output all reads. The flag <code class="docutils literal notranslate"><span class="pre">--bt2-unal</span></code> causes all reads
(including those that did not align) to appear in the temporary SAM file
that is output directly from Bowtie 2. Because this file is located in
the temporary directory, the <code class="docutils literal notranslate"><span class="pre">--save-temp</span></code> flag must also be used, or
else the SAM file (and everything else in the directory) will be deleted
when the alignment step finishes.</p>
</section>
<section id="paired-end-options">
<h3>Paired-end options<a class="headerlink" href="#paired-end-options" title="Permalink to this heading"></a></h3>
<p>Bowtie 2 considers paired-end reads to align “concordantly” when their
relative positions match expectations and “discordantly” otherwise. See
the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#concordant-pairs-match-pair-expectations-discordant-pairs-dont">Bowtie 2 manual for details on concordant/discordant alignments</a>.
By default, SEISMIC-RNA treats only concordantly aligning pairs as valid
alignments. To also treat discordant pairs as valid alignments, use the
flag <code class="docutils literal notranslate"><span class="pre">--bt2-discordant</span></code>.</p>
<p>Several options control which types of alignments are concordant. First,
the expected orientation of paired mates is set using <code class="docutils literal notranslate"><span class="pre">--bt2-orient</span></code>.
It can be <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the 5’-most mate is forward, the 3’-most is reversed),
<code class="docutils literal notranslate"><span class="pre">rf</span></code> (the 5’-most mate is reversed, the 3’-most is forward), or <code class="docutils literal notranslate"><span class="pre">ff</span></code>
(both mates are forward). The default is <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the most common type).
Second, the mates may <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mates-can-overlap-contain-or-dovetail-each-other">overlap partially or completely, or dovetail</a>.
By default, overlaps (partial and complete) are considered concordant,
and dovetailing is considered discordant. The flag <code class="docutils literal notranslate"><span class="pre">--bt2-no-contain</span></code>
treats as discordant pairs where one mate completely overlaps the other,
while <code class="docutils literal notranslate"><span class="pre">--bt2-dovetail</span></code> treats dovetailed pairs as concordant. Pairs
that overlap partially are always considered concordant in SEISMIC-RNA.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The flags <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]contain</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]dovetail</span></code> choose
whether to treat these types of overlaps as concordant (yes) or
discordant (no). If they are treated as discordant, then the flag
<code class="docutils literal notranslate"><span class="pre">--bt2-[no-]discordant</span></code> determines whether they are considered
valid alignments (yes) or invalid (no).</p>
</div>
<p>The option <code class="docutils literal notranslate"><span class="pre">--bt2-mixed</span></code> enables <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a> wherein, for pairs that
fail to produce a valid paired-end alignment, Bowtie 2 attempts to align
each mate individually (as if it were a single-end read).</p>
</section>
</section>
<section id="troubleshooting-alignment">
<h2>Troubleshooting alignment<a class="headerlink" href="#troubleshooting-alignment" title="Permalink to this heading"></a></h2>
<section id="alignment-rate-is-low">
<h3>Alignment rate is low<a class="headerlink" href="#alignment-rate-is-low" title="Permalink to this heading"></a></h3>
<p>If the percent of reads aligning to the reference is less than expected,
then try the following steps (in this order):</p>
<ol class="arabic simple">
<li><p>Ensure you are using Bowtie version 2.5.1 or later (version 2.5.0
has a bug that affects alignment rate). You can check the version by
running <code class="docutils literal notranslate"><span class="pre">bowtie2</span> <span class="pre">--version</span> <span class="pre">|</span> <span class="pre">head</span> <span class="pre">-n</span> <span class="pre">1</span></code>.</p></li>
<li><p>Double check that the FASTA has the correct reference sequence(s)
and that, if the Bowtie 2 index was pre-built before the align step,
that the correct FASTA file was used.</p></li>
<li><p>Rerun alignment using the flags <code class="docutils literal notranslate"><span class="pre">--bt2-unal</span></code> and <code class="docutils literal notranslate"><span class="pre">--save-temp</span></code>,
which will write all the unaligned reads to the temporary SAM file
and keep that file after alignment ends. Find unaligned reads with
<code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">view</span> <span class="pre">-f</span> <span class="pre">4</span> <span class="pre">temp/sample/align/align-2_align/refs.sam</span> <span class="pre">-o</span> <span class="pre">x</span></code>
where <code class="docutils literal notranslate"><span class="pre">sample</span></code>, <code class="docutils literal notranslate"><span class="pre">refs</span></code>, and <code class="docutils literal notranslate"><span class="pre">x</span></code> are replaced with the name of
the sample, name of the FASTA file, and name of the SAM file into
which to write the unaligned reads, respectively. Use <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&amp;PAGE_TYPE=BlastSearch&amp;LINK_LOC=blasthome">BLAST</a> to
identify the source of the unaligned reads, to infer the problem.</p></li>
</ol>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>