<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Steps of the workflow &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Input file patterns" href="../inputs.html" />
    <link rel="prev" title="Manuals" href="../index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">Manuals</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Steps of the workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#align-the-sequencing-reads">Align the sequencing reads</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fasta-input-file">FASTA input file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#options-for-fastq-input">Options for FASTQ input</a></li>
<li class="toctree-l4"><a class="reference internal" href="#options-for-phred-score-encoding">Options for Phred score encoding</a></li>
<li class="toctree-l4"><a class="reference internal" href="#options-for-quality-assessment-with-fastqc">Options for quality assessment with FastQC</a></li>
<li class="toctree-l4"><a class="reference internal" href="#options-for-trimming-with-cutadapt">Options for trimming with Cutadapt</a></li>
<li class="toctree-l4"><a class="reference internal" href="#options-for-alignment-with-bowtie-2">Options for alignment with Bowtie 2</a></li>
<li class="toctree-l4"><a class="reference internal" href="#troubleshooting-alignment">Troubleshooting alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#relate-each-read-to-every-reference-position">Relate each read to every reference position</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#input-files-for-relate">Input files for relate</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#all-steps-with-one-command">All steps with one command</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">Input file patterns</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Logging messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Manuals</a></li>
      <li class="breadcrumb-item active">Steps of the workflow</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/manuals/steps/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="steps-of-the-workflow">
<h1>Steps of the workflow<a class="headerlink" href="#steps-of-the-workflow" title="Permalink to this headline"></a></h1>
<section id="align-the-sequencing-reads">
<h2>Align the sequencing reads<a class="headerlink" href="#align-the-sequencing-reads" title="Permalink to this headline"></a></h2>
<section id="fasta-input-file">
<h3>FASTA input file<a class="headerlink" href="#fasta-input-file" title="Permalink to this headline"></a></h3>
<p>Alignment requires exactly one file of reference sequences, which must
be in FASTA format.</p>
</section>
<section id="options-for-fastq-input">
<h3>Options for FASTQ input<a class="headerlink" href="#options-for-fastq-input" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA can align single-end reads and paired-end reads where mate 1
and mate 2 reads are either interleaved in one FASTQ file or in separate
FASTQ files labeled 1 and 2.
It can also align both “regular” FASTQ files that contain reads from a
mixture of references and “demultiplexed” FASTQ files that contain reads
from only one reference (e.g. generated by <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">demult</span></code>).</p>
<section id="fastq-files-of-single-end-or-paired-end-reads">
<h4>FASTQ files of single-end or paired-end reads<a class="headerlink" href="#fastq-files-of-single-end-or-paired-end-reads" title="Permalink to this headline"></a></h4>
<p>Sequencing reads can be single-end or paired-end (for more details, see
<a class="reference internal" href="../../formats/data/index.html#fastq-endedness"><span class="std std-ref">Endedness: single-end and paired-end reads</span></a>).
The SEISMIC-RNA workflow can handle both single- and paired-end reads.
Each FASTQ file must contain one type or the other, not a mix of both.
Thus, we refer to entire FASTQ files as “paired-end”
However, SEISMIC-RNA can process FASTQ files of single-end reads and
other FASTQ files of paired-end reads simultaneously.</p>
</section>
<section id="fastq-files-of-whole-samples-and-demultiplexed-samples">
<h4>FASTQ files of whole samples and demultiplexed samples<a class="headerlink" href="#fastq-files-of-whole-samples-and-demultiplexed-samples" title="Permalink to this headline"></a></h4>
</section>
<section id="fastq-s-of-one-sample">
<h4>FASTQ(s) of one sample<a class="headerlink" href="#fastq-s-of-one-sample" title="Permalink to this headline"></a></h4>
<p>Give the path to a FASTQ file with the <code class="docutils literal notranslate"><span class="pre">-x</span></code>, <code class="docutils literal notranslate"><span class="pre">-y</span></code>, or <code class="docutils literal notranslate"><span class="pre">-z</span></code> option,
depending on the contents of the FASTQ file.</p>
<p>FASTQ files of paired-end reads, with mates 1 and 2 in separate files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ file of paired-end reads, with mates 1 and 2 interleaved:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ file of single-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="fastqs-of-multiple-samples">
<h4>FASTQs of multiple samples<a class="headerlink" href="#fastqs-of-multiple-samples" title="Permalink to this headline"></a></h4>
<p>Give the <code class="docutils literal notranslate"><span class="pre">-x</span></code>, <code class="docutils literal notranslate"><span class="pre">-y</span></code>, or <code class="docutils literal notranslate"><span class="pre">-z</span></code> option multiple times to process more
than one sample simultaneously.</p>
<p>FASTQ files of paired-end reads, with mates 1 and 2 in separate files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_1_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_2_R1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">x</span> <span class="n">sample_2_R2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ files of paired-end reads, with mates 1 and 2 interleaved:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_3</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>FASTQ files of single-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_3</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="directory-of-fastq-files">
<h4>Directory of FASTQ files<a class="headerlink" href="#directory-of-fastq-files" title="Permalink to this headline"></a></h4>
<p>If any given path is a directory instead of a FASTQ file, then it will
be searched for FASTQ files recursively (any subdirectories will also be
searched, any of their subdirectories will be searched, and so on).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Any file that does not have a valid FASTQ extension will be ignored,
so if the directory contains a mixture of FASTQ and non-FASTQ files,
then the program will still work.</p>
</div>
<p>Directory containing FASTQ files of paired-end reads, with mates 1 and 2
in separate files; note that if a directory contains both the mate 1 and
the mate 2 FASTQ files, then you only need to give it once:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">samples</span><span class="o">/</span>
</pre></div>
</div>
<p>Directory containing FASTQ files of paired-end reads, with mates 1 and 2
interleaved:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">y</span> <span class="n">samples</span><span class="o">/</span>
</pre></div>
</div>
<p>Directory containing FASTQ files of single-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">z</span> <span class="n">samples</span><span class="o">/</span>
</pre></div>
</div>
</section>
<section id="align-any-combination-of-the-above">
<h4>Align any combination of the above<a class="headerlink" href="#align-any-combination-of-the-above" title="Permalink to this headline"></a></h4>
<p>You may any combination of the above options for specifying FASTQ files
in a single command.
For example, to align</p>
<ul class="simple">
<li><p>two directories of paired-end reads with mates 1 and 2 separated</p></li>
<li><p>one sample of paired-end reads with mates 1 and 2 interleaved</p></li>
<li><p>two samples of single-end reads</p></li>
</ul>
<p>one could type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">x</span> <span class="n">samples_1</span><span class="o">-</span><span class="mi">10</span><span class="o">/</span> <span class="o">-</span><span class="n">x</span> <span class="n">samples_11</span><span class="o">-</span><span class="mi">20</span><span class="o">/</span> <span class="o">-</span><span class="n">y</span> <span class="n">sample_21</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_22</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">z</span> <span class="n">sample_23</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
</section>
<section id="options-for-phred-score-encoding">
<h3>Options for Phred score encoding<a class="headerlink" href="#options-for-phred-score-encoding" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA defaults to using Phred+33 encoding for FASTQ files, which
is standard on modern Illumina sequencers.
To change the Phred score encoding, use the option <code class="docutils literal notranslate"><span class="pre">--phred-enc</span></code>.</p>
</section>
<section id="options-for-quality-assessment-with-fastqc">
<h3>Options for quality assessment with FastQC<a class="headerlink" href="#options-for-quality-assessment-with-fastqc" title="Permalink to this headline"></a></h3>
<p>By default, each FASTQ file is processed with <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, both before and
after trimming, in order to find any potential problems.
FastQC can be disabled with the flag <code class="docutils literal notranslate"><span class="pre">--no-fastqc</span></code>.
To enable automatic extraction of the zipped output files from FastQC,
add the flag <code class="docutils literal notranslate"><span class="pre">--qc-extract</span></code>.</p>
</section>
<section id="options-for-trimming-with-cutadapt">
<h3>Options for trimming with Cutadapt<a class="headerlink" href="#options-for-trimming-with-cutadapt" title="Permalink to this headline"></a></h3>
<p>By default, each FASTQ file and pair of mated FASTQ files is trimmed for
adapters and low-quality bases using <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">Cutadapt</a>. To disable adapter
trimming, add the flag <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code>.</p>
<section id="adapter-trimming">
<h4>Adapter trimming<a class="headerlink" href="#adapter-trimming" title="Permalink to this headline"></a></h4>
<p>By default, SEISMIC-RNA uses the standard, minimal adapter sequences for
Illumina sequencing runs for both read 1 and (if paired-end) read 2:</p>
<ul class="simple">
<li><p>5’: <code class="docutils literal notranslate"><span class="pre">GCTCTTCCGATCT</span></code></p></li>
<li><p>3’: <code class="docutils literal notranslate"><span class="pre">AGATCGGAAGAGC</span></code></p></li>
</ul>
<p>To use another adapter, type its sequence after the appropriate option:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 23%" />
<col style="width: 23%" />
<col style="width: 54%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Side</p></th>
<th class="head"><p>Read</p></th>
<th class="head"><p>Option</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>5’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>5’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-g2</span></code></p></td>
</tr>
<tr class="row-even"><td><p>3’</p></td>
<td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a1</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>3’</p></td>
<td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--cut-a2</span></code></p></td>
</tr>
</tbody>
</table>
</section>
<section id="quality-trimming">
<h4>Quality trimming<a class="headerlink" href="#quality-trimming" title="Permalink to this headline"></a></h4>
<p>Base calls on either end of a read that fall below a minimum Phred score
quality are trimmed with Cutadapt.
The default minimum quality is 25, which corresponds to a confidence of
1 - 10 <sup>-2.5</sup> = 0.997 that the base call is correct.
To change the quality threshold, use the option <code class="docutils literal notranslate"><span class="pre">--min-phred</span></code>.</p>
</section>
<section id="dark-cycle-trimming">
<h4>Dark cycle trimming<a class="headerlink" href="#dark-cycle-trimming" title="Permalink to this headline"></a></h4>
<p>On some Illumina sequencers (e.g. NextSeq, iSeq), the probes used to
detect G bases emit no light. Hence, these instruments will label a base
call as a G if it appears dark. If sequencing reaches the end of a read,
then there will be no more bases to sequence, so every cycle thereafter
will be dark, causing a string of Gs to be added to the 3’ end of the
read. Using the option <code class="docutils literal notranslate"><span class="pre">--cut-nextseq</span></code> tells Cutadapt to trim off any
high-quality G bases from the 3’ end of each read. This may improve the
alignment (especially in end-to-end mode) but also removes real G bases
from the 3’ ends of reads (since they cannot be distinguished from any
artefactual G bases).</p>
</section>
<section id="additional-options-for-cutadapt">
<h4>Additional options for Cutadapt<a class="headerlink" href="#additional-options-for-cutadapt" title="Permalink to this headline"></a></h4>
<p>See <a class="reference internal" href="../../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can
use with Cutadapt, and the <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/reference.html">Cutadapt reference guide</a> for details on
each of these options.
These options should suffice for most users.
If you require a more customized adapter trimming workflow, you can trim
your FASTQ files outside of SEISMIC-RNA, then perform alignment within
SEISMIC-RNA using the option <code class="docutils literal notranslate"><span class="pre">--no-cut</span></code> to disable additional adapter
trimming.</p>
</section>
</section>
<section id="options-for-alignment-with-bowtie-2">
<h3>Options for alignment with Bowtie 2<a class="headerlink" href="#options-for-alignment-with-bowtie-2" title="Permalink to this headline"></a></h3>
<section id="bowtie-2-index-files-automatic-and-pre-built">
<h4>Bowtie 2 index files: automatic and pre-built<a class="headerlink" href="#bowtie-2-index-files-automatic-and-pre-built" title="Permalink to this headline"></a></h4>
<p>SEISMIC-RNA searches for a set of Bowtie 2 index files in the directory
where the FASTA file is.
A valid Bowtie 2 index comprises six files, all with the same name as
the FASTA file, with the extensions <code class="docutils literal notranslate"><span class="pre">.1.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.2.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.3.bt2</span></code>,
<code class="docutils literal notranslate"><span class="pre">.4.bt2</span></code>, <code class="docutils literal notranslate"><span class="pre">.rev.1.bt2</span></code>, and <code class="docutils literal notranslate"><span class="pre">.rev.2.bt2</span></code>.
If this set of files exists, then SEISMIC-RNA uses it as the index.
Otherwise, it calls <code class="docutils literal notranslate"><span class="pre">bowtie2-build</span></code> to build an index in a temporary
directory, which is deleted after alignment finishes.
Indexing a small FASTA file takes several seconds; thus, it is generally
more convenient to build a temporary .
However, indexing a very large FASTA (e.g. a whole transcriptome) can
take hours, so it is advantageous to pre-build your index in the same
directory as the FASTA file, to save time in case you need to align more
than once.
You can pre-build an index with this command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bowtie2</span><span class="o">-</span><span class="n">build</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="p">{</span><span class="n">refs</span><span class="p">}</span>
</pre></div>
</div>
<p>replacing <code class="docutils literal notranslate"><span class="pre">{refs}</span></code> with the path to and name of your FASTA file.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer">Bowtie 2 Indexer manual</a> for more details.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If SEISMIC-RNA finds a pre-built Bowtie 2 index, then it does <em>not</em>
verify that the index was actually built from the FASTA file of the
same name.
You must verify this yourself if using a pre-built index.
You can assume the index is correct if you build it using the above
command and avoid modifying or replacing the FASTA and index files.</p>
</div>
</section>
<section id="alignment-modes-local-and-end-to-end">
<h4>Alignment modes: local and end-to-end<a class="headerlink" href="#alignment-modes-local-and-end-to-end" title="Permalink to this headline"></a></h4>
<p>During alignment, Bowtie 2 can either align the entirety of each read
(end-to-end mode) or find and align only the section of the read that
yields the best alignment score (local mode).
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#end-to-end-alignment-versus-local-alignment">description of alignment modes in Bowtie 2</a> for more details.</p>
<p>Generally, end-to-end mode yields spurious mutations (false positives)
at the ends of reads if the reads contain artifacts such as low-quality
base calls or untrimmed or improperly trimmed adapters.
Conversely, local mode misses real mutations (false negatives) within
several nucleotides of the ends of reads because such mutations are not,
by definition, part of the best local alignment.</p>
<p>Concerning RNA mutational profiling, false positives are generally much
more problematic than false negatives, so SEISMIC-RNA uses local mode
(<code class="docutils literal notranslate"><span class="pre">--bt2-local</span></code>) by default.
Use end-to-end mode (<code class="docutils literal notranslate"><span class="pre">--bt2-end-to-end</span></code>) only if you have a compelling
reason to do so (e.g. if it is essential to detect mutations at the ends
of reads) and only after carefully trimming any extraneous sequences
from the ends of the reads.</p>
</section>
<section id="filtering-alignments">
<h4>Filtering alignments<a class="headerlink" href="#filtering-alignments" title="Permalink to this headline"></a></h4>
<p>Alignments can be filtered by <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">alignment score</a> and <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">mapping quality</a>,
which are distinct properties.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#scores-higher-more-similar">Alignment score</a> measures how well a read aligns to a given location
in the reference.
It is calculated from the number of matches, substitutions, and gaps
using the score parameters.
The minimum alignment scores for local and end-to-end modes can be set
using <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-loc</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-score-min-e2e</span></code>, respectively.
See the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#valid-alignments-meet-or-exceed-the-minimum-score-threshold">section of the Bowtie 2 manual on alignment scores</a> for advice
on setting this parameter.</p>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mapping-quality-higher-more-unique">Mapping quality</a> measures how unique an alignment is: high quality if
the read aligns with a high score to exactly one location, low quality
if it aligns with similar scores to multiple locations in the reference.
The default minimum quality is 25, which corresponds to a confidence of
1 - 10 <sup>-2.5</sup> = 0.997 that the read has aligned correctly.
To change the quality threshold, use the option <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>.
For those searching for this option in Bowtie 2, you will not find it.
Instead, reads with insufficient mapping quality are filtered out after
alignment using the <a class="reference external" href="https://www.htslib.org/doc/samtools-view.html">view command in Samtools</a>.</p>
</section>
<section id="examining-unaligned-reads-pairs">
<h4>Examining unaligned reads/pairs<a class="headerlink" href="#examining-unaligned-reads-pairs" title="Permalink to this headline"></a></h4>
<p>For some reasons, including troubleshooting low alignment rates, it can
be helpful to examine the reads that did not align. Depending on the</p>
</section>
<section id="paired-end-options">
<h4>Paired-end options<a class="headerlink" href="#paired-end-options" title="Permalink to this headline"></a></h4>
<p>Bowtie 2 considers paired-end reads to align “concordantly” when their
relative positions match expectations and “discordantly” otherwise. See
the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#concordant-pairs-match-pair-expectations-discordant-pairs-dont">Bowtie 2 manual for details on concordant/discordant alignments</a>.
By default, SEISMIC-RNA treats only concordantly aligning pairs as valid
alignments. To also treat discordant pairs as valid alignments, use the
flag <code class="docutils literal notranslate"><span class="pre">--bt2-discordant</span></code>.</p>
<p>Several options control which types of alignments are concordant. First,
the expected orientation of paired mates is set using <code class="docutils literal notranslate"><span class="pre">--bt2-orient</span></code>.
It can be <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the 5’-most mate is forward, the 3’-most is reversed),
<code class="docutils literal notranslate"><span class="pre">rf</span></code> (the 5’-most mate is reversed, the 3’-most is forward), or <code class="docutils literal notranslate"><span class="pre">ff</span></code>
(both mates are forward). The default is <code class="docutils literal notranslate"><span class="pre">fr</span></code> (the most common type).
Second, the mates may <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mates-can-overlap-contain-or-dovetail-each-other">overlap partially or completely, or dovetail</a>.
By default, overlaps (partial and complete) are considered concordant,
and dovetailing is considered discordant. The flag <code class="docutils literal notranslate"><span class="pre">--bt2-no-contain</span></code>
treats as discordant pairs where one mate completely overlaps the other,
while <code class="docutils literal notranslate"><span class="pre">--bt2-dovetail</span></code> treats dovetailed pairs as concordant. Pairs
that overlap partially are always considered concordant in SEISMIC-RNA.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The flags <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]contain</span></code> and <code class="docutils literal notranslate"><span class="pre">--bt2-[no-]dovetail</span></code> choose
whether to treat these types of overlaps as concordant (yes) or
discordant (no). If they are treated as discordant, then the flag
<code class="docutils literal notranslate"><span class="pre">--bt2-[no-]discordant</span></code> determines whether they are considered
valid alignments (yes) or invalid (no).</p>
</div>
<p>The option <code class="docutils literal notranslate"><span class="pre">--bt2-mixed</span></code> enables <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a> wherein, for pairs that
fail to produce a valid paired-end alignment, Bowtie 2 attempts to align
each mate individually (as if it were a single-end read).</p>
</section>
<section id="additional-options-for-bowtie-2">
<h4>Additional options for Bowtie 2<a class="headerlink" href="#additional-options-for-bowtie-2" title="Permalink to this headline"></a></h4>
<p>See <a class="reference internal" href="../../cli.html#cli-align"><span class="std std-ref">seismic align</span></a> for the full list of options that SEISMIC-RNA can
use with Bowtie 2, and the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">Bowtie 2 manual</a> for details on each of
these options.
These options should suffice for most users.
If you require a more customized alignment workflow, you can align your
your FASTQ files outside of SEISMIC-RNA, then pass the resulting XAM
files into SEISMIC-RNA at the step <a class="reference internal" href="relate.html#step-relate"><span class="std std-ref">Relate each read to every reference position</span></a>.</p>
</section>
</section>
<section id="troubleshooting-alignment">
<h3>Troubleshooting alignment<a class="headerlink" href="#troubleshooting-alignment" title="Permalink to this headline"></a></h3>
<section id="alignment-rate-is-low">
<span id="low-align-rate"></span><h4>Alignment rate is low<a class="headerlink" href="#alignment-rate-is-low" title="Permalink to this headline"></a></h4>
<p>If the percent of reads aligning to the reference is less than expected,
then try the following steps (in this order):</p>
<ol class="arabic simple">
<li><p>Ensure you are using Bowtie version 2.5.1 or later (version 2.5.0
has a bug that affects alignment rate). You can check the version by
running <code class="docutils literal notranslate"><span class="pre">bowtie2</span> <span class="pre">--version</span> <span class="pre">|</span> <span class="pre">head</span> <span class="pre">-n</span> <span class="pre">1</span></code>.</p></li>
<li><p>Double check that the FASTA has the correct reference sequence(s)
and that, if the Bowtie 2 index was pre-built before the align step,
that the correct FASTA file was used.</p></li>
<li><p>Reads that failed to align with Bowtie2 are written to the file
<code class="docutils literal notranslate"><span class="pre">out/{sample}/align/unaligned.</span></code> Open the SAM file,
process several unaligned reads randomly, and use <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&amp;PAGE_TYPE=BlastSearch&amp;LINK_LOC=blasthome">BLAST</a> to discern
their origins, which can help in deducing what went wrong.</p></li>
</ol>
</section>
</section>
</section>
<section id="relate-each-read-to-every-reference-position">
<span id="step-relate"></span><h2>Relate each read to every reference position<a class="headerlink" href="#relate-each-read-to-every-reference-position" title="Permalink to this headline"></a></h2>
<section id="input-files-for-relate">
<h3>Input files for relate<a class="headerlink" href="#input-files-for-relate" title="Permalink to this headline"></a></h3>
<p>Relate requires exactly one FASTA file containing one or more reference
sequences and any number of alignment map files in SAM/BAM/CRAM format
(referred to collectively as “XAM” format).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The references in the FASTA file must match those to which the reads
were aligned to produce the XAM file(s); the names and the sequences
must be identical. If the names differ, then the XAM files using the
old names will be ignored; while if the sequences differ, then reads
can yield erroneous relation vectors or fail to yield any output.</p>
</div>
<section id="one-xam-file">
<h4>One XAM file<a class="headerlink" href="#one-xam-file" title="Permalink to this headline"></a></h4>
<p>A single XAM file (<code class="docutils literal notranslate"><span class="pre">sample_1/align/ref_1.bam</span></code>) can be run as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">relate</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="n">sample_1</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref_1</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</section>
<section id="multiple-xam-files">
<h4>Multiple XAM files<a class="headerlink" href="#multiple-xam-files" title="Permalink to this headline"></a></h4>
<p>Multiple XAM files can be run in parallel by giving multiple paths:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">relate</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="n">sample_1</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref_1</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>and/or by using <a href="#id1"><span class="problematic" id="id2">`glob patterns`_</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">relate</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="n">sample_</span><span class="o">*/</span><span class="n">align</span><span class="o">/</span><span class="n">ref_1</span><span class="o">.</span><span class="n">bam</span> <span class="n">sample_</span><span class="o">*/</span><span class="n">align</span><span class="o">/</span><span class="n">ref_2</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</section>
<section id="xam-file-content-requirements">
<h4>XAM file content requirements<a class="headerlink" href="#xam-file-content-requirements" title="Permalink to this headline"></a></h4>
<p>Generally, a XAM file can contain reads that have aligned to any number
of reference, as well as unaligned reads. However, SEISMIC-RNA requires
that each XAM file contain reads aligned to exactly one reference. This
restriction enables the relate step to process XAM files in parallel,
which increases the speed. If the XAM files were created using <code class="docutils literal notranslate"><span class="pre">seismic</span>
<span class="pre">align</span></code>, then they are guaranteed follow this convention.</p>
</section>
<section id="xam-file-path-requirements">
<h4>XAM file path requirements<a class="headerlink" href="#xam-file-path-requirements" title="Permalink to this headline"></a></h4>
<p>The name of the XAM file (minus the file extension) must be the name of
the reference to which it was aligned. It must be inside a directory
named <code class="docutils literal notranslate"><span class="pre">align</span></code>, which must be inside a directory named after the sample
from which the reads came. If the BAM files were created using <code class="docutils literal notranslate"><span class="pre">seismic</span>
<span class="pre">align</span></code>, then they will already follow this convention.</p>
</section>
</section>
</section>
<section id="all-steps-with-one-command">
<h2>All steps with one command<a class="headerlink" href="#all-steps-with-one-command" title="Permalink to this headline"></a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">all</span></code> accepts FASTQ, SAM/BAM/CRAM, relate/mask/cluster report, and
table files and directories as inputs.</p>
</div>
<p>From BAM, report, and/or table file(s):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="nb">all</span> <span class="n">refs</span><span class="o">.</span><span class="n">fa</span> <span class="n">out</span><span class="o">/</span><span class="n">sample</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">Ref</span><span class="o">.</span><span class="n">bam</span> <span class="n">out</span><span class="o">/</span><span class="n">sample</span><span class="o">/*/</span><span class="n">report</span><span class="o">-*.</span><span class="n">json</span> <span class="n">out</span><span class="o">/</span><span class="n">sample</span><span class="o">/</span><span class="n">table</span><span class="o">/*/*.</span><span class="n">csv</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only the align, relate, mask, and table steps run by default. Enable
clustering by specifying <code class="docutils literal notranslate"><span class="pre">--max-clusters</span></code> (<code class="docutils literal notranslate"><span class="pre">-k</span></code>) followed by the
maximum number of clusters to attempt. Enable structure prediction
with the flag <code class="docutils literal notranslate"><span class="pre">--fold</span></code>.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../index.html" class="btn btn-neutral float-left" title="Manuals" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../inputs.html" class="btn btn-neutral float-right" title="Input file patterns" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>